{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "d993aec1-bc38-4621-a4b5-bce9826fdfd7",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Load libraries\n",
    "import os\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from ast import literal_eval\n",
    "from tqdm import tqdm\n",
    "import random\n",
    "import tensorflow as tf\n",
    "import re\n",
    "import matplotlib\n",
    "from matplotlib import pyplot as plt\n",
    "import sys\n",
    "import copy\n",
    "matplotlib.rcParams['pdf.fonttype'] = 42\n",
    "matplotlib.rcParams['ps.fonttype'] = 42\n",
    "\n",
    "import tensorflow as tf\n",
    "from tensorflow.keras.models import Sequential\n",
    "from tensorflow.keras.layers import Dense, Conv2D, Dropout, Flatten, MaxPooling2D\n",
    "#from tensorflow.python.ops.numpy_ops import np_config\n",
    "#np_config.enable_numpy_behavior()\n",
    "from tensorflow.keras.regularizers import l2\n",
    "from tensorflow.keras.optimizers import Adam\n",
    "from tensorflow.keras.layers import Embedding, LSTM, Dense, Dropout, InputLayer, Reshape\n",
    "from sklearn.metrics import roc_curve, auc, average_precision_score, matthews_corrcoef, f1_score, precision_score\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "c30e12d8-6dfc-4d97-b33c-9af87e77a924",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Add your PTM amino acids\n",
    "PR=\"S\"\n",
    "PR2=\"T\"\n",
    "sym=\"@\"\n",
    "sym2=\"&\"\n",
    "#If only have modification at one site then leave PR2 and sym2 blank i.e. \"\"\n",
    "#Add folder\n",
    "folder=f\"Example_data\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "ebbca413-dc50-4e3c-9225-508177740cd3",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\Clair\\AppData\\Local\\Temp\\ipykernel_32740\\2230625363.py:13: ParserWarning: Falling back to the 'python' engine because the 'c' engine does not support regex separators (separators > 1 char and different from '\\s+' are interpreted as regex); you can avoid this warning by specifying engine='python'.\n",
      "  df = pd.read_csv(file_name, sep='delimiter', header=None)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ID array has been made!\n",
      "Lines have been separated!\n",
      "protein_dict has been made!\n"
     ]
    }
   ],
   "source": [
    "#Specify a reference proteome and generate dictionay\n",
    "\n",
    "######################################################################################################################\n",
    "#\n",
    "#Reference proteome should contain sequences for all the proteins in your dataset. In the case of an organism specific dataset\n",
    "#it could also be useful to have all the proteins for the organism. \n",
    "\n",
    "file_name='20240529_Ref_seq_Blazev.fasta' #This is an example reference proteome which contains all human proteins in their common isoform. \n",
    "\n",
    "#Reference proteome should be in FASTA format in a text file, where the name for each sequence is: >sp|UniprotID|Mnemonic_ID opt_info\n",
    "#Refernece proteomes or ID mappings of Uniprot IDs (esp. if using multiple organism proteins) are easy to get off of Uniprot\n",
    "\n",
    "df = pd.read_csv(file_name, sep='delimiter', header=None)\n",
    "\n",
    "#Load all possible Uniprot IDs from reference proteome\n",
    "df.fillna(\"NA\", inplace=True) # just in case any \"NA\" got turned into NaNs\n",
    "ID_df = df[df[0].str.contains(\">\")][0]\n",
    "ID_list = []\n",
    "for i in range(len(ID_df)):\n",
    "    ID_list.append(ID_df.iloc[i].split(\"|\")[1]) #Change to index 2 if using Mnemonic_ID instead of UniprotID\n",
    "ID_array = np.array(ID_list)\n",
    "print(\"ID array has been made!\")\n",
    "\n",
    "#Makes array of sequences\n",
    "sequences = []\n",
    "indices_of_uniprot_ids = list(df[df[0].str.contains(\">\")].index)\n",
    "for i in range(len(indices_of_uniprot_ids)-1):\n",
    "    # concatenate all into one string\n",
    "    sequences.append(df[indices_of_uniprot_ids[i]+1:indices_of_uniprot_ids[i+1]][0].sum())\n",
    "    if i == len(indices_of_uniprot_ids)-2:\n",
    "        sequences.append(df[indices_of_uniprot_ids[i+1]+1:][0].sum())\n",
    "sequences = np.array(sequences)\n",
    "print(\"Lines have been separated!\")\n",
    "\n",
    "# Creates dictionary that maps Uniprot ID to sequence\n",
    "pairs = zip(ID_array, sequences)\n",
    "protein_dict = {A:B for A,B in pairs}\n",
    "print(\"protein_dict has been made!\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "a51a2478-131c-476e-ad07-ac0594d8bc87",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\Clair\\anaconda3\\envs\\tf\\lib\\site-packages\\pandas\\core\\base.py:662: RuntimeWarning: invalid value encountered in cast\n",
      "  result = np.asarray(values, dtype=dtype)\n"
     ]
    }
   ],
   "source": [
    "#Specify known sites of modification\n",
    "#See example excel sheet for format\n",
    "excel_sheet=\"20240529_Blazev_2022_excercise_dataset.xlsx\"\n",
    "\n",
    "sheet = pd.read_excel(excel_sheet, sheet_name='Spre')\n",
    "Data=sheet[[\"Uniprot ID\", \"Position\"]]\n",
    "\n",
    "IDs=Data.iloc[:, 0] \n",
    "IDs=IDs.to_numpy(dtype=str)\n",
    "\n",
    "Position=Data.iloc[:, 1] \n",
    "Position=Position.to_numpy(dtype=int)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "68e84155-10fa-48c6-8ef3-e7f020f531bb",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading...: 100%|██████████████████████████████████████████████████████████████| 8284/8284 [00:00<00:00, 102228.16it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "H7BXK9 not in protein_dict\n",
      "A0A0A0MRL6 not in protein_dict\n",
      "Site outside of seq bounds, site: 458, sequnce length: 455\n",
      "A0A0A0MRL6\n",
      "Site outside of seq bounds, site: 608, sequnce length: 455\n",
      "A0A0A0MRL6\n",
      "Site outside of seq bounds, site: 623, sequnce length: 455\n",
      "A0A0A0MRL6\n",
      "Site outside of seq bounds, site: 470, sequnce length: 455\n",
      "A0A0A0MRL6\n",
      "A0A087X0W7 not in protein_dict\n",
      "A0A087X1U1 not in protein_dict\n",
      "M0QYL4 not in protein_dict\n",
      "Site outside of seq bounds, site: 1390, sequnce length: 383\n",
      "Q96L96\n",
      "Site outside of seq bounds, site: 1394, sequnce length: 383\n",
      "Q96L96\n",
      "Site outside of seq bounds, site: 1401, sequnce length: 383\n",
      "Q96L96\n",
      "Site outside of seq bounds, site: 1406, sequnce length: 383\n",
      "Q96L96\n",
      "Site outside of seq bounds, site: 1424, sequnce length: 383\n",
      "Q96L96\n",
      "Site outside of seq bounds, site: 1855, sequnce length: 383\n",
      "Q96L96\n",
      "Site outside of seq bounds, site: 1865, sequnce length: 383\n",
      "Q96L96\n",
      "Site outside of seq bounds, site: 430, sequnce length: 383\n",
      "Q96L96\n",
      "Site outside of seq bounds, site: 658, sequnce length: 383\n",
      "Q96L96\n",
      "Site outside of seq bounds, site: 1190, sequnce length: 383\n",
      "Q96L96\n",
      "Site outside of seq bounds, site: 1383, sequnce length: 383\n",
      "Q96L96\n",
      "Site outside of seq bounds, site: 1384, sequnce length: 383\n",
      "Q96L96\n",
      "Site outside of seq bounds, site: 673, sequnce length: 383\n",
      "Q96L96\n",
      "Site outside of seq bounds, site: 871, sequnce length: 383\n",
      "Q96L96\n",
      "A0A2R8Y611 not in protein_dict\n",
      "A0A087WUX8 not in protein_dict\n",
      "H0Y7S3 not in protein_dict\n",
      "J3KN59 not in protein_dict\n",
      "Site outside of seq bounds, site: 279, sequnce length: 43\n",
      "D6RFH8\n",
      "REV__Q6ZUS5 not in protein_dict\n",
      "REV__E5RG94 not in protein_dict\n",
      "J3KN18 not in protein_dict\n",
      "E9PR44 not in protein_dict\n",
      "A0A087WUB9 not in protein_dict\n",
      "H0Y3A3 not in protein_dict\n",
      "A0A087X0H7 not in protein_dict\n",
      "A0A087WYN9 not in protein_dict\n",
      "E7EPU2 not in protein_dict\n",
      "Site outside of seq bounds, site: 245, sequnce length: 225\n",
      "E7EPU2\n",
      "G8JLD5 not in protein_dict\n",
      "A0A0A0MSY4 not in protein_dict\n",
      "A0A087WUT6 not in protein_dict\n",
      "H0YL12 not in protein_dict\n",
      "A0A087WY82 not in protein_dict\n",
      "Site outside of seq bounds, site: 373, sequnce length: 275\n",
      "H7C0W7\n",
      "A0A0A0MR12 not in protein_dict\n",
      "A0A1W2PRQ0 not in protein_dict\n",
      "A0A0A0MT60 not in protein_dict\n",
      "Site outside of seq bounds, site: 1186, sequnce length: 37\n",
      "A0A0A0MT60\n",
      "A0A0A0MTS2 not in protein_dict\n",
      "Site outside of seq bounds, site: 470, sequnce length: 280\n",
      "A0A0A0MTS2\n",
      "Site outside of seq bounds, site: 547, sequnce length: 280\n",
      "A0A0A0MTS2\n",
      "Site outside of seq bounds, site: 552, sequnce length: 280\n",
      "A0A0A0MTS2\n",
      "Site outside of seq bounds, site: 551, sequnce length: 280\n",
      "A0A0A0MTS2\n",
      "O75367-2 not in protein_dict\n",
      "Site outside of seq bounds, site: 313, sequnce length: 213\n",
      "O75367-2\n",
      "E7EVH9 not in protein_dict\n",
      "Site outside of seq bounds, site: 1571, sequnce length: 205\n",
      "A0A087X2H1\n",
      "Site outside of seq bounds, site: 1575, sequnce length: 205\n",
      "A0A087X2H1\n",
      "Site outside of seq bounds, site: 1776, sequnce length: 205\n",
      "A0A087X2H1\n",
      "Site outside of seq bounds, site: 632, sequnce length: 205\n",
      "A0A087X2H1\n",
      "E5RGH4 not in protein_dict\n",
      "Q5TCC6 not in protein_dict\n",
      "A0A087WWK8 not in protein_dict\n",
      "Site outside of seq bounds, site: 940, sequnce length: 814\n",
      "A0A087WWK8\n",
      "Site outside of seq bounds, site: 941, sequnce length: 814\n",
      "A0A087WWK8\n",
      "Q6NY19-2 not in protein_dict\n",
      "A0A0A0MSC0 not in protein_dict\n",
      "Site outside of seq bounds, site: 822, sequnce length: 746\n",
      "A0A0A0MSC0\n",
      "H0YI08 not in protein_dict\n",
      "A0A087WWJ1 not in protein_dict\n",
      "A0A087X023 not in protein_dict\n",
      "Site outside of seq bounds, site: 2645, sequnce length: 1141\n",
      "A0A499FJI4\n",
      "A0A087X0T3 not in protein_dict\n",
      "Q15170-2 not in protein_dict\n",
      "A0A1B0GVK0 not in protein_dict\n",
      "A0A088AWL3 not in protein_dict\n",
      "Site outside of seq bounds, site: 2088, sequnce length: 249\n",
      "A0A088AWL3\n",
      "P20929-2 not in protein_dict\n",
      "Site outside of seq bounds, site: 1243, sequnce length: 473\n",
      "P20929-2\n",
      "Site outside of seq bounds, site: 1275, sequnce length: 473\n",
      "P20929-2\n",
      "Site outside of seq bounds, site: 1420, sequnce length: 473\n",
      "P20929-2\n",
      "Site outside of seq bounds, site: 1652, sequnce length: 473\n",
      "P20929-2\n",
      "Site outside of seq bounds, site: 1836, sequnce length: 473\n",
      "P20929-2\n",
      "Site outside of seq bounds, site: 1908, sequnce length: 473\n",
      "P20929-2\n",
      "Site outside of seq bounds, site: 2031, sequnce length: 473\n",
      "P20929-2\n",
      "Site outside of seq bounds, site: 2209, sequnce length: 473\n",
      "P20929-2\n",
      "Site outside of seq bounds, site: 2359, sequnce length: 473\n",
      "P20929-2\n",
      "Site outside of seq bounds, site: 2944, sequnce length: 473\n",
      "P20929-2\n",
      "Site outside of seq bounds, site: 2948, sequnce length: 473\n",
      "P20929-2\n",
      "Site outside of seq bounds, site: 3017, sequnce length: 473\n",
      "P20929-2\n",
      "Site outside of seq bounds, site: 3172, sequnce length: 473\n",
      "P20929-2\n",
      "Site outside of seq bounds, site: 3485, sequnce length: 473\n",
      "P20929-2\n",
      "Site outside of seq bounds, site: 3673, sequnce length: 473\n",
      "P20929-2\n",
      "Site outside of seq bounds, site: 3948, sequnce length: 473\n",
      "P20929-2\n",
      "Site outside of seq bounds, site: 4025, sequnce length: 473\n",
      "P20929-2\n",
      "Site outside of seq bounds, site: 4111, sequnce length: 473\n",
      "P20929-2\n",
      "Site outside of seq bounds, site: 4156, sequnce length: 473\n",
      "P20929-2\n",
      "Site outside of seq bounds, site: 4159, sequnce length: 473\n",
      "P20929-2\n",
      "Site outside of seq bounds, site: 4173, sequnce length: 473\n",
      "P20929-2\n",
      "Site outside of seq bounds, site: 4177, sequnce length: 473\n",
      "P20929-2\n",
      "Site outside of seq bounds, site: 4309, sequnce length: 473\n",
      "P20929-2\n",
      "Site outside of seq bounds, site: 4402, sequnce length: 473\n",
      "P20929-2\n",
      "Site outside of seq bounds, site: 4420, sequnce length: 473\n",
      "P20929-2\n",
      "Site outside of seq bounds, site: 4455, sequnce length: 473\n",
      "P20929-2\n",
      "Site outside of seq bounds, site: 4545, sequnce length: 473\n",
      "P20929-2\n",
      "Site outside of seq bounds, site: 508, sequnce length: 473\n",
      "P20929-2\n",
      "Site outside of seq bounds, site: 518, sequnce length: 473\n",
      "P20929-2\n",
      "Site outside of seq bounds, site: 5629, sequnce length: 473\n",
      "P20929-2\n",
      "Site outside of seq bounds, site: 5649, sequnce length: 473\n",
      "P20929-2\n",
      "Site outside of seq bounds, site: 5760, sequnce length: 473\n",
      "P20929-2\n",
      "Site outside of seq bounds, site: 5844, sequnce length: 473\n",
      "P20929-2\n",
      "Site outside of seq bounds, site: 6117, sequnce length: 473\n",
      "P20929-2\n",
      "Site outside of seq bounds, site: 613, sequnce length: 473\n",
      "P20929-2\n",
      "Site outside of seq bounds, site: 6139, sequnce length: 473\n",
      "P20929-2\n",
      "Site outside of seq bounds, site: 6421, sequnce length: 473\n",
      "P20929-2\n",
      "Site outside of seq bounds, site: 6422, sequnce length: 473\n",
      "P20929-2\n",
      "Site outside of seq bounds, site: 6647, sequnce length: 473\n",
      "P20929-2\n",
      "Site outside of seq bounds, site: 6648, sequnce length: 473\n",
      "P20929-2\n",
      "Site outside of seq bounds, site: 6759, sequnce length: 473\n",
      "P20929-2\n",
      "Site outside of seq bounds, site: 684, sequnce length: 473\n",
      "P20929-2\n",
      "Site outside of seq bounds, site: 6871, sequnce length: 473\n",
      "P20929-2\n",
      "Site outside of seq bounds, site: 6875, sequnce length: 473\n",
      "P20929-2\n",
      "Site outside of seq bounds, site: 701, sequnce length: 473\n",
      "P20929-2\n",
      "Site outside of seq bounds, site: 7015, sequnce length: 473\n",
      "P20929-2\n",
      "Site outside of seq bounds, site: 7089, sequnce length: 473\n",
      "P20929-2\n",
      "Site outside of seq bounds, site: 7222, sequnce length: 473\n",
      "P20929-2\n",
      "Site outside of seq bounds, site: 7365, sequnce length: 473\n",
      "P20929-2\n",
      "Site outside of seq bounds, site: 7404, sequnce length: 473\n",
      "P20929-2\n",
      "Site outside of seq bounds, site: 7515, sequnce length: 473\n",
      "P20929-2\n",
      "Site outside of seq bounds, site: 755, sequnce length: 473\n",
      "P20929-2\n",
      "Site outside of seq bounds, site: 7560, sequnce length: 473\n",
      "P20929-2\n",
      "Site outside of seq bounds, site: 7682, sequnce length: 473\n",
      "P20929-2\n",
      "Site outside of seq bounds, site: 7763, sequnce length: 473\n",
      "P20929-2\n",
      "Site outside of seq bounds, site: 7792, sequnce length: 473\n",
      "P20929-2\n",
      "Site outside of seq bounds, site: 7860, sequnce length: 473\n",
      "P20929-2\n",
      "Site outside of seq bounds, site: 7944, sequnce length: 473\n",
      "P20929-2\n",
      "Site outside of seq bounds, site: 7945, sequnce length: 473\n",
      "P20929-2\n",
      "Site outside of seq bounds, site: 8254, sequnce length: 473\n",
      "P20929-2\n",
      "Site outside of seq bounds, site: 8255, sequnce length: 473\n",
      "P20929-2\n",
      "Site outside of seq bounds, site: 8261, sequnce length: 473\n",
      "P20929-2\n",
      "Site outside of seq bounds, site: 8325, sequnce length: 473\n",
      "P20929-2\n",
      "Site outside of seq bounds, site: 8363, sequnce length: 473\n",
      "P20929-2\n",
      "Site outside of seq bounds, site: 8378, sequnce length: 473\n",
      "P20929-2\n",
      "Site outside of seq bounds, site: 8396, sequnce length: 473\n",
      "P20929-2\n",
      "Site outside of seq bounds, site: 8398, sequnce length: 473\n",
      "P20929-2\n",
      "Site outside of seq bounds, site: 8401, sequnce length: 473\n",
      "P20929-2\n",
      "Site outside of seq bounds, site: 8403, sequnce length: 473\n",
      "P20929-2\n",
      "Site outside of seq bounds, site: 8409, sequnce length: 473\n",
      "P20929-2\n",
      "Site outside of seq bounds, site: 8412, sequnce length: 473\n",
      "P20929-2\n",
      "Site outside of seq bounds, site: 8420, sequnce length: 473\n",
      "P20929-2\n",
      "Site outside of seq bounds, site: 8423, sequnce length: 473\n",
      "P20929-2\n",
      "Site outside of seq bounds, site: 8431, sequnce length: 473\n",
      "P20929-2\n",
      "Site outside of seq bounds, site: 8449, sequnce length: 473\n",
      "P20929-2\n",
      "Site outside of seq bounds, site: 8458, sequnce length: 473\n",
      "P20929-2\n",
      "Site outside of seq bounds, site: 8459, sequnce length: 473\n",
      "P20929-2\n",
      "Site outside of seq bounds, site: 8462, sequnce length: 473\n",
      "P20929-2\n",
      "Site outside of seq bounds, site: 8465, sequnce length: 473\n",
      "P20929-2\n",
      "Site outside of seq bounds, site: 8486, sequnce length: 473\n",
      "P20929-2\n",
      "Site outside of seq bounds, site: 860, sequnce length: 473\n",
      "P20929-2\n",
      "Site outside of seq bounds, site: 976, sequnce length: 473\n",
      "P20929-2\n",
      "Site outside of seq bounds, site: 995, sequnce length: 473\n",
      "P20929-2\n",
      "Site outside of seq bounds, site: 2254, sequnce length: 473\n",
      "P20929-2\n",
      "Site outside of seq bounds, site: 2946, sequnce length: 473\n",
      "P20929-2\n",
      "Site outside of seq bounds, site: 2983, sequnce length: 473\n",
      "P20929-2\n",
      "Site outside of seq bounds, site: 3947, sequnce length: 473\n",
      "P20929-2\n",
      "Site outside of seq bounds, site: 4158, sequnce length: 473\n",
      "P20929-2\n",
      "Site outside of seq bounds, site: 4163, sequnce length: 473\n",
      "P20929-2\n",
      "Site outside of seq bounds, site: 4198, sequnce length: 473\n",
      "P20929-2\n",
      "Site outside of seq bounds, site: 4406, sequnce length: 473\n",
      "P20929-2\n",
      "Site outside of seq bounds, site: 4441, sequnce length: 473\n",
      "P20929-2\n",
      "Site outside of seq bounds, site: 5648, sequnce length: 473\n",
      "P20929-2\n",
      "Site outside of seq bounds, site: 5858, sequnce length: 473\n",
      "P20929-2\n",
      "Site outside of seq bounds, site: 6128, sequnce length: 473\n",
      "P20929-2\n",
      "Site outside of seq bounds, site: 6739, sequnce length: 473\n",
      "P20929-2\n",
      "Site outside of seq bounds, site: 6949, sequnce length: 473\n",
      "P20929-2\n",
      "Site outside of seq bounds, site: 6970, sequnce length: 473\n",
      "P20929-2\n",
      "Site outside of seq bounds, site: 7559, sequnce length: 473\n",
      "P20929-2\n",
      "Site outside of seq bounds, site: 7894, sequnce length: 473\n",
      "P20929-2\n",
      "Site outside of seq bounds, site: 7930, sequnce length: 473\n",
      "P20929-2\n",
      "Site outside of seq bounds, site: 7987, sequnce length: 473\n",
      "P20929-2\n",
      "Site outside of seq bounds, site: 7992, sequnce length: 473\n",
      "P20929-2\n",
      "Site outside of seq bounds, site: 8142, sequnce length: 473\n",
      "P20929-2\n",
      "Site outside of seq bounds, site: 8147, sequnce length: 473\n",
      "P20929-2\n",
      "Site outside of seq bounds, site: 8204, sequnce length: 473\n",
      "P20929-2\n",
      "Site outside of seq bounds, site: 8209, sequnce length: 473\n",
      "P20929-2\n",
      "Site outside of seq bounds, site: 8235, sequnce length: 473\n",
      "P20929-2\n",
      "Site outside of seq bounds, site: 8240, sequnce length: 473\n",
      "P20929-2\n",
      "Site outside of seq bounds, site: 8276, sequnce length: 473\n",
      "P20929-2\n",
      "Site outside of seq bounds, site: 8306, sequnce length: 473\n",
      "P20929-2\n",
      "Site outside of seq bounds, site: 8310, sequnce length: 473\n",
      "P20929-2\n",
      "Site outside of seq bounds, site: 8350, sequnce length: 473\n",
      "P20929-2\n",
      "Site outside of seq bounds, site: 8421, sequnce length: 473\n",
      "P20929-2\n",
      "Site outside of seq bounds, site: 8439, sequnce length: 473\n",
      "P20929-2\n",
      "Site outside of seq bounds, site: 8452, sequnce length: 473\n",
      "P20929-2\n",
      "Site outside of seq bounds, site: 8456, sequnce length: 473\n",
      "P20929-2\n",
      "Site outside of seq bounds, site: 1818, sequnce length: 473\n",
      "P20929-2\n",
      "Site outside of seq bounds, site: 2549, sequnce length: 473\n",
      "P20929-2\n",
      "Site outside of seq bounds, site: 3035, sequnce length: 473\n",
      "P20929-2\n",
      "Site outside of seq bounds, site: 3278, sequnce length: 473\n",
      "P20929-2\n",
      "Site outside of seq bounds, site: 554, sequnce length: 473\n",
      "P20929-2\n",
      "Site outside of seq bounds, site: 6262, sequnce length: 473\n",
      "P20929-2\n",
      "Site outside of seq bounds, site: 6679, sequnce length: 473\n",
      "P20929-2\n",
      "Site outside of seq bounds, site: 7097, sequnce length: 473\n",
      "P20929-2\n",
      "Site outside of seq bounds, site: 801, sequnce length: 473\n",
      "P20929-2\n",
      "Site outside of seq bounds, site: 8367, sequnce length: 473\n",
      "P20929-2\n",
      "Site outside of seq bounds, site: 8422, sequnce length: 473\n",
      "P20929-2\n",
      "Site outside of seq bounds, site: 8477, sequnce length: 473\n",
      "P20929-2\n",
      "A0A087WYW3 not in protein_dict\n",
      "C9J5P6 not in protein_dict\n",
      "F5H7V1 not in protein_dict\n",
      "A0A087WY61 not in protein_dict\n",
      "Site outside of seq bounds, site: 1741, sequnce length: 695\n",
      "A0A087WY61\n",
      "Site outside of seq bounds, site: 1984, sequnce length: 695\n",
      "A0A087WY61\n",
      "A0A087WU39 not in protein_dict\n",
      "F5H1M8 not in protein_dict\n",
      "A0A0A0MRM1 not in protein_dict\n",
      "Site outside of seq bounds, site: 2113, sequnce length: 687\n",
      "A0A0A0MRM1\n",
      "F6VCX5 not in protein_dict\n",
      "A0A0D9SF60 not in protein_dict\n",
      "H0YJ75 not in protein_dict\n",
      "A0A087X0I9 not in protein_dict\n",
      "F5H7R9 not in protein_dict\n",
      "H0Y4Z8 not in protein_dict\n",
      "Site outside of seq bounds, site: 641, sequnce length: 504\n",
      "H0Y4Z8\n",
      "Site outside of seq bounds, site: 600, sequnce length: 504\n",
      "H0Y4Z8\n",
      "A0A0D9SEU5 not in protein_dict\n",
      "nan not in protein_dict\n",
      "A0A087WX02 not in protein_dict\n",
      "H0Y507 not in protein_dict\n",
      "Site outside of seq bounds, site: 1227, sequnce length: 729\n",
      "A0A0U1RR93\n",
      "Site outside of seq bounds, site: 919, sequnce length: 67\n",
      "I3L4N6\n",
      "A0A087X2D8 not in protein_dict\n",
      "Site outside of seq bounds, site: 719, sequnce length: 267\n",
      "A0A087X2D8\n",
      "Site outside of seq bounds, site: 721, sequnce length: 267\n",
      "A0A087X2D8\n",
      "Site outside of seq bounds, site: 724, sequnce length: 267\n",
      "A0A087X2D8\n",
      "Site outside of seq bounds, site: 586, sequnce length: 267\n",
      "A0A087X2D8\n",
      "A0A087WUZ3 not in protein_dict\n",
      "Site outside of seq bounds, site: 2140, sequnce length: 2137\n",
      "A0A087WUZ3\n",
      "Site outside of seq bounds, site: 2167, sequnce length: 2137\n",
      "A0A087WUZ3\n",
      "Site outside of seq bounds, site: 2171, sequnce length: 2137\n",
      "A0A087WUZ3\n",
      "Site outside of seq bounds, site: 2342, sequnce length: 2137\n",
      "A0A087WUZ3\n",
      "Site outside of seq bounds, site: 2343, sequnce length: 2137\n",
      "A0A087WUZ3\n",
      "Site outside of seq bounds, site: 2360, sequnce length: 2137\n",
      "A0A087WUZ3\n",
      "Site outside of seq bounds, site: 509, sequnce length: 508\n",
      "Q86TD4\n",
      "Site outside of seq bounds, site: 513, sequnce length: 508\n",
      "Q86TD4\n",
      "Site outside of seq bounds, site: 526, sequnce length: 508\n",
      "Q86TD4\n",
      "Site outside of seq bounds, site: 533, sequnce length: 508\n",
      "Q86TD4\n",
      "Site outside of seq bounds, site: 551, sequnce length: 508\n",
      "Q86TD4\n",
      "Site outside of seq bounds, site: 591, sequnce length: 508\n",
      "Q86TD4\n",
      "Site outside of seq bounds, site: 614, sequnce length: 508\n",
      "Q86TD4\n",
      "Site outside of seq bounds, site: 870, sequnce length: 508\n",
      "Q86TD4\n",
      "Site outside of seq bounds, site: 871, sequnce length: 508\n",
      "Q86TD4\n",
      "Site outside of seq bounds, site: 904, sequnce length: 508\n",
      "Q86TD4\n",
      "Site outside of seq bounds, site: 550, sequnce length: 508\n",
      "Q86TD4\n",
      "Site outside of seq bounds, site: 512, sequnce length: 508\n",
      "Q86TD4\n",
      "Site outside of seq bounds, site: 580, sequnce length: 508\n",
      "Q86TD4\n",
      "Site outside of seq bounds, site: 838, sequnce length: 508\n",
      "Q86TD4\n",
      "A0A0A0MTD1 not in protein_dict\n",
      "E7ENN3 not in protein_dict\n",
      "Site outside of seq bounds, site: 7900, sequnce length: 476\n",
      "E7ENN3\n",
      "Site outside of seq bounds, site: 7955, sequnce length: 476\n",
      "E7ENN3\n",
      "J3KQ96 not in protein_dict\n",
      "Site outside of seq bounds, site: 1312, sequnce length: 200\n",
      "J3KQ96\n",
      "Site outside of seq bounds, site: 1340, sequnce length: 200\n",
      "J3KQ96\n",
      "Site outside of seq bounds, site: 1278, sequnce length: 482\n",
      "G3V1L9\n",
      "Site outside of seq bounds, site: 1577, sequnce length: 482\n",
      "G3V1L9\n",
      "Site outside of seq bounds, site: 1579, sequnce length: 482\n",
      "G3V1L9\n",
      "Site outside of seq bounds, site: 1617, sequnce length: 482\n",
      "G3V1L9\n",
      "Site outside of seq bounds, site: 1619, sequnce length: 482\n",
      "G3V1L9\n",
      "H0YLT6 not in protein_dict\n",
      "Site outside of seq bounds, site: 617, sequnce length: 482\n",
      "G3V1L9\n",
      "Site outside of seq bounds, site: 912, sequnce length: 482\n",
      "G3V1L9\n",
      "Site outside of seq bounds, site: 927, sequnce length: 482\n",
      "G3V1L9\n",
      "Site outside of seq bounds, site: 968, sequnce length: 482\n",
      "G3V1L9\n",
      "A0A087WT47 not in protein_dict\n",
      "A0A087WXS0 not in protein_dict\n",
      "C9JM61 not in protein_dict\n",
      "A0A087X0I0 not in protein_dict\n",
      "P60174-1 not in protein_dict\n",
      "Site outside of seq bounds, site: 195, sequnce length: 62\n",
      "P60174-1\n",
      "Site outside of seq bounds, site: 198, sequnce length: 62\n",
      "P60174-1\n",
      "Site outside of seq bounds, site: 204, sequnce length: 62\n",
      "P60174-1\n",
      "Site outside of seq bounds, site: 212, sequnce length: 62\n",
      "P60174-1\n",
      "Site outside of seq bounds, site: 223, sequnce length: 62\n",
      "P60174-1\n",
      "Site outside of seq bounds, site: 80, sequnce length: 62\n",
      "P60174-1\n",
      "Site outside of seq bounds, site: 178, sequnce length: 62\n",
      "P60174-1\n",
      "Site outside of seq bounds, site: 214, sequnce length: 62\n",
      "P60174-1\n",
      "Site outside of seq bounds, site: 71, sequnce length: 62\n",
      "P60174-1\n",
      "Site outside of seq bounds, site: 76, sequnce length: 62\n",
      "P60174-1\n",
      "Site outside of seq bounds, site: 68, sequnce length: 62\n",
      "P60174-1\n",
      "Site outside of seq bounds, site: 72, sequnce length: 23\n",
      "A0A087WV45\n",
      "A0A1W2PR67 not in protein_dict\n",
      "A0A087WTB8 not in protein_dict\n",
      "H7C1J4 not in protein_dict\n",
      "Site outside of seq bounds, site: 1111, sequnce length: 794\n",
      "H7C1J4\n",
      "Site outside of seq bounds, site: 922, sequnce length: 794\n",
      "H7C1J4\n",
      "Site outside of seq bounds, site: 1101, sequnce length: 794\n",
      "H7C1J4\n",
      "G8JLK1 not in protein_dict\n",
      "Q93008-1 not in protein_dict\n",
      "Site outside of seq bounds, site: 2443, sequnce length: 1012\n",
      "Q93008-1\n"
     ]
    }
   ],
   "source": [
    "PR_site_array=[]\n",
    "PR2_site_array=[]\n",
    "for i in tqdm(range(len(IDs)), desc=\"Loading...\", position=0, leave=True):\n",
    "    ID=IDs[i]\n",
    "    location=int(Position[i])\n",
    "    try:\n",
    "        seq=protein_dict[ID]\n",
    "        if location > len(seq): #Will not add to list of kmers\n",
    "            print (\"Site outside of seq bounds, site: \"+str(location)+\", sequnce length: \"+str(len(seq)))\n",
    "            print (ID)\n",
    "            newseq=seq\n",
    "        elif seq[location-1] == PR or seq[location-1]==sym:\n",
    "            PR_site_array.append ((ID, location))\n",
    "            newseq=seq[:location-1] + sym + seq[location:]\n",
    "        elif PR2 != \"\":\n",
    "            if seq[int(location-1)] == PR2 or seq[int(location-1)]==sym2:\n",
    "                PR2_site_array.append ((ID, location))\n",
    "                newseq=seq[:location-1] + sym2 + seq[location:]\n",
    "        else:\n",
    "            print(seq[location-1])\n",
    "            print (\"Sequence key error for \"+ID+\" at \" + str(location)) #Will not add it to list of sites or dictionary\n",
    "            newseq=seq\n",
    "    except KeyError:\n",
    "        print(ID+\" not in protein_dict\")    \n",
    "    protein_dict.update({ID: newseq})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "a95b5dae-4894-4230-91a1-4b11985d917f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "MKKFFDSRREQGGSGLGSGS@GGGGSTSGLGSGYIGRVFGIGRQQVTVDEVLAEGGFAIVFLVRTSNGMKCALKRMFVNNEHDLQVCKREIQIMRDLSGHKNIVGYIDSSINNVSSGDVWEVLILMDFCRGGQVVNLMNQRLQTGFTENEVLQIFCDTCEAVARLHQCKTPIIHRDLKVENILLHDRGHYVLCDFGSATNKFQNPQTEGVNAVEDEIKKYTTLSYRAPEMVNLYSGKIITTKADIWALGCLLYKLCYFTLPFGESQVAICDGNFTIPDNSRYSQDMHCLIRYMLEPDPDKRPDIYQVSYFSFKLLKKECPIPNVQNSPIPAKLPEPVKASEAAAKKTQPKARLTDPIPTTETSIAPRQRPKAGQTQPNPGILPIQPALTPRKRATVQPPPQAAGSSNQPGLLASVPQPKPQAPPSQPLPQTQAKQPQAPPTPQQTPSTQAQGLPAQAQATPQHQQQLFLKQQQQQQQPPPAQQQPAGTFYQQQQAQTQQFQAVHPATQKPAIAQFPVVSQGGSQQQLMQNFYQQQQQQQQQQQQQQLATALHQQQLMTQQAALQQKPTMAAGQQPQPQPAAAPQPAPAQEPAIQAPVRQQPKVQ&&PPPAVQGQKVGSL&PPS@PKTQRAGHRRILSDVTHSAVFGVPASKSTQLLQAAAAEASLNKSKSATTTPSGSPRTSQQNVYNPSEGSTWNPFDDDNFSKLTAEELLNKDFAKLGEGKHPEKLGGSAESLIPGFQSTQGDAFATTSFSAGTAEKRKGGQTVDSGLPLLSVSDPFIPLQVPDAPEKLIEGLKSPDTSLLLPDLLPMTDPFGSTSDAVIEKADVAVESLIPGLEPPVPQRLP@QTESVTSNRTDSLTGEDSLLDCSLLSNPTTDLLEEFAPTAISAPVHKAAEDSNLISGFDVPEGSDKVAEDEFDPIPVLITKNPQGGHSRNSSGSSESSLPNLARSLLLVDQLIDL\n",
      "Number of PR sites: 335259\n",
      "PR sites saved\n",
      "Number of PR2 sites: 220178\n",
      "PR2 sites saved\n"
     ]
    }
   ],
   "source": [
    "#Now we need a list of all the S and T sites in the Uniprot file\n",
    "#These will be the negatives as the positives have been been changed from PR, PR2 to sym, sym2 so they won't appear\n",
    "print (protein_dict['Q2M2I8'])\n",
    "PR_list=[]\n",
    "PR2_list=[]\n",
    "for key in protein_dict:\n",
    "    seq=protein_dict[key]\n",
    "    for match in re.finditer(PR, seq):\n",
    "        PR_list.append((key, match.start()+1))\n",
    "\n",
    "print (\"Number of PR sites: \"+str(len(PR_list)))\n",
    "np.save(f'{folder}/All_PR_sites', PR_list)\n",
    "print (\"PR sites saved\")\n",
    "\n",
    "#!!!!!!!!\n",
    "#Don't do if there is no PR2 site\n",
    "for key in protein_dict:\n",
    "    #print (key)\n",
    "    seq=protein_dict[key]\n",
    "    for matchPR2 in re.finditer(PR2, seq):\n",
    "        PR2_list.append((key, matchPR2.start()+1))\n",
    "        #print (matchT.start()+1)\n",
    "\n",
    "print (\"Number of PR2 sites: \"+str(len(PR2_list)))\n",
    "np.save(f'{folder}/All_PR2_sites', PR2_list)\n",
    "print (\"PR2 sites saved\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "24a60b11-890d-466c-a28a-13649c99c2cd",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "44bfd2a1-a303-4984-9ec4-4c42173485a9",
   "metadata": {},
   "outputs": [],
   "source": [
    "################################\n",
    "#Various functions to define\n",
    "###############################\n",
    "\n",
    "#Define function to make kmer from a location and sequence\n",
    "def ndtolot (ndarray):\n",
    "    \"\"\"\n",
    "    converts 2D ndarray (n, 2) to a list of tuples\n",
    "    \"\"\"\n",
    "    list=[]\n",
    "    for i in range (0, len(ndarray[:, 0])):\n",
    "        holding=(ndarray[i, 0], int(ndarray[i, 1]))\n",
    "        list.append(holding)\n",
    "    return list\n",
    "\n",
    "def kmer_from_site(seq, location, k=53, verbose=\"no\"):\n",
    "    \"\"\"\n",
    "    Returns kmer of length k from a location and sequence. Ends are padded with \"-\"\n",
    "\n",
    "    \"\"\"\n",
    "    half=int((k-1)/2)\n",
    "    if location > len(seq): #Will not add to list of kmers\n",
    "        if verbose==\"yes\":\n",
    "            print (\"Site outside of seq bounds, site: \"+str(location)+\", sequnce length: \"+str(len(seq)))\n",
    "        kmer=''\n",
    "    elif location <= half: #To deal with sites near the n terminus\n",
    "        if location > len(seq)-half: #To deal with sequences shorter than k\n",
    "            gap=\"-\"*(half-location+1)\n",
    "            gap2=\"-\"*int(half-(len(seq)-location))\n",
    "            kmer=seq[0 : int(location+half)]\n",
    "            kmer=gap+kmer+gap2\n",
    "        else: \n",
    "            gap=\"-\"*(half-location+1)\n",
    "            kmer=seq[0 : int(location+half)]\n",
    "            kmer=gap+kmer\n",
    "    elif location > len(seq)-half: #To deal with sites near the C terminus\n",
    "        gap=\"-\"*int(half-(len(seq)-location))\n",
    "        kmer=seq[int(location-half-1): len(seq)]\n",
    "        kmer=kmer+gap\n",
    "    else:\n",
    "        kmer=seq[int(location-half-1): int(location+half)]\n",
    "    return kmer\n",
    "    \n",
    "#Define function to make positive and negative kmers\n",
    "def make_kmers(protein_dict, PR_sites, PR2_sites, k=53, verbose=\"no\"):\n",
    "    \"\"\"\n",
    "    Makes positive and negative kmers (with known PTM locations) for a list of sites\n",
    "    Can have up to two site lists for modifications at 2 different amino acids: PR_sites and PR2_sites, PR2_sites can be empty\n",
    "        -Site lists must be a list of tuples of the form (ID, site) where site is the bio indexing (starts from 1)\n",
    "    Protein_dict: dictionary of UniprotIDs with the sequences as keys\n",
    "    k: length of kmer, default is 53\n",
    "    There must be files for all the negatives named: All_PR_sites.npy and All_PR_sites.npy which are made in a previous code cell\n",
    "    verbose: If this is set to \"yes\" then it will print dataset comments, otherwise prints will be off. \"yes\" is helpful for seeing what sequences\n",
    "    are incorrect. \n",
    "    \"\"\"\n",
    "    pos_sites=PR_sites+PR2_sites\n",
    "    pos_kmers=[]\n",
    "    neg_kmers=[]\n",
    "\n",
    "    nPR=ndtolot(np.load(f'{folder}/All_PR_sites.npy', allow_pickle=True))\n",
    "    print (nPR)\n",
    "    nPR2=ndtolot(np.load(f'{folder}/All_PR2_sites.npy', allow_pickle=True))\n",
    "\n",
    "    neg_PR_sites= random.sample(nPR, len(PR_sites))\n",
    "    neg_PR2_sites= random.sample(nPR2, len(PR2_sites))\n",
    "\n",
    "    neg_sites=neg_PR_sites+neg_PR2_sites\n",
    "    print (len(neg_sites))\n",
    "    print (len(pos_sites))\n",
    "    \n",
    "    for i in range(0, len(pos_sites)):\n",
    "        ID=pos_sites[i][0]\n",
    "        location=pos_sites[i][1]\n",
    "        seq=protein_dict[ID]\n",
    "        kmer=kmer_from_site(seq, location, k=k, verbose=verbose)\n",
    "        if location > len(seq): #Will not add to list of kmers\n",
    "            if verbose ==\"yes\":\n",
    "                print (\"Site outside of seq bounds, site: \"+str(location)+\", sequnce length: \"+str(len(seq)))\n",
    "        elif seq[int(location-1)] == PR or seq[int(location-1)]==sym:\n",
    "            pos_kmers.append(kmer)\n",
    "        elif seq[int(location-1)] == PR2 or seq[int(location-1)]==sym2:\n",
    "            pos_kmers.append(kmer)\n",
    "        else:\n",
    "            if verbose==\"yes\":\n",
    "                print(seq[int(location-1)])\n",
    "                print (\"Sequence key error for \"+ID+\" at \" + str(location)) #Will not add it to list of kmers because doesn't match uniprot\n",
    "                #raise Exception(\"Sequence key error for \"+ID+\" at \" + str(location))\n",
    "        if len(kmer) != k:\n",
    "            print (kmer)\n",
    "            print (len(kmer))\n",
    "\n",
    "    for i in range(0, len(neg_sites)):\n",
    "        ID=neg_sites[i][0]\n",
    "        location=neg_sites[i][1]\n",
    "        if verbose==\"yes\":\n",
    "            print (ID)\n",
    "        seq=protein_dict[ID]\n",
    "        kmer=kmer_from_site(seq, location, k=k, verbose=verbose)\n",
    "        if location > len(seq): #Will not add to list of kmers\n",
    "            if verbose ==\"yes\":\n",
    "                print (\"Site outside of seq bounds, site: \"+str(location)+\", sequnce length: \"+str(len(seq)))\n",
    "        elif seq[int(location-1)] == PR or seq[int(location-1)]==sym:\n",
    "            neg_kmers.append(kmer)\n",
    "        elif seq[int(location-1)] == PR2 or seq[int(location-1)]==sym2:\n",
    "            neg_kmers.append(kmer)\n",
    "        else:\n",
    "            if verbose ==\"yes\":\n",
    "                print(seq[int(location-1)])\n",
    "                print (\"Sequence key error for \"+ID+\" at \" + str(location)) #Will not add it to list of kmers because doesn't match uniprot\n",
    "                #raise Exception(\"Sequence key error for \"+ID+\" at \" + str(location))\n",
    "        if len(kmer) != k:\n",
    "            print (kmer)\n",
    "            print (len(kmer))\n",
    "    \n",
    "    return pos_kmers, neg_kmers     \n",
    "\n",
    "def unlabeling (kmers):\n",
    "    \"\"\"\n",
    "    Removes sym and sym2 from kmers to make dataset without PTM locations\n",
    "    \"\"\"\n",
    "    new_kmers=[]\n",
    "    for kmer in kmers:\n",
    "        new_kmer=kmer.replace(sym, PR).replace(sym2,PR2)\n",
    "        new_kmers.append(new_kmer)\n",
    "    return new_kmers\n",
    "\n",
    "def unlabel_center (kmers, k=53):\n",
    "    \"\"\"\n",
    "    Removes sym and sym2 from kmer center\n",
    "    \"\"\"\n",
    "    new_kmers=[]\n",
    "    half=int((k-1)/2)\n",
    "    for kmer in kmers:\n",
    "        if kmer[half] == sym:\n",
    "            new_kmer=kmer[:half]+PR+kmer[half+1:]\n",
    "        elif kmer[half] == sym2:\n",
    "            new_kmer=kmer[:half]+PR2+kmer[half+1:]\n",
    "        else:\n",
    "            new_kmer=kmer\n",
    "        assert len(new_kmer)==k\n",
    "        new_kmers.append(new_kmer)\n",
    "    \n",
    "    return new_kmers\n",
    "\n",
    "def convert_to_onehot(data, alphabet):\n",
    "    #Creates a dict, that maps to every char of alphabet an unique int based on position\n",
    "    global char_to_int\n",
    "    char_to_int = dict((c,i) for i,c in enumerate(alphabet))\n",
    "    encoded_data = []\n",
    "    #Replaces every char in data with the mapped int\n",
    "    encoded_data.extend([char_to_int[char] for char in data])\n",
    "    return encoded_data\n",
    "\n",
    "def tensor_encoding(x_data, depth, type, alphabet):\n",
    "    indices = []\n",
    "    t2 = []\n",
    "    for i in tqdm(range(len(x_data))):\n",
    "        indices.append(convert_to_onehot(x_data[i,0], alphabet))\n",
    "        if len(convert_to_onehot(x_data[i,0], alphabet)) != 53:\n",
    "            print (x_data[i,0])\n",
    "            print (\"Length off\")\n",
    "    array=np.stack(indices, axis=0)\n",
    "    if type == 'emb':\n",
    "        return array\n",
    "    for i in tqdm(range(len(indices))):\n",
    "        t1 = tf.one_hot(indices[i], depth) # output: [9 x 23]\n",
    "        t2.append(t1)\n",
    "    return t2\n",
    "\n",
    "def make_datasets_train_models (protein_dict, PR_sites, PR2_sites, trial_num, k=53, folder=folder, verbose=\"no\"):\n",
    "    \"\"\"\n",
    "    Given a protein_dict that has been labeled with PTM sites and a list of sites it will make test, train, and val datasets and train a simple CNN \n",
    "    model\n",
    "\n",
    "    Datasets and models are saved in specified folder\n",
    "    \"\"\"\n",
    "    #Make kmers\n",
    "    lpos_kmers, lneg_kmers=make_kmers(protein_dict, PR_site_array, PR2_site_array, k=k, verbose=verbose)\n",
    "    lpos_kmers, lneg_kmers=unlabel_center(lpos_kmers, k=k), unlabel_center(lneg_kmers, k=k)\n",
    "    pos_kmers, neg_kmers=unlabeling(lpos_kmers), unlabeling (lneg_kmers)\n",
    "\n",
    "    y_pos=np.ones((len(pos_kmers)), dtype=np.int32)\n",
    "    y_neg=np.zeros((len(neg_kmers)), dtype=np.int32)\n",
    "\n",
    "    pos_kmers=np.array(pos_kmers)\n",
    "    neg_kmers=np.array(neg_kmers)\n",
    "    lpos_kmers=np.array(lpos_kmers)\n",
    "    lneg_kmers=np.array(lneg_kmers)\n",
    "    \n",
    "    pos_k = np.stack((pos_kmers, y_pos), axis=1)\n",
    "    neg_k = np.stack((neg_kmers, y_neg), axis=1)\n",
    "    lpos_k = np.stack((lpos_kmers, y_pos), axis=1)\n",
    "    lneg_k = np.stack((lneg_kmers, y_neg), axis=1)\n",
    "\n",
    "    full= np.vstack([pos_k,neg_k])\n",
    "    lfull = np.vstack([lpos_k,lneg_k])\n",
    "    np.random.shuffle(full)\n",
    "    np.random.shuffle(lfull)\n",
    "\n",
    "    print (full[:, 1].dtype)\n",
    "    #print (np.shape(full))\n",
    "    #print (np.shape(lfull))\n",
    "\n",
    "    x_data, y_data = np.split(full, 2, axis=1)\n",
    "    x_datal, y_datal = np.split(lfull, 2, axis=1)\n",
    "\n",
    "    #print (np.shape(x_datal))\n",
    "    y_data=y_data.astype(np.int32)\n",
    "    y_datal=y_datal.astype(np.int32)\n",
    "    #print (y_data.dtype)\n",
    "    #print (y_data)\n",
    "\n",
    "    # Split into train, validation, and test datasets\n",
    "    N = len(full)\n",
    "    x_train = x_data[0:int(0.8*N), :]\n",
    "    #print (x_train)\n",
    "    y_train = y_data[0:int(0.8*N)]\n",
    "    #print (y_train)\n",
    "\n",
    "    x_validation = x_data[int(0.8*N):int(0.8*N)+int(0.1*N), :]\n",
    "    y_validation = y_data[int(0.8*N):int(0.8*N)+int(0.1*N)]\n",
    "\n",
    "    x_test = x_data[int(0.8*N)+int(0.1*N):, :]\n",
    "    y_test = y_data[int(0.8*N)+int(0.1*N):]\n",
    "\n",
    "    x_trainl = x_datal[0:int(0.8*N), :]\n",
    "    y_trainl = y_datal[0:int(0.8*N)]\n",
    "\n",
    "    x_validationl = x_datal[int(0.8*N):int(0.8*N)+int(0.1*N), :]\n",
    "    y_validationl = y_datal[int(0.8*N):int(0.8*N)+int(0.1*N)]\n",
    "\n",
    "    x_testl = x_datal[int(0.8*N)+int(0.1*N):, :]\n",
    "    y_testl = y_datal[int(0.8*N)+int(0.1*N):]\n",
    "\n",
    "    #Make tensors of x data\n",
    "    tensor1 = tensor_encoding(x_train, 23, 'emb', alphabet_without_labels)\n",
    "    tensor2 = tensor_encoding(x_validation, 23, 'emb', alphabet_without_labels)\n",
    "    tensor3 = tensor_encoding(x_test, 23, 'emb', alphabet_without_labels)\n",
    "    tensor4 = tensor_encoding(x_trainl, 23, 'emb', alphabet_with_labels)\n",
    "    tensor5 = tensor_encoding(x_validationl, 23, 'emb', alphabet_with_labels)\n",
    "    tensor6 = tensor_encoding(x_testl, 23, 'emb', alphabet_with_labels)\n",
    "\n",
    "    print (type(tensor1))\n",
    "    print (tensor1.dtype)\n",
    "    print (y_train.dtype)\n",
    "    #print (np.shape(tensor2))\n",
    "    #print (np.shape(tensor3))\n",
    "    #print (np.shape(tensor4))\n",
    "    #print (np.shape(tensor5))\n",
    "    #print (np.shape(tensor6))\n",
    "\n",
    "    directory = f\"Trial{trial_num}\"\n",
    "    parent_dir = folder+'/'\n",
    "    path = os.path.join(parent_dir, directory) \n",
    "    try:\n",
    "        os.mkdir(path) \n",
    "    except:\n",
    "        pass\n",
    "    \n",
    "    #Save datasets\n",
    "    np.save(f'{path}/x_train_no_labels_{trial_num}', tensor1)\n",
    "    np.save(f'{path}/y_train_no_labels_{trial_num}', y_train)\n",
    "    np.save(f'{path}/x_validation_no_labels_{trial_num}', tensor2)\n",
    "    np.save(f'{path}/y_validation_no_labels_{trial_num}', y_validation)\n",
    "    np.save(f'{path}/x_test_no_labels_{trial_num}', tensor3)\n",
    "    np.save(f'{path}/y_test_no_labels_{trial_num}', y_test)\n",
    "    np.save(f'{path}/x_train_with_labels_{trial_num}', tensor4)\n",
    "    np.save(f'{path}/y_train_with_labels_{trial_num}', y_trainl)\n",
    "    np.save(f'{path}/x_validation_with_labels_{trial_num}', tensor5)\n",
    "    np.save(f'{path}/y_validation_with_labels_{trial_num}', y_validationl)\n",
    "    np.save(f'{path}/x_test_with_labels_{trial_num}', tensor6)\n",
    "    np.save(f'{path}/y_test_with_labels_{trial_num}', y_testl)\n",
    "\n",
    "    #Add the model\n",
    "    callback = tf.keras.callbacks.EarlyStopping(monitor='val_loss', mode='min', verbose=1, patience=15, restore_best_weights=True)\n",
    "\n",
    "    model = Sequential()\n",
    "    model.add(Embedding(input_dim=256, output_dim=21, embeddings_initializer={'class_name': 'RandomUniform',\n",
    "         'config': {'minval': -0.05, 'maxval': 0.05, 'seed': None}}, input_length=53, name=\"emb_layer\"))\n",
    "    model.add(Reshape((53, 21, 1), input_shape=(53, 21)))\n",
    "    model.add(Conv2D(16, (3, 3), padding='valid', input_shape=(53, 21, 1), activation='relu', kernel_regularizer=l2(1e-06)))\n",
    "    model.add(MaxPooling2D(pool_size=(2,2), padding='valid'))\n",
    "    model.add(Flatten())\n",
    "    model.add(Dense(32, activation=\"relu\"))\n",
    "    model.add(Dropout(0.1))\n",
    "    model.add(Dense(1, activation='sigmoid'))\n",
    "    model.summary()\n",
    "\n",
    "    model.compile(optimizer=Adam(1e-03), \n",
    "                  loss='BinaryCrossentropy', \n",
    "                  metrics=['accuracy'])\n",
    "    history = model.fit(x=tensor1,y=y_train,\n",
    "              batch_size=100, epochs=400,\n",
    "              validation_data=(tensor2, y_validation), \n",
    "              callbacks=[callback], verbose=1)\n",
    "    model.save(f'{path}/emb_CNN_no_labels_{trial_num}.h5')\n",
    "    model.evaluate(tensor3, y_test)\n",
    "\n",
    "     # The predictions are between 0 and 1 because of the sigmoid activation\n",
    "    # function of the last layer in the NN\n",
    "    pred = model.predict(tensor3).flatten()\n",
    "\n",
    "    # We can also calculate the AUC to get an estimate how good the NN actually learned\n",
    "    print (\"Without labels\")\n",
    "    fpr, tpr, thresholds = roc_curve(y_test, pred)\n",
    "    print(\"AUC:\", round(auc(fpr, tpr), 4))\n",
    "    print(\"AUPRC:\", round(average_precision_score(y_test, pred), 4))\n",
    "    pred2 = (pred > 0.5)\n",
    "    pred2 = np.array(pred2)\n",
    "    pred2 = pred2.astype(int)\n",
    "    print(\"MCC:\", round(matthews_corrcoef(y_test, pred2), 4))\n",
    "    print(\"F1:\", round(f1_score(y_test, pred2), 4))\n",
    "    print(\"Sensitivity:\", round(np.average(tpr), 4))\n",
    "    print(\"Specificity:\", round(np.average(1-fpr), 4))\n",
    "    print(\"Precision:\", round(precision_score(y_test, pred2), 4))\n",
    "\n",
    "    #Now with labels\n",
    "    historyl = model.fit(x=tensor4,y=y_trainl,\n",
    "              batch_size=100, epochs=400,\n",
    "              validation_data=(tensor5, y_validationl), \n",
    "              callbacks=[callback], verbose=1)\n",
    "    model.save(f'{path}/emb_CNN_with_labels_{trial_num}.h5')\n",
    "    model.evaluate(tensor6, y_testl)\n",
    "\n",
    "     # The predictions are between 0 and 1 because of the sigmoid activation\n",
    "    # function of the last layer in the NN\n",
    "    predl = model.predict(tensor6).flatten()\n",
    "\n",
    "    # We can also calculate the AUC to get an estimate how good the NN actually learned\n",
    "    print (\"With labels\")\n",
    "    fprl, tprl, thresholdsl = roc_curve(y_testl, predl)\n",
    "    print(\"AUC:\", round(auc(fprl, tprl), 4))\n",
    "    print(\"AUPRC:\", round(average_precision_score(y_testl, predl), 4))\n",
    "    pred2l = (predl > 0.5)\n",
    "    pred2l = np.array(pred2l)\n",
    "    pred2l = pred2l.astype(int)\n",
    "    print(\"MCC:\", round(matthews_corrcoef(y_testl, pred2l), 4))\n",
    "    print(\"F1:\", round(f1_score(y_testl, pred2l), 4))\n",
    "    print(\"Sensitivity:\", round(np.average(tprl), 4))\n",
    "    print(\"Specificity:\", round(np.average(1-fprl), 4))\n",
    "    print(\"Precision:\", round(precision_score(y_testl, pred2l), 4))\n",
    "\n",
    "    \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "37bb7bea-ee7a-4617-9ad8-ca93c4a0ef32",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "IOPub data rate exceeded.\n",
      "The Jupyter server will temporarily stop sending output\n",
      "to the client in order to avoid crashing it.\n",
      "To change this limit, set the config variable\n",
      "`--ServerApp.iopub_data_rate_limit`.\n",
      "\n",
      "Current values:\n",
      "ServerApp.iopub_data_rate_limit=1000000.0 (bytes/sec)\n",
      "ServerApp.rate_limit_window=3.0 (secs)\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "7499\n",
      "7499\n",
      "\n",
      "0\n",
      "\n",
      "0\n",
      "\n",
      "0\n",
      "<U53\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|█████████████████████████████████████████████████████████████████████████| 11900/11900 [00:00<00:00, 34043.54it/s]\n",
      "100%|███████████████████████████████████████████████████████████████████████████| 1487/1487 [00:00<00:00, 25085.09it/s]\n",
      "100%|███████████████████████████████████████████████████████████████████████████| 1489/1489 [00:00<00:00, 22180.81it/s]\n",
      "100%|█████████████████████████████████████████████████████████████████████████| 11900/11900 [00:00<00:00, 37701.55it/s]\n",
      "100%|███████████████████████████████████████████████████████████████████████████| 1487/1487 [00:00<00:00, 41351.82it/s]\n",
      "100%|███████████████████████████████████████████████████████████████████████████| 1489/1489 [00:00<00:00, 28283.04it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'numpy.ndarray'>\n",
      "int32\n",
      "int32\n",
      "Model: \"sequential_9\"\n",
      "_________________________________________________________________\n",
      " Layer (type)                Output Shape              Param #   \n",
      "=================================================================\n",
      " emb_layer (Embedding)       (None, 53, 21)            5376      \n",
      "                                                                 \n",
      " reshape_9 (Reshape)         (None, 53, 21, 1)         0         \n",
      "                                                                 \n",
      " conv2d_9 (Conv2D)           (None, 51, 19, 16)        160       \n",
      "                                                                 \n",
      " max_pooling2d_9 (MaxPooling  (None, 25, 9, 16)        0         \n",
      " 2D)                                                             \n",
      "                                                                 \n",
      " flatten_9 (Flatten)         (None, 3600)              0         \n",
      "                                                                 \n",
      " dense_18 (Dense)            (None, 32)                115232    \n",
      "                                                                 \n",
      " dropout_9 (Dropout)         (None, 32)                0         \n",
      "                                                                 \n",
      " dense_19 (Dense)            (None, 1)                 33        \n",
      "                                                                 \n",
      "=================================================================\n",
      "Total params: 120,801\n",
      "Trainable params: 120,801\n",
      "Non-trainable params: 0\n",
      "_________________________________________________________________\n",
      "Epoch 1/400\n",
      "119/119 [==============================] - 7s 39ms/step - loss: 0.6682 - accuracy: 0.5845 - val_loss: 0.6256 - val_accuracy: 0.6483\n",
      "Epoch 2/400\n",
      "119/119 [==============================] - 2s 19ms/step - loss: 0.6124 - accuracy: 0.6553 - val_loss: 0.6231 - val_accuracy: 0.6510\n",
      "Epoch 3/400\n",
      "119/119 [==============================] - 3s 21ms/step - loss: 0.5900 - accuracy: 0.6807 - val_loss: 0.5944 - val_accuracy: 0.6873\n",
      "Epoch 4/400\n",
      "119/119 [==============================] - 3s 29ms/step - loss: 0.5810 - accuracy: 0.6882 - val_loss: 0.5900 - val_accuracy: 0.6907\n",
      "Epoch 5/400\n",
      "119/119 [==============================] - 3s 24ms/step - loss: 0.5683 - accuracy: 0.6990 - val_loss: 0.5865 - val_accuracy: 0.6960\n",
      "Epoch 6/400\n",
      "119/119 [==============================] - 5s 38ms/step - loss: 0.5553 - accuracy: 0.7102 - val_loss: 0.5826 - val_accuracy: 0.6940\n",
      "Epoch 7/400\n",
      "119/119 [==============================] - 3s 23ms/step - loss: 0.5443 - accuracy: 0.7180 - val_loss: 0.5817 - val_accuracy: 0.6967\n",
      "Epoch 8/400\n",
      "119/119 [==============================] - 4s 30ms/step - loss: 0.5334 - accuracy: 0.7250 - val_loss: 0.5781 - val_accuracy: 0.6960\n",
      "Epoch 9/400\n",
      "119/119 [==============================] - 3s 25ms/step - loss: 0.5212 - accuracy: 0.7387 - val_loss: 0.5802 - val_accuracy: 0.7001\n",
      "Epoch 10/400\n",
      "119/119 [==============================] - 7s 63ms/step - loss: 0.5082 - accuracy: 0.7520 - val_loss: 0.5862 - val_accuracy: 0.6765\n",
      "Epoch 11/400\n",
      "119/119 [==============================] - 3s 21ms/step - loss: 0.4937 - accuracy: 0.7625 - val_loss: 0.5927 - val_accuracy: 0.6920\n",
      "Epoch 12/400\n",
      "119/119 [==============================] - 4s 31ms/step - loss: 0.4789 - accuracy: 0.7689 - val_loss: 0.6517 - val_accuracy: 0.6590\n",
      "Epoch 13/400\n",
      "119/119 [==============================] - 3s 25ms/step - loss: 0.4694 - accuracy: 0.7751 - val_loss: 0.5943 - val_accuracy: 0.6967\n",
      "Epoch 14/400\n",
      "119/119 [==============================] - 4s 36ms/step - loss: 0.4511 - accuracy: 0.7877 - val_loss: 0.5982 - val_accuracy: 0.7048\n",
      "Epoch 15/400\n",
      "119/119 [==============================] - 5s 40ms/step - loss: 0.4405 - accuracy: 0.7939 - val_loss: 0.6021 - val_accuracy: 0.7021\n",
      "Epoch 16/400\n",
      "119/119 [==============================] - 3s 28ms/step - loss: 0.4255 - accuracy: 0.8005 - val_loss: 0.6317 - val_accuracy: 0.6859\n",
      "Epoch 17/400\n",
      "119/119 [==============================] - 3s 23ms/step - loss: 0.4062 - accuracy: 0.8154 - val_loss: 0.6067 - val_accuracy: 0.6947\n",
      "Epoch 18/400\n",
      "119/119 [==============================] - 3s 24ms/step - loss: 0.3938 - accuracy: 0.8206 - val_loss: 0.6324 - val_accuracy: 0.6893\n",
      "Epoch 19/400\n",
      "119/119 [==============================] - 3s 26ms/step - loss: 0.3841 - accuracy: 0.8295 - val_loss: 0.6566 - val_accuracy: 0.6772\n",
      "Epoch 20/400\n",
      "119/119 [==============================] - 2s 21ms/step - loss: 0.3642 - accuracy: 0.8360 - val_loss: 0.6853 - val_accuracy: 0.6718\n",
      "Epoch 21/400\n",
      "119/119 [==============================] - 2s 21ms/step - loss: 0.3464 - accuracy: 0.8497 - val_loss: 0.6971 - val_accuracy: 0.6846\n",
      "Epoch 22/400\n",
      "119/119 [==============================] - 4s 33ms/step - loss: 0.3305 - accuracy: 0.8582 - val_loss: 0.6991 - val_accuracy: 0.6792\n",
      "Epoch 23/400\n",
      "118/119 [============================>.] - ETA: 0s - loss: 0.3261 - accuracy: 0.8619Restoring model weights from the end of the best epoch: 8.\n",
      "119/119 [==============================] - 3s 25ms/step - loss: 0.3257 - accuracy: 0.8621 - val_loss: 0.7067 - val_accuracy: 0.6651\n",
      "Epoch 23: early stopping\n",
      "47/47 [==============================] - 0s 6ms/step - loss: 0.5829 - accuracy: 0.6931\n",
      "47/47 [==============================] - 1s 16ms/step\n",
      "Without labels\n",
      "AUC: 0.7583\n",
      "AUPRC: 0.7497\n",
      "MCC: 0.3861\n",
      "F1: 0.6963\n",
      "Sensitivity: 0.6398\n",
      "Specificity: 0.6456\n",
      "Precision: 0.6968\n",
      "Epoch 1/400\n",
      "119/119 [==============================] - 4s 32ms/step - loss: 0.4485 - accuracy: 0.7896 - val_loss: 0.3953 - val_accuracy: 0.8184\n",
      "Epoch 2/400\n",
      "119/119 [==============================] - 2s 21ms/step - loss: 0.3684 - accuracy: 0.8392 - val_loss: 0.3734 - val_accuracy: 0.8325\n",
      "Epoch 3/400\n",
      "119/119 [==============================] - 2s 20ms/step - loss: 0.3505 - accuracy: 0.8506 - val_loss: 0.3528 - val_accuracy: 0.8473\n",
      "Epoch 4/400\n",
      "119/119 [==============================] - 3s 26ms/step - loss: 0.3426 - accuracy: 0.8560 - val_loss: 0.3538 - val_accuracy: 0.8426\n",
      "Epoch 5/400\n",
      "119/119 [==============================] - 2s 19ms/step - loss: 0.3338 - accuracy: 0.8599 - val_loss: 0.3532 - val_accuracy: 0.8521\n",
      "Epoch 6/400\n",
      "119/119 [==============================] - 2s 20ms/step - loss: 0.3257 - accuracy: 0.8659 - val_loss: 0.3589 - val_accuracy: 0.8473\n",
      "Epoch 7/400\n",
      "119/119 [==============================] - 2s 17ms/step - loss: 0.3216 - accuracy: 0.8638 - val_loss: 0.3528 - val_accuracy: 0.8514\n",
      "Epoch 8/400\n",
      "119/119 [==============================] - 2s 15ms/step - loss: 0.3135 - accuracy: 0.8676 - val_loss: 0.3656 - val_accuracy: 0.8467\n",
      "Epoch 9/400\n",
      "119/119 [==============================] - 2s 18ms/step - loss: 0.3037 - accuracy: 0.8776 - val_loss: 0.3660 - val_accuracy: 0.8514\n",
      "Epoch 10/400\n",
      "119/119 [==============================] - 2s 18ms/step - loss: 0.2952 - accuracy: 0.8773 - val_loss: 0.3855 - val_accuracy: 0.8467\n",
      "Epoch 11/400\n",
      "119/119 [==============================] - 2s 17ms/step - loss: 0.2891 - accuracy: 0.8795 - val_loss: 0.3668 - val_accuracy: 0.8305\n",
      "Epoch 12/400\n",
      "119/119 [==============================] - 2s 17ms/step - loss: 0.2795 - accuracy: 0.8836 - val_loss: 0.3686 - val_accuracy: 0.8467\n",
      "Epoch 13/400\n",
      "119/119 [==============================] - 2s 15ms/step - loss: 0.2694 - accuracy: 0.8908 - val_loss: 0.3776 - val_accuracy: 0.8426\n",
      "Epoch 14/400\n",
      "119/119 [==============================] - 2s 16ms/step - loss: 0.2577 - accuracy: 0.8934 - val_loss: 0.3784 - val_accuracy: 0.8299\n",
      "Epoch 15/400\n",
      "119/119 [==============================] - 2s 19ms/step - loss: 0.2479 - accuracy: 0.9013 - val_loss: 0.3895 - val_accuracy: 0.8433\n",
      "Epoch 16/400\n",
      "119/119 [==============================] - 2s 18ms/step - loss: 0.2389 - accuracy: 0.9034 - val_loss: 0.3933 - val_accuracy: 0.8211\n",
      "Epoch 17/400\n",
      "119/119 [==============================] - 2s 19ms/step - loss: 0.2299 - accuracy: 0.9082 - val_loss: 0.3894 - val_accuracy: 0.8319\n",
      "Epoch 18/400\n",
      "119/119 [==============================] - 2s 16ms/step - loss: 0.2139 - accuracy: 0.9173 - val_loss: 0.4075 - val_accuracy: 0.8252\n",
      "Epoch 19/400\n",
      "119/119 [==============================] - 2s 18ms/step - loss: 0.2136 - accuracy: 0.9138 - val_loss: 0.4156 - val_accuracy: 0.8278\n",
      "Epoch 20/400\n",
      "119/119 [==============================] - 2s 18ms/step - loss: 0.1943 - accuracy: 0.9239 - val_loss: 0.4392 - val_accuracy: 0.8325\n",
      "Epoch 21/400\n",
      "119/119 [==============================] - 2s 19ms/step - loss: 0.1858 - accuracy: 0.9262 - val_loss: 0.4511 - val_accuracy: 0.8272\n",
      "Epoch 22/400\n",
      "118/119 [============================>.] - ETA: 0s - loss: 0.1779 - accuracy: 0.9318Restoring model weights from the end of the best epoch: 7.\n",
      "119/119 [==============================] - 2s 17ms/step - loss: 0.1780 - accuracy: 0.9319 - val_loss: 0.4554 - val_accuracy: 0.8258\n",
      "Epoch 22: early stopping\n",
      "47/47 [==============================] - 0s 7ms/step - loss: 0.3891 - accuracy: 0.8334\n",
      "47/47 [==============================] - 0s 5ms/step\n",
      "With labels\n",
      "AUC: 0.9081\n",
      "AUPRC: 0.9133\n",
      "MCC: 0.6694\n",
      "F1: 0.8278\n",
      "Sensitivity: 0.7831\n",
      "Specificity: 0.7444\n",
      "Precision: 0.865\n"
     ]
    }
   ],
   "source": [
    "alphabet_with_labels = \"ARNDCEQGHILKMFPSTWYV@&-UX\" \n",
    "alphabet_without_labels = \"ARNDCEQGHILKMFPSTWYV-UX\"\n",
    "\n",
    "for trial_num in range(1, 2):#Do number of reps for number of validation runs\n",
    "    make_datasets_train_models (protein_dict, PR_site_array, PR2_site_array, trial_num, k=53, folder=f'{folder}/PostSprint')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "0de76677-f27e-42dc-aeac-eda23f7c0967",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "IOPub data rate exceeded.\n",
      "The Jupyter server will temporarily stop sending output\n",
      "to the client in order to avoid crashing it.\n",
      "To change this limit, set the config variable\n",
      "`--ServerApp.iopub_data_rate_limit`.\n",
      "\n",
      "Current values:\n",
      "ServerApp.iopub_data_rate_limit=1000000.0 (bytes/sec)\n",
      "ServerApp.rate_limit_window=3.0 (secs)\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "7499\n",
      "7499\n",
      "\n",
      "0\n",
      "\n",
      "0\n",
      "\n",
      "0\n",
      "[[  3 375]\n",
      " [  9 192]\n",
      " [  6 251]\n",
      " [ 10 207]\n",
      " [ 15 145]\n",
      " [ 12 165]\n",
      " [ 18 142]\n",
      " [ 24 145]\n",
      " [ 13 165]\n",
      " [  7 235]\n",
      " [  5 262]\n",
      " [  2 503]\n",
      " [  1 470]\n",
      " [  4 401]\n",
      " [ 17 148]\n",
      " [ 16 161]\n",
      " [ 23 146]\n",
      " [ 19 140]\n",
      " [  8 216]\n",
      " [ 20 161]\n",
      " [ 22 140]\n",
      " [ 14 172]\n",
      " [ 21 152]\n",
      " [ 26 153]\n",
      " [ 25 146]\n",
      " [ 11 170]]\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAjsAAAHFCAYAAAAUpjivAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjguMCwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy81sbWrAAAACXBIWXMAAA9hAAAPYQGoP6dpAABSlUlEQVR4nO3dd1wU1/4//tdKBwEBhYWIgBGwoMYWFY2ogNiwJVFjw5aYWLHHaK5Ygi2iXo0tKqh8Lcm1l6DYSJR4rSgarxLFDmIhNHEROL8//LEfl7qDi+D4ej4e+3i4M2fOvGc5LC/PzOwqhBACRERERDJVqbwLICIiIipLDDtEREQkaww7REREJGsMO0RERCRrDDtEREQkaww7REREJGsMO0RERCRrDDtEREQkaww7REREJGsMO/TeCAsLg0KhKPQxadKk8i7vvXX06FE0bdoUZmZmUCgU2L17d3mXJEneuDp37lyZ7icoKEhjzBoaGsLFxQXjxo3DP//8AwBFju/8jxMnTuD27dvq50FBQYXuc+jQoeo2utC2bVu0bdtW/fz58+cICgrCiRMndNI/UVH0y7sAorctNDQUtWvX1ljm4OBQTtW834QQ6N27N9zc3LB3716YmZnB3d29vMuq0CIiImBpaYm0tDQcPHgQy5Ytw5kzZxAdHY0///xTo+2cOXNw/PhxHDt2TGN53bp18ezZMwCAubk5wsLC8K9//QuVKv3f/3/T09Px66+/wsLCAqmpqTqpfeXKlRrPnz9/jlmzZgGARggi0jWGHXrveHh4oGnTplq1ffnyJRQKBfT1+atSFh4+fIhnz56hZ8+e8Pb2Lu9yJMkbG29bkyZNULVqVQCAr68vnj59is2bNyM6OhqtWrXSaFutWjVUqlQJLVq0KNBPXtjp06cP1q1bh6NHj8LX11e9fvv27cjJyUGPHj0QHh6uk9rr1q2rk36IpOJpLKL/34kTJ6BQKLB582ZMnDgRH3zwAYyMjPD3338DAI4cOQJvb29YWFjA1NQUrVq1wtGjRwv0c+DAAXz00UcwMjKCi4sLfvzxR/UpiDx5pxDCwsIKbF/YaYW4uDj069cPtra2MDIyQp06dfDTTz8VWv/WrVsxffp0ODg4wMLCAj4+Prh+/XqB/URERMDb2xuWlpYwNTVFnTp1MG/ePADA5s2boVAoCswUAMDs2bNhYGCAhw8fFvt6njx5Et7e3jA3N4epqSk8PT1x4MAB9fqgoCBUr14dADB16lQoFAo4OzsX2Z/U49Pm5/X3339jyJAhcHV1hampKT744AP4+/sjNja20H0XNTYAIDk5GUOGDIG1tTXMzMzg7++PW7duqdfPmTMH+vr6uHfvXoFahw4dChsbG7x48aLY17QweUHmzp07krcFAHd3d3h6emLDhg0ayzds2IBevXrB0tJSq35u3bqFvn37wsHBAUZGRrCzs4O3tzdiYmLUbV4/jXX79m1Uq1YNADBr1iz16bLBgwer22sz7nNzczF37ly4u7vDxMQEVapUQYMGDbBs2TLpLwbJFsMOvXdycnKQnZ2t8XjdtGnTcPfuXaxevRr79u2Dra0twsPD0aFDB1hYWGDjxo345ZdfYG1tDT8/P40/oEePHkX37t1hbm6Obdu2YdGiRfjll18QGhpa6nr/+usvNGvWDFeuXMHixYuxf/9+dOnSBWPHjlWfAnjdd999hzt37mDdunVYu3Yt4uLi4O/vj5ycHHWb9evXo3PnzsjNzVUf59ixY3H//n0Ar/63r1QqC/xhyc7Oxpo1a9CzZ89iT/1FRUWhffv2SElJwfr167F161aYm5vD398f27dvBwAMHz4cO3fuBACMGTMGf/75J3bt2lXi66HN8Wn783r48CFsbGwwf/58RERE4KeffoK+vj6aN29eaIAqbGzkGTZsGCpVqoQtW7Zg6dKlOHPmDNq2bau+nmbEiBHQ19fHmjVrNPp89uwZtm3bhmHDhsHY2LjE488vL3DlBYfSGDZsGHbv3o3k5GQAwPXr1xEdHY1hw4Zp3Ufnzp1x/vx5LFy4EJGRkVi1ahUaNWqkPv787O3tERERod7/n3/+iT///BPff/89AO3H/cKFCxEUFIQvvvgCBw4cwPbt2zFs2LAi90vvKUH0nggNDRUACn28fPlSHD9+XAAQbdq00dguIyNDWFtbC39/f43lOTk5omHDhuLjjz9WL2vevLlwcHAQmZmZ6mWpqanC2tpavP7rFh8fLwCI0NDQAnUCEDNnzlQ/9/PzE9WrVxcpKSka7UaPHi2MjY3Fs2fPhBBCXX/nzp012v3yyy8CgPjzzz+FEEKkpaUJCwsL0bp1a5Gbm1vk6zVz5kxhaGgoHj16pF62fft2AUBERUUVuZ0QQrRo0ULY2tqKtLQ09bLs7Gzh4eEhqlevrt5v3uuwaNGiYvuTcnxSfl75ZWdni6ysLOHq6irGjx9fYN/5x4YQ/zeuevbsqbH81KlTAoCYO3euellAQICwtbUVKpVKvWzBggWiUqVKIj4+vtjjnzlzpgAgEhMTxcuXL0VycrIIDw8XJiYmwtHRUWPMvb4/MzOzQvt7/bVPS0sTlStXFitWrBBCCDF58mTh4uIicnNzxahRo0RJfyqePHkiAIilS5cW287Ly0t4eXmpnz9+/LjAeM+j7bjv2rWr+Oijj4rdLxFndui9s2nTJpw9e1bj8fo1OZ9++qlG++joaDx79gwBAQEas0G5ubno2LEjzp49i4yMDGRkZODs2bPo1auXxv/Q82Y0SuPFixc4evQoevbsCVNTU439d+7cGS9evMDp06c1tunWrZvG8wYNGgD4v9Mc0dHRSE1NxciRI4u95uSbb74BAPz888/qZStWrED9+vXRpk2bIrfLyMjAf//7X3z22WeoXLmyermenh4GDhyI+/fvFzproi1tjk+bnxfwaqYqODgYdevWhaGhIfT19WFoaIi4uDhcu3atwL7zj43X9e/fX+O5p6cnnJyccPz4cfWycePGISkpCb/++iuAV6dgVq1ahS5duhR7Cu91SqUSBgYGsLKywoABA9C4cWNERESUalYoT+XKlfH5559jw4YNyM7OxqZNmzBkyBCtr0mytrbGhx9+iEWLFiEkJAQXL15Ebm5uqeuRMu4//vhjXLp0CSNHjsShQ4d0djE1yQuvuqT3Tp06dYq9QNne3l7j+aNHjwAAn332WZHbPHv2DAqFArm5uVAqlQXWF7ZMG0+fPkV2djaWL1+O5cuXF9rmyZMnGs9tbGw0nhsZGQEAMjMzAQCPHz8GAPX1MkWxs7NDnz59sGbNGnz77be4evUq/vjjjwKnYfJLTk6GEKLA6wj8311vT58+LbaP4pR0fNr+vMzMzDBhwgT89NNPmDp1Kry8vGBlZYVKlSph+PDh6v5eV9gx5Snq5/76sTZq1AiffPIJfvrpJ/Tv3x/79+/H7du3S3xNX3fkyBFYWlrCwMAA1atXL/B6lNawYcPQunVr/PDDD3j8+LHGtTMlUSgUOHr0KGbPno2FCxdi4sSJsLa2Rv/+/fHDDz/A3NxcUi1Sxv20adNgZmaG8PBwrF69Gnp6emjTpg0WLFig9Y0IJH8MO0T55P/fbN6dL8uXLy/0rhbgVTDIuzsnMTGxwPr8y/L+F65SqTSW5w8BVlZW6hmRUaNGFbpvFxeXYo6moLxrO/KuzynOuHHjsHnzZuzZswcRERGoUqVKgRmM/PICQ0JCQoF1eRc1572mZUHbnxfw6tqeQYMGITg4WGP9kydPUKVKlQLbFTfTUdTPvVatWhrLxo4di88//xwXLlzAihUr4ObmpnEXVEkaNmxYJq9fq1at4O7ujtmzZ8PX1xeOjo6StndycsL69esBADdu3MAvv/yCoKAgZGVlYfXq1ZL6kjLu9fX1MWHCBEyYMAH//PMPjhw5gu+++w5+fn64d+8eTE1NJe2b5Ilhh6gErVq1QpUqVfDXX39h9OjRRbYzNDTExx9/jJ07d2LRokXqQJOWloZ9+/ZptLWzs4OxsTEuX76ssXzPnj0az01NTdGuXTtcvHgRDRo0gKGh4Rsfj6enJywtLbF69Wr07du32D/gTZo0gaenJxYsWIArV67gq6++gpmZWbH9m5mZoXnz5ti5cyd+/PFHmJiYAHh1yiY8PBzVq1eHm5vbGx9HUbT9eQGvwkvezFCeAwcO4MGDBwVCSkn+3//7fxqnuaKjo3Hnzh0MHz5co13Pnj1Ro0YNTJw4EVFRUViyZEm53MJemBkzZuA///lPkQFDW25ubpgxYwZ27NiBCxcuFNku/6xcntKO+ypVquCzzz7DgwcPEBgYiNu3b/N2dwLAsENUosqVK2P58uUICAjAs2fP8Nlnn8HW1haPHz/GpUuX8PjxY6xatQrAq9uLO3bsCF9fX0ycOBE5OTlYsGABzMzM1J9rArz6IztgwABs2LABH374IRo2bIgzZ85gy5YtBfa/bNkytG7dGp988gm++eYbODs7Iy0tDX///Tf27dtX4APjtDmexYsXY/jw4fDx8cGXX34JOzs7/P3337h06RJWrFih0X7cuHHo06cPFAoFRo4cqdU+5s2bB19fX7Rr1w6TJk2CoaEhVq5ciStXrmDr1q1l+sddys+ra9euCAsLQ+3atdGgQQOcP38eixYtKvEUX2HOnTuH4cOH4/PPP8e9e/cwffp0fPDBBwVeMz09PYwaNQpTp06FmZmZpNNFZW3AgAEYMGCA5O0uX76M0aNH4/PPP4erqysMDQ1x7NgxXL58Gd9++22R25mbm8PJyQl79uyBt7c3rK2tUbVqVTg7O2s97v39/dWfnVWtWjXcuXMHS5cuhZOTE1xdXUv9WpC8MOwQaWHAgAGoUaMGFi5ciBEjRiAtLQ22trb46KOPNP5Y+fr6Yvfu3ZgxY4b69u2RI0ciMzOzwG3iixcvBvDq1tn09HS0b98e+/fvL3Chat26dXHhwgXMmTMHM2bMQFJSEqpUqQJXV1d07ty5VMczbNgwODg4YMGCBRg+fDiEEHB2dkZAQECBtj169ICRkRHatWun9R8PLy8vHDt2DDNnzsTgwYORm5uLhg0bYu/evejatWupapZC25/XsmXLYGBggHnz5iE9PR2NGzfGzp07MWPGDMn7XL9+PTZv3oy+fftCpVKhXbt2WLZsGaytrQu07dOnD6ZOnYqBAwdq/Tk2FZlSqcSHH36IlStX4t69e1AoFKhZsyYWL16MMWPGFLvt+vXrMXnyZHTr1g0qlQoBAQEICwvTety3a9cOO3bswLp165CamgqlUglfX198//33MDAwKOtDp3eEQgghyrsIIrkLCgrCrFmz8C7+uu3btw/dunXDgQMHSh2uSNPy5csxduxYXLlyBfXq1SvvcohkjzM7RFSov/76C3fu3MHEiRPx0UcfoVOnTuVd0jvv4sWLiI+Px+zZs9G9e3cGHaK3hGGHiAo1cuRInDp1Co0bN8bGjRsrzEW077KePXsiMTERn3zyieQ7lIio9Hgai4iIiGSNn6BMREREssawQ0RERLLGsENERESyxguU8eqTXR8+fAhzc3NehElERPSOEEIgLS0NDg4OqFSp6Pkbhh28+r4eqd8DQ0RERBXDvXv3iv3kc4YdQP2NvPfu3YOFhUU5V0NERETaSE1NhaOjo/rveFEYdvB/32RsYWHBsENERPSOKekSFF6gTERERLLGsENERESyxrBDREREssawQ0RERLLGsENERESyxrBDREREssawQ0RERLLGsENERESyxrBDREREssawQ0RERLLGsENERESyxrBDREREssawQ0RERLJWrmEnKCgICoVC46FUKtXrhRAICgqCg4MDTExM0LZtW1y9elWjD5VKhTFjxqBq1aowMzNDt27dcP/+/bd9KERERFRBlfvMTr169ZCQkKB+xMbGqtctXLgQISEhWLFiBc6ePQulUglfX1+kpaWp2wQGBmLXrl3Ytm0bTp48ifT0dHTt2hU5OTnlcThERERUweiXewH6+hqzOXmEEFi6dCmmT5+OXr16AQA2btwIOzs7bNmyBSNGjEBKSgrWr1+PzZs3w8fHBwAQHh4OR0dHHDlyBH5+fm/1WMqS87cHSrXd7flddFwJERHRu6XcZ3bi4uLg4OAAFxcX9O3bF7du3QIAxMfHIzExER06dFC3NTIygpeXF6KjowEA58+fx8uXLzXaODg4wMPDQ92mMCqVCqmpqRoPIiIikqdyDTvNmzfHpk2bcOjQIfz8889ITEyEp6cnnj59isTERACAnZ2dxjZ2dnbqdYmJiTA0NISVlVWRbQozb948WFpaqh+Ojo46PjIiIiKqKMo17HTq1Amffvop6tevDx8fHxw48OpUzcaNG9VtFAqFxjZCiALL8iupzbRp05CSkqJ+3Lt37w2OgoiIiCqycj+N9TozMzPUr18fcXFx6ut48s/QJCUlqWd7lEolsrKykJycXGSbwhgZGcHCwkLjQURERPJUocKOSqXCtWvXYG9vDxcXFyiVSkRGRqrXZ2VlISoqCp6engCAJk2awMDAQKNNQkICrly5om5DRERE77dyvRtr0qRJ8Pf3R40aNZCUlIS5c+ciNTUVAQEBUCgUCAwMRHBwMFxdXeHq6org4GCYmpqiX79+AABLS0sMGzYMEydOhI2NDaytrTFp0iT1aTEiIiKicg079+/fxxdffIEnT56gWrVqaNGiBU6fPg0nJycAwJQpU5CZmYmRI0ciOTkZzZs3x+HDh2Fubq7uY8mSJdDX10fv3r2RmZkJb29vhIWFQU9Pr7wOi4iIiCoQhRBClHcR5S01NRWWlpZISUmpsNfv8HN2iIiINGn797tCXbNDREREpGsMO0RERCRrDDtEREQkaww7REREJGsMO0RERCRrDDtEREQkaww7REREJGsMO0RERCRrDDtEREQkaww7REREJGsMO0RERCRrDDtEREQkaww7REREJGsMO0RERCRr+uVdgNw5f3ugVNvdnt9Fx5UQERG9nzizQ0RERLLGsENERESyxrBDREREssawQ0RERLLGsENERESyxrBDREREssawQ0RERLLGsENERESyxrBDREREssawQ0RERLLGsENERESyxrBDREREssawQ0RERLLGsENERESyxrBDREREssawQ0RERLLGsENERESyxrBDREREssawQ0RERLLGsENERESyxrBDREREssawQ0RERLLGsENERESyxrBDREREssawQ0RERLLGsENERESyxrBDREREssawQ0RERLLGsENERESyxrBDREREssawQ0RERLLGsENERESyxrBDREREssawQ0RERLLGsENERESyxrBDREREssawQ0RERLLGsENERESyxrBDREREssawQ0RERLLGsENERESyxrBDREREssawQ0RERLLGsENERESyVmHCzrx586BQKBAYGKheJoRAUFAQHBwcYGJigrZt2+Lq1asa26lUKowZMwZVq1aFmZkZunXrhvv377/l6omIiKiiqhBh5+zZs1i7di0aNGigsXzhwoUICQnBihUrcPbsWSiVSvj6+iItLU3dJjAwELt27cK2bdtw8uRJpKeno2vXrsjJyXnbh0FEREQVULmHnfT0dPTv3x8///wzrKys1MuFEFi6dCmmT5+OXr16wcPDAxs3bsTz58+xZcsWAEBKSgrWr1+PxYsXw8fHB40aNUJ4eDhiY2Nx5MiR8jokIiIiqkD0y7uAUaNGoUuXLvDx8cHcuXPVy+Pj45GYmIgOHTqolxkZGcHLywvR0dEYMWIEzp8/j5cvX2q0cXBwgIeHB6Kjo+Hn51foPlUqFVQqlfp5ampqGRyZfDl/e6BU292e30XHlRAREZWsXMPOtm3bcOHCBZw9e7bAusTERACAnZ2dxnI7OzvcuXNH3cbQ0FBjRiivTd72hZk3bx5mzZr1puUTERHRO6DcTmPdu3cP48aNQ3h4OIyNjYtsp1AoNJ4LIQosy6+kNtOmTUNKSor6ce/ePWnFExER0Tuj3MLO+fPnkZSUhCZNmkBfXx/6+vqIiorCv//9b+jr66tndPLP0CQlJanXKZVKZGVlITk5ucg2hTEyMoKFhYXGg4iIiOSp3MKOt7c3YmNjERMTo340bdoU/fv3R0xMDGrWrAmlUonIyEj1NllZWYiKioKnpycAoEmTJjAwMNBok5CQgCtXrqjbEBER0fut3K7ZMTc3h4eHh8YyMzMz2NjYqJcHBgYiODgYrq6ucHV1RXBwMExNTdGvXz8AgKWlJYYNG4aJEyfCxsYG1tbWmDRpEurXrw8fH5+3fkxERERU8ZT73VjFmTJlCjIzMzFy5EgkJyejefPmOHz4MMzNzdVtlixZAn19ffTu3RuZmZnw9vZGWFgY9PT0yrFyIiIiqigqVNg5ceKExnOFQoGgoCAEBQUVuY2xsTGWL1+O5cuXl21xRERE9E4q9w8VJCIiIipLDDtEREQkaww7REREJGsMO0RERCRrDDtEREQkaww7REREJGsMO0RERCRrDDtEREQkaww7REREJGsMO0RERCRrFerrIqhsOX97oFTb3Z7fRceVEBERvT2c2SEiIiJZY9ghIiIiWWPYISIiIllj2CEiIiJZY9ghIiIiWWPYISIiIllj2CEiIiJZY9ghIiIiWWPYISIiIllj2CEiIiJZY9ghIiIiWWPYISIiIllj2CEiIiJZkxx2Nm3aBJVKVWB5VlYWNm3apJOiiIiIiHRFctgZMmQIUlJSCixPS0vDkCFDdFIUERERka5IDjtCCCgUigLL79+/D0tLS50URURERKQr+to2bNSoERQKBRQKBby9vaGv/3+b5uTkID4+Hh07diyTIomIiIhKS+uw06NHDwBATEwM/Pz8ULlyZfU6Q0NDODs749NPP9V5gURERERvQuuwM3PmTACAs7Mz+vTpA2Nj4zIrioiIiEhXtA47eQICAgC8uvsqKSkJubm5Gutr1Kihm8qIiIiIdEBy2ImLi8PQoUMRHR2tsTzvwuWcnBydFUdERET0piSHncGDB0NfXx/79++Hvb19oXdmEREREVUUksNOTEwMzp8/j9q1a5dFPUREREQ6JflzdurWrYsnT56URS1EREREOic57CxYsABTpkzBiRMn8PTpU6Smpmo8iIiIiCoSyaexfHx8AADe3t4ay3mBMhEREVVEksPO8ePHy6IOIiIiojIhOex4eXmVRR1EREREZUJy2Pn999+LXd+mTZtSF0NERESka5LDTtu2bQsse/2zdnjNDhEREVUkku/GSk5O1ngkJSUhIiICzZo1w+HDh8uiRiIiIqJSkzyzY2lpWWCZr68vjIyMMH78eJw/f14nhRERERHpguSZnaJUq1YN169f11V3RERERDoheWbn8uXLGs+FEEhISMD8+fPRsGFDnRVGREREpAuSw85HH30EhUIBIYTG8hYtWmDDhg06K4yIiIhIFySHnfj4eI3nlSpVQrVq1WBsbKyzooiIiIh0RXLYcXJyKos6iIiIiMpEqS5QjoqKgr+/P2rVqgVXV1d069YNf/zxh65rIyIiInpjksNOeHg4fHx8YGpqirFjx2L06NEwMTGBt7c3tmzZUhY1EhEREZWa5NNYP/zwAxYuXIjx48erl40bNw4hISGYM2cO+vXrp9MCiYiIiN6E5JmdW7duwd/fv8Dybt26Fbh4mYiIiKi8SQ47jo6OOHr0aIHlR48ehaOjo06KIiIiItIVyaexJk6ciLFjxyImJgaenp5QKBQ4efIkwsLCsGzZsrKokYiIiKjUJIedb775BkqlEosXL8Yvv/wCAKhTpw62b9+O7t2767xAIiIiojchOewAQM+ePdGzZ09d10JERESkc6UKO3nS09ORm5ursczCwuKNCiIiIiLSJckXKMfHx6NLly4wMzODpaUlrKysYGVlhSpVqsDKyqosaiQiIiIqNclhp3///khOTsaGDRtw9OhRHDt2DMeOHcPx48dx7NgxSX2tWrUKDRo0gIWFBSwsLNCyZUv89ttv6vVCCAQFBcHBwQEmJiZo27Ytrl69qtGHSqXCmDFjULVqVZiZmaFbt264f/++1MMiIiIimZJ8Guvy5cs4f/483N3d33jn1atXx/z581GrVi0AwMaNG9G9e3dcvHgR9erVw8KFCxESEoKwsDC4ublh7ty58PX1xfXr12Fubg4ACAwMxL59+7Bt2zbY2Nhg4sSJ6Nq1K86fPw89Pb03rpGIiIjebZJndpo1a4Z79+7pZOf+/v7o3Lkz3Nzc4Obmhh9++AGVK1fG6dOnIYTA0qVLMX36dPTq1QseHh7YuHEjnj9/rv5aipSUFKxfvx6LFy+Gj48PGjVqhPDwcMTGxuLIkSM6qZGIiIjebZJndtatW4evv/4aDx48gIeHBwwMDDTWN2jQoFSF5OTk4Ndff0VGRgZatmyJ+Ph4JCYmokOHDuo2RkZG8PLyQnR0NEaMGIHz58/j5cuXGm0cHBzg4eGB6Oho+Pn5laoWIiIikg/JYefx48e4efMmhgwZol6mUCgghIBCoUBOTo6k/mJjY9GyZUu8ePEClStXxq5du1C3bl1ER0cDAOzs7DTa29nZ4c6dOwCAxMREGBoaFrgw2s7ODomJiUXuU6VSQaVSqZ+npqZKqpmIiIjeHZLDztChQ9GoUSNs3boVdnZ2UCgUb1SAu7s7YmJi8M8//2DHjh0ICAhAVFSUen3+/vNCVXFKajNv3jzMmjXrjeomIiKid4PksHPnzh3s3btXfVHxmzI0NFT31bRpU5w9exbLli3D1KlTAbyavbG3t1e3T0pKUs/2KJVKZGVlITk5WWN2JykpCZ6enkXuc9q0aZgwYYL6eWpqKr/Xi4iISKYkX6Dcvn17XLp0qSxqAfBqVkalUsHFxQVKpRKRkZHqdVlZWYiKilIHmSZNmsDAwECjTUJCAq5cuVJs2DEyMlLf7p73ICIiInmSPLPj7++P8ePHIzY2FvXr1y9wgXK3bt207uu7775Dp06d4OjoiLS0NGzbtg0nTpxAREQEFAoFAgMDERwcDFdXV7i6uiI4OBimpqbo168fAMDS0hLDhg3DxIkTYWNjA2tra0yaNAn169eHj4+P1EMjIiIiGZIcdr7++msAwOzZswusk3qB8qNHjzBw4EAkJCTA0tISDRo0QEREBHx9fQEAU6ZMQWZmJkaOHInk5GQ0b94chw8fVn/GDgAsWbIE+vr66N27NzIzM+Ht7Y2wsDB+xg4REREBKEXYyf9dWG9i/fr1xa5XKBQICgpCUFBQkW2MjY2xfPlyLF++XGd1ERERkXxIvmaHiIiI6F1Sqm89z8jIQFRUFO7evYusrCyNdWPHjtVJYURERES6IDnsXLx4EZ07d8bz58+RkZEBa2trPHnyBKamprC1tWXYISIiogpF8mms8ePHw9/fH8+ePYOJiQlOnz6NO3fuoEmTJvjxxx/LokYiIiKiUpMcdmJiYjBx4kTo6elBT08PKpUKjo6OWLhwIb777ruyqJGIiIio1CSHHQMDA/VXMdjZ2eHu3bsAXn3mTd6/iYiIiCoKydfsNGrUCOfOnYObmxvatWuHf/3rX3jy5Ak2b96M+vXrl0WNRERERKUmeWYnODhY/V1Vc+bMgY2NDb755hskJSVh7dq1Oi+QiIiI6E1Intlp2rSp+t/VqlXDwYMHdVoQERERkS7xQwWJiIhI1hh2iIiISNYYdoiIiEjWGHaIiIhI1iSHnfj4+LKog4iIiKhMSA47tWrVQrt27RAeHo4XL16URU1EREREOiM57Fy6dAmNGjXCxIkToVQqMWLECJw5c6YsaiMiIiJ6Y5LDjoeHB0JCQvDgwQOEhoYiMTERrVu3Rr169RASEoLHjx+XRZ1EREREpVLqC5T19fXRs2dP/PLLL1iwYAFu3ryJSZMmoXr16hg0aBASEhJ0WScRERFRqZQ67Jw7dw4jR46Evb09QkJCMGnSJNy8eRPHjh3DgwcP0L17d13WSURERFQqkr8uIiQkBKGhobh+/To6d+6MTZs2oXPnzqhU6VVucnFxwZo1a1C7dm2dF0tEREQkleSws2rVKgwdOhRDhgyBUqkstE2NGjWwfv36Ny6O5Mv52wOl2u72/C46roSIiOROctiJi4srsY2hoSECAgJKVRARERGRLkm+Zic0NBS//vprgeW//vorNm7cqJOiiIiIiHRFctiZP38+qlatWmC5ra0tgoODdVIUERERka5IDjt37tyBi4tLgeVOTk64e/euTooiIiIi0hXJYcfW1haXL18usPzSpUuwsbHRSVFEREREuiI57PTt2xdjx47F8ePHkZOTg5ycHBw7dgzjxo1D3759y6JGIiIiolKTfDfW3LlzcefOHXh7e0Nf/9Xmubm5GDRoEK/ZISIiogpHctgxNDTE9u3bMWfOHFy6dAkmJiaoX78+nJycyqI+omLx83qIiKgkksNOHjc3N7i5uemyFiIiIiKdkxx2cnJyEBYWhqNHjyIpKQm5ubka648dO6az4oiIiIjelOSwM27cOISFhaFLly7w8PCAQqEoi7qIiIiIdEJy2Nm2bRt++eUXdO7cuSzqISIiItIpybeeGxoaolatWmVRCxEREZHOSQ47EydOxLJlyyCEKIt6iIiIiHRK8mmskydP4vjx4/jtt99Qr149GBgYaKzfuXOnzoojIiIielOSw06VKlXQs2fPsqiFiIiISOckh53Q0NCyqIOIiIioTEi+ZgcAsrOzceTIEaxZswZpaWkAgIcPHyI9PV2nxRERERG9KckzO3fu3EHHjh1x9+5dqFQq+Pr6wtzcHAsXLsSLFy+wevXqsqiTqMzwKyeIiORN8szOuHHj0LRpUyQnJ8PExES9vGfPnjh69KhOiyMiIiJ6U6W6G+vUqVMwNDTUWO7k5IQHDx7orDAiIiIiXZA8s5Obm4ucnJwCy+/fvw9zc3OdFEVERESkK5LDjq+vL5YuXap+rlAokJ6ejpkzZ/IrJIiIiKjCkXwaa8mSJWjXrh3q1q2LFy9eoF+/foiLi0PVqlWxdevWsqiRiIiIqNQkhx0HBwfExMRg69atuHDhAnJzczFs2DD0799f44JlIiIioopActgBABMTEwwdOhRDhw7VdT1EREREOiU57GzatKnY9YMGDSp1MURERES6JjnsjBs3TuP5y5cv8fz5cxgaGsLU1JRhh4iIiCoUyXdjJScnazzS09Nx/fp1tG7dmhcoExERUYVTqu/Gys/V1RXz588vMOtDREREVN50EnYAQE9PDw8fPtRVd0REREQ6Ifmanb1792o8F0IgISEBK1asQKtWrXRWGBEREZEuSA47PXr00HiuUChQrVo1tG/fHosXL9ZVXUTvHH57OhFRxSQ57OTm5pZFHURERERlQmfX7BARERFVRJJndiZMmKB125CQEKndExEREemU5LBz8eJFXLhwAdnZ2XB3dwcA3LhxA3p6emjcuLG6nUKh0F2VRERERKUkOez4+/vD3NwcGzduhJWVFYBXHzQ4ZMgQfPLJJ5g4caLOiyQiIiIqLclhZ/HixTh8+LA66ACAlZUV5s6diw4dOkgKO/PmzcPOnTvxv//9DyYmJvD09MSCBQvUM0bAq1vbZ82ahbVr1yI5ORnNmzfHTz/9hHr16qnbqFQqTJo0CVu3bkVmZia8vb2xcuVKVK9eXerhEckC7wwjIvo/ki9QTk1NxaNHjwosT0pKQlpamqS+oqKiMGrUKJw+fRqRkZHIzs5Ghw4dkJGRoW6zcOFChISEYMWKFTh79iyUSiV8fX019hUYGIhdu3Zh27ZtOHnyJNLT09G1a1fk5ORIPTwiIiKSGckzOz179sSQIUOwePFitGjRAgBw+vRpTJ48Gb169ZLUV0REhMbz0NBQ2Nra4vz582jTpg2EEFi6dCmmT5+u7nvjxo2ws7PDli1bMGLECKSkpGD9+vXYvHkzfHx8AADh4eFwdHTEkSNH4OfnJ/UQiYiISEYkz+ysXr0aXbp0wYABA+Dk5AQnJyf0798fnTp1wsqVK9+omJSUFACAtbU1ACA+Ph6JiYno0KGDuo2RkRG8vLwQHR0NADh//jxevnyp0cbBwQEeHh7qNkRERPT+kjyzY2pqipUrV2LRokW4efMmhBCoVasWzMzM3qgQIQQmTJiA1q1bw8PDAwCQmJgIALCzs9Noa2dnhzt37qjbGBoaalxDlNcmb/v8VCoVVCqV+nlqauob1U5EREQVV6k/VDAhIQEJCQlwc3ODmZkZhBBvVMjo0aNx+fJlbN26tcC6/LexCyFKvLW9uDbz5s2DpaWl+uHo6Fj6womIiKhCkxx2nj59Cm9vb7i5uaFz585ISEgAAAwfPrzUt52PGTMGe/fuxfHjxzXuoFIqlQBQYIYmKSlJPdujVCqRlZWF5OTkItvkN23aNKSkpKgf9+7dK1XdREREVPFJDjvjx4+HgYEB7t69C1NTU/XyPn36FLjguCRCCIwePRo7d+7EsWPH4OLiorHexcUFSqUSkZGR6mVZWVmIioqCp6cnAKBJkyYwMDDQaJOQkIArV66o2+RnZGQECwsLjQcRERHJk+Rrdg4fPoxDhw4V+AwbV1dX9XU02ho1ahS2bNmCPXv2wNzcXD2DY2lpCRMTEygUCgQGBiI4OBiurq5wdXVFcHAwTE1N0a9fP3XbYcOGYeLEibCxsYG1tTUmTZqE+vXrq+/OIiIioveX5LCTkZGhMaOT58mTJzAyMpLU16pVqwAAbdu21VgeGhqKwYMHAwCmTJmCzMxMjBw5Uv2hgocPH4a5ubm6/ZIlS6Cvr4/evXurP1QwLCwMenp60g6OiIiIZEdy2GnTpg02bdqEOXPmAHh18XBubi4WLVqEdu3aSepLm4uaFQoFgoKCEBQUVGQbY2NjLF++HMuXL5e0f6KKhp98TESke5LDzqJFi9C2bVucO3cOWVlZmDJlCq5evYpnz57h1KlTZVEjERERUalJvkC5bt26uHz5Mj7++GP4+voiIyMDvXr1wsWLF/Hhhx+WRY1EREREpSZpZifvk4rXrFmDWbNmlVVNRERERDojaWbHwMAAV65cKfED/YiIiIgqCsmnsQYNGoT169eXRS1EREREOif5AuWsrCysW7cOkZGRaNq0aYHvxAoJCdFZcURERERvSnLYuXLlCho3bgwAuHHjhsY6nt4iIiKiikbrsHPr1i24uLjg+PHjZVkPERERkU5pfc2Oq6srHj9+rH7ep08fPHr0qEyKIiIiItIVrcNO/k87PnjwIDIyMnReEBEREZEuSb4bi4iIiOhdonXYUSgUBS5A5gXJREREVNFpfYGyEAKDBw9Wf7P5ixcv8PXXXxe49Xznzp26rZCIiIjoDWgddgICAjSeDxgwQOfFEBGVJX6rPNH7SeuwExoaWpZ1EBEREZUJyR8qSETvB86CEJFc8G4sIiIikjWGHSIiIpI1nsYiojKli9NhPKVGRG+CMztEREQkaww7REREJGs8jUVEVA54aq5wpXld5P6a0JvjzA4RERHJGsMOERERyRrDDhEREckaww4RERHJGsMOERERyRrvxiIiIlnhnW5l5119bTmzQ0RERLLGsENERESyxtNYRETvqHf1lALR28aZHSIiIpI1zuwQEUnA2RSSQhfjRVdj7n0eu5zZISIiIlnjzA4R0Xvuff4fP70fOLNDREREssaZHSIiemNymx2S2/G87zizQ0RERLLGsENERESyxrBDREREssawQ0RERLLGsENERESyxrBDREREssawQ0RERLLGsENERESyxrBDREREssawQ0RERLLGsENERESyxrBDREREssawQ0RERLLGsENERESyxrBDREREssawQ0RERLLGsENERESyxrBDREREssawQ0RERLLGsENERESyxrBDREREssawQ0RERLJWrmHn999/h7+/PxwcHKBQKLB7926N9UIIBAUFwcHBASYmJmjbti2uXr2q0UalUmHMmDGoWrUqzMzM0K1bN9y/f/8tHgURERFVZOUadjIyMtCwYUOsWLGi0PULFy5ESEgIVqxYgbNnz0KpVMLX1xdpaWnqNoGBgdi1axe2bduGkydPIj09HV27dkVOTs7bOgwiIiKqwPTLc+edOnVCp06dCl0nhMDSpUsxffp09OrVCwCwceNG2NnZYcuWLRgxYgRSUlKwfv16bN68GT4+PgCA8PBwODo64siRI/Dz83trx0JEREQVU4W9Zic+Ph6JiYno0KGDepmRkRG8vLwQHR0NADh//jxevnyp0cbBwQEeHh7qNkRERPR+K9eZneIkJiYCAOzs7DSW29nZ4c6dO+o2hoaGsLKyKtAmb/vCqFQqqFQq9fPU1FRdlU1EREQVTIWd2cmjUCg0ngshCizLr6Q28+bNg6Wlpfrh6Oiok1qJiIio4qmwYUepVAJAgRmapKQk9WyPUqlEVlYWkpOTi2xTmGnTpiElJUX9uHfvno6rJyIiooqiwoYdFxcXKJVKREZGqpdlZWUhKioKnp6eAIAmTZrAwMBAo01CQgKuXLmiblMYIyMjWFhYaDyIiIhInsr1mp309HT8/fff6ufx8fGIiYmBtbU1atSogcDAQAQHB8PV1RWurq4IDg6Gqakp+vXrBwCwtLTEsGHDMHHiRNjY2MDa2hqTJk1C/fr11XdnERER0futXMPOuXPn0K5dO/XzCRMmAAACAgIQFhaGKVOmIDMzEyNHjkRycjKaN2+Ow4cPw9zcXL3NkiVLoK+vj969eyMzMxPe3t4ICwuDnp7eWz8eIiIiqnjKNey0bdsWQogi1ysUCgQFBSEoKKjINsbGxli+fDmWL19eBhUSERHRu67CXrNDREREpAsMO0RERCRrDDtEREQkaww7REREJGsMO0RERCRrDDtEREQkaww7REREJGsMO0RERCRrDDtEREQkaww7REREJGsMO0RERCRrDDtEREQkaww7REREJGsMO0RERCRrDDtEREQkaww7REREJGsMO0RERCRrDDtEREQkaww7REREJGsMO0RERCRrDDtEREQkaww7REREJGsMO0RERCRrDDtEREQkaww7REREJGsMO0RERCRrDDtEREQkaww7REREJGsMO0RERCRrDDtEREQkaww7REREJGsMO0RERCRrDDtEREQkaww7REREJGsMO0RERCRrDDtEREQkaww7REREJGsMO0RERCRrDDtEREQkaww7REREJGsMO0RERCRrDDtEREQkaww7REREJGsMO0RERCRrDDtEREQkaww7REREJGsMO0RERCRrDDtEREQkaww7REREJGsMO0RERCRrDDtEREQkaww7REREJGsMO0RERCRrDDtEREQkaww7REREJGsMO0RERCRrDDtEREQkaww7REREJGsMO0RERCRrsgk7K1euhIuLC4yNjdGkSRP88ccf5V0SERERVQCyCDvbt29HYGAgpk+fjosXL+KTTz5Bp06dcPfu3fIujYiIiMqZLMJOSEgIhg0bhuHDh6NOnTpYunQpHB0dsWrVqvIujYiIiMrZOx92srKycP78eXTo0EFjeYcOHRAdHV1OVREREVFFoV/eBbypJ0+eICcnB3Z2dhrL7ezskJiYWOg2KpUKKpVK/TwlJQUAkJqaqvP6clXPS7Vd/lp00Q9rYS2shbXIsRa5Hc/7UIuu5PUrhCi+oXjHPXjwQAAQ0dHRGsvnzp0r3N3dC91m5syZAgAffPDBBx988CGDx71794rNCu/8zE7VqlWhp6dXYBYnKSmpwGxPnmnTpmHChAnq57m5uXj27BlsbGygUCjKtN48qampcHR0xL1792BhYVGu/bAW1sJaWAtreXt9sBbdEUIgLS0NDg4OxbZ758OOoaEhmjRpgsjISPTs2VO9PDIyEt27dy90GyMjIxgZGWksq1KlSlmWWSQLCwudDAxd9MNaWAtrYS2s5e31wVp0w9LSssQ273zYAYAJEyZg4MCBaNq0KVq2bIm1a9fi7t27+Prrr8u7NCIiIipnsgg7ffr0wdOnTzF79mwkJCTAw8MDBw8ehJOTU3mXRkREROVMFmEHAEaOHImRI0eWdxlaMzIywsyZMwucTiuPflgLa2EtrIW1vL0+WMvbpxCipPu1iIiIiN5d7/yHChIREREVh2GHiIiIZI1hh4iIiGSNYYeIiIhkjWGnHPz+++/w9/eHg4MDFAoFdu/eLWn7efPmoVmzZjA3N4etrS169OiB69evS65j1apVaNCggfpDoFq2bInffvtNcj/5a1MoFAgMDJS0XVBQEBQKhcZDqVRK3v+DBw8wYMAA2NjYwNTUFB999BHOnz8vqQ9nZ+cCtSgUCowaNUrrPrKzszFjxgy4uLjAxMQENWvWxOzZs5Gbmyv1kJCWlobAwEA4OTnBxMQEnp6eOHv2bJHtSxpfQggEBQXBwcEBJiYmaNu2La5evSq5n507d8LPzw9Vq1aFQqFATEyMpD5evnyJqVOnon79+jAzM4ODgwMGDRqEhw8fSq4lKCgItWvXhpmZGaysrODj44P//ve/kvp43YgRI6BQKLB06VLJtQwePLjA2GnRooXkWq5du4Zu3brB0tIS5ubmaNGiBe7evSupn8LGsUKhwKJFi7TuIz09HaNHj0b16tVhYmKCOnXqYNWqVZJfl0ePHmHw4MFwcHCAqakpOnbsiLi4OI022ry3lTR+telDm7FbUj/ajF9tatFm7Ep9zy9s/GrTR0ljV9s6tBm7bxvDTjnIyMhAw4YNsWLFilJtHxUVhVGjRuH06dOIjIxEdnY2OnTogIyMDEn9VK9eHfPnz8e5c+dw7tw5tG/fHt27dy/0D582zp49i7Vr16JBgwal2r5evXpISEhQP2JjYyVtn5ycjFatWsHAwAC//fYb/vrrLyxevFjyp2OfPXtWo47IyEgAwOeff651HwsWLMDq1auxYsUKXLt2DQsXLsSiRYuwfPlySbUAwPDhwxEZGYnNmzcjNjYWHTp0gI+PDx48eFBo+5LG18KFCxESEoIVK1bg7NmzUCqV8PX1RVpamqR+MjIy0KpVK8yfP7/I2ovr4/nz57hw4QK+//57XLhwATt37sSNGzfQrVs3ycfk5uaGFStWIDY2FidPnoSzszM6dOiAx48fa91Hnt27d+O///1vkR8/r00/HTt21BhDBw8elNTHzZs30bp1a9SuXRsnTpzApUuX8P3338PY2FhSP6/XkJCQgA0bNkChUODTTz/Vuo/x48cjIiIC4eHhuHbtGsaPH48xY8Zgz549WtcihECPHj1w69Yt7NmzBxcvXoSTkxN8fHw03re0eW8rafxq04c2Y7ekfrQZv9rUos3YlfKeX9T41baP4sauNn1oO3bfOh18Fye9AQBi165db9RHUlKSACCioqLeuB4rKyuxbt06ydulpaUJV1dXERkZKby8vMS4ceMkbT9z5kzRsGFDyft93dSpU0Xr1q3fqI/CjBs3Tnz44YciNzdX6226dOkihg4dqrGsV69eYsCAAZL2/fz5c6Gnpyf279+vsbxhw4Zi+vTpJW6ff3zl5uYKpVIp5s+fr1724sULYWlpKVavXq11P6+Lj48XAMTFixcl1VKYM2fOCADizp07b9RPSkqKACCOHDkiqY/79++LDz74QFy5ckU4OTmJJUuWFLufwvoJCAgQ3bt3L3a7kvro06eP5LGizevSvXt30b59e0l91KtXT8yePVtjWePGjcWMGTO07uf69esCgLhy5Yp6WXZ2trC2thY///xzkf3kf28rzfgt7v1R27FbUj95Shq/2vRR0tgtrh8p47ewPqSO3cL6KM3YfRs4syMDKSkpAABra+tS95GTk4Nt27YhIyMDLVu2lLz9qFGj0KVLF/j4+JS6hri4ODg4OMDFxQV9+/bFrVu3JG2/d+9eNG3aFJ9//jlsbW3RqFEj/Pzzz6WuBwCysrIQHh6OoUOHSvqS2NatW+Po0aO4ceMGAODSpUs4efIkOnfuLGn/2dnZyMnJKfC/IhMTE5w8eVJSXwAQHx+PxMREdOjQQb3MyMgIXl5eiI6OltyfrqWkpEChULzRd9VlZWVh7dq1sLS0RMOGDbXeLjc3FwMHDsTkyZNRr169Uu8fAE6cOAFbW1u4ubnhyy+/RFJSkqQ6Dhw4ADc3N/j5+cHW1hbNmzeXfLo7v0ePHuHAgQMYNmyYpO1at26NvXv34sGDBxBC4Pjx47hx4wb8/Py07kOlUgGAxjjW09ODoaFhseM4/3tbacavLt4fte2npPFbUh/ajt3C+pE6fouqRcrYzd9HWY1dnSjvtPW+wxvO7OTm5gp/f/9Sz2hcvnxZmJmZCT09PWFpaSkOHDgguY+tW7cKDw8PkZmZKYQQpZrZOXjwoPjPf/4jLl++rJ4dsrOzE0+ePNG6DyMjI2FkZCSmTZsmLly4IFavXi2MjY3Fxo0bJdXyuu3btws9PT3x4MEDSdvl5uaKb7/9VigUCqGvry8UCoUIDg4uVQ0tW7YUXl5e4sGDByI7O1ts3rxZKBQK4ebmVuK2+cfXqVOnBIACx/Pll1+KDh06aN3P63Q1s5OZmSmaNGki+vfvX6p+9u3bJ8zMzIRCoRAODg7izJkzkvoIDg4Wvr6+6hm80s7sbNu2Tezfv1/ExsaKvXv3ioYNG4p69eqJFy9eaNVHQkKCACBMTU1FSEiIuHjxopg3b55QKBTixIkTkmp53YIFC4SVlZX691TbPlQqlRg0aJAAIPT19YWhoaHYtGlTkX0U1k9WVpZwcnISn3/+uXj27JlQqVRi3rx5AkCR466w9zap47ek90dtx64277Mljd/i+pAydovqR8r4LaoPKWO3sD5KO3bfBoadcvamYWfkyJHCyclJ3Lt3r1Tbq1QqERcXJ86ePSu+/fZbUbVqVXH16lWtt797966wtbUVMTEx6mWlCTv5paenCzs7O7F48WKttzEwMBAtW7bUWDZmzBjRokWLUtfRoUMH0bVrV8nbbd26VVSvXl1s3bpVXL58WWzatElYW1uLsLAwyX39/fffok2bNgKA0NPTE82aNRP9+/cXderUKXHbosLOw4cPNdoNHz5c+Pn5ad3P63QRdrKyskT37t1Fo0aNREpKSqn6SU9PF3FxceLPP/8UQ4cOFc7OzuLRo0da9XHu3DlhZ2en8Ue0tGEnv4cPHwoDAwOxY8cOrfp48OCBACC++OILjXb+/v6ib9++pa7F3d1djB49uthaC+tj0aJFws3NTezdu1dcunRJLF++XFSuXFlERkZK6ufcuXOiYcOG6nHs5+cnOnXqJDp16lRoH4W9t0kdvyW9P2o7dkvqR5vxW1wfUsZuYf1IHb/a/t0obuwW1kdpx+7bwLBTzt4k7IwePVpUr15d3Lp1S2f1eHt7i6+++krr9rt27VK/eeU9AAiFQiH09PREdnZ2qWvx8fERX3/9tdbta9SoIYYNG6axbOXKlcLBwaFU+799+7aoVKmS2L17t+Rtq1evLlasWKGxbM6cOcLd3b1UtQjx6g0x702+d+/eonPnziVuk3983bx5UwAQFy5c0GjXrVs3MWjQIK37ed2bhp2srCzRo0cP0aBBA61m8rT9nalVq1aRs2n5+1iyZIl6zL4+jitVqiScnJx0Usvr15kU14dKpRL6+vpizpw5Gu2mTJkiPD09S1XL77//LgBo/KdEmz6eP38uDAwMClwzNmzYsFKH43/++UckJSUJIYT4+OOPxciRIwu0Keq9Tcr41eb9UZuxW1I/2oxfqe/VRY3dovqRMn5LU0v+sVtUH6Udu28Dr9l5BwkhMHr0aOzcuRPHjh2Di4uLTvvOO7+uDW9vb8TGxiImJkb9aNq0Kfr374+YmBjo6emVqg6VSoVr167B3t5e621atWpV4DbIGzduwMnJqVQ1hIaGwtbWFl26dJG87fPnz1Gpkuavl56eXqluPc9jZmYGe3t7JCcn49ChQ+jevbvkPlxcXKBUKtV3mAGvrhOIioqCp6dnqWsrrZcvX6J3796Ii4vDkSNHYGNjo7O+pYzlgQMH4vLlyxrj2MHBAZMnT8ahQ4feqI6nT5/i3r17Wo9lQ0NDNGvWTKdjef369WjSpImka5iAVz+fly9f6nQsW1paolq1aoiLi8O5c+c0xnFJ723ajF9dvT9q009J47e0teQfuyX1o834LU0t+cduSX2UxdjVFdl86/m7JD09HX///bf6eXx8PGJiYmBtbY0aNWqUuP2oUaOwZcsW7NmzB+bm5khMTATw6k3ExMRE6zq+++47dOrUCY6OjkhLS8O2bdtw4sQJREREaN2Hubk5PDw8NJaZmZnBxsamwPLiTJo0Cf7+/qhRowaSkpIwd+5cpKamIiAgQOs+xo8fD09PTwQHB6N37944c+YM1q5di7Vr12rdR57c3FyEhoYiICAA+vrSf038/f3xww8/oEaNGqhXrx4uXryIkJAQDB06VHJfhw4dghAC7u7u+PvvvzF58mS4u7tjyJAhhbYvaXwFBgYiODgYrq6ucHV1RXBwMExNTdGvXz9J/Tx79gx3795Vf65I3hucUqlUf0ZScX04ODjgs88+w4ULF7B//37k5OSox7K1tTUMDQ21qsXGxgY//PADunXrBnt7ezx9+hQrV67E/fv3NT4uoKTjyf+HysDAAEqlEu7u7lq/LtbW1ggKCsKnn34Ke3t73L59G9999x2qVq2Knj17al3L5MmT0adPH7Rp0wbt2rVDREQE9u3bhxMnTkj6GQFAamoqfv31VyxevBiFKakPLy8vTJ48GSYmJnByckJUVBQ2bdqEkJAQSf38+uuvqFatGmrUqIHY2FiMGzcOPXr00LjYuKT3trzP8Cpu/Grz/qjN2C2pn+zs7BLHb0l9ZGRkaDV2S+rHxsamxPFbUh/p6ekljl1tXlttx+5bVz4TSu+348ePCwAFHgEBAVptX9i2AERoaKikOoYOHSqcnJyEoaGhqFatmvD29haHDx+WfkD5lOaanT59+gh7e3thYGAgHBwcRK9evSRdO5Rn3759wsPDQxgZGYnatWuLtWvXSu5DCCEOHTokAIjr16+XavvU1FQxbtw4UaNGDWFsbCxq1qwppk+fLlQqleS+tm/fLmrWrCkMDQ2FUqkUo0aNEv/880+R7UsaX7m5uWLmzJlCqVQKIyMj0aZNGxEbGyu5n9DQ0ELXz5w5U6s+8k4hFPY4fvy41rVkZmaKnj17CgcHB2FoaCjs7e1Ft27dClzkKfX3rqhrHorr5/nz56JDhw6iWrVqwsDAQNSoUUMEBASIu3fvSq5l/fr1olatWsLY2Fg0bNiw0NOp2vSzZs0aYWJiUuSYKamPhIQEMXjwYOHg4CCMjY2Fu7u7WLx4cYGPYiipn2XLlonq1aurX5cZM2YU+H3Q5r2tpPGrTR/ajN2S+tFm/JbUh7ZjtzTv+fnHb0l9aDN2ta1Dm7H7timEEAJEREREMsVrdoiIiEjWGHaIiIhI1hh2iIiISNYYdoiIiEjWGHaIiIhI1hh2iIiISNYYdoiIiEjWGHaICqFQKLB79+7yLuOd07ZtWwQGBr6VfYWFhaFKlSqSthk8eDB69OhRJvUUZu3atXB0dESlSpWwdOnSAutv374NhUKBmJiYt1YT0fuIYYfeG4MHD4ZCoYBCoYCBgQHs7Ozg6+uLDRs2FPien4SEBHTq1Emrft+lYKSrP/YnTpyAQqHAP//888Z9lVafPn1w48YNnffr7OxcaDCRKjU1FaNHj8bUqVPx4MEDfPXVV29e3HuCIZB0jWGH3isdO3ZEQkICbt++jd9++w3t2rXDuHHj0LVrV2RnZ6vbKZVKGBkZlWOlVJyXL1/CxMQEtra25V1Kke7evYuXL1+iS5cusLe3h6mpaXmXJNnLly/Lu4Q3JodjoDfHsEPvFSMjIyiVSnzwwQdo3LgxvvvuO+zZswe//fYbwsLC1O1en63JysrC6NGjYW9vD2NjYzg7O2PevHkAXs0CAEDPnj2hUCjUz2/evInu3bvDzs4OlStXRrNmzXDkyBGNWpydnREcHIyhQ4fC3NwcNWrUKPClpffv30ffvn1hbW0NMzMzNG3aFP/973/V6/ft24cmTZrA2NgYNWvWxKxZszRC2+uCgoKwceNG7NmzRz3DlfflfLGxsWjfvr36SwW/+uorpKenF9rP7du30a5dOwCAlZUVFAoFBg8erF6fm5uLKVOmwNraGkqlEkFBQRrbp6Sk4KuvvoKtrS0sLCzQvn17XLp0qdB95e1PoVDgl19+Qdu2bWFsbIzw8PBCT2PNnTsXtra2MDc3x/Dhw/Htt9/io48+KtDnjz/+CHt7e9jY2GDUqFHqP4ht27bFnTt3MH78ePVrVJS7d++ie/fuqFy5MiwsLNC7d288evQIwKtTbPXr1wcA1KxZEwqFArdv3y6yr9dfuy+//BJubm64c+dOia/HJ598AhMTEzRr1gw3btzA2bNn0bRpU1SuXBkdO3bE48ePNbYNDQ1FnTp1YGxsjNq1a2PlypWF9vv661yYf/75B1999RXs7OxgbGwMDw8P7N+/X70+Ojoabdq0gYmJCRwdHTF27FhkZGSo15c09vO+TbtRo0ZQKBRo27atzo+B3jPl/eVcRG9LQECA6N69e6HrGjZsKDp16qR+DkDs2rVLCCHEokWLhKOjo/j999/F7du3xR9//CG2bNkihBAiKSlJ/UV4CQkJIikpSQghRExMjFi9erW4fPmyuHHjhpg+fbowNjYWd+7cUe/DyclJWFtbi59++knExcWJefPmiUqVKolr164JIYRIS0sTNWvWFJ988on4448/RFxcnNi+fbuIjo4WQggREREhLCwsRFhYmLh586Y4fPiwcHZ2FkFBQYUeY1pamujdu7fo2LGjSEhIEAkJCUKlUomMjAz1l6/GxsaKo0ePChcXlyK/IDM7O1vs2LFD/UWpCQkJ6i+Z9PLyEhYWFiIoKEjcuHFDbNy4USgUCvUXzObm5opWrVoJf39/cfbsWXHjxg0xceJEYWNjI54+fVro/vK+cNHZ2Vns2LFD3Lp1Szx48ECEhoYKS0tLdbvw8HBhbGwsNmzYIK5fvy5mzZolLCwsRMOGDdVtAgIChIWFhfj666/FtWvXxL59+4Spqan6C2OfPn0qqlevLmbPnq1+jQqTm5srGjVqJFq3bi3OnTsnTp8+LRo3biy8vLyEEK++VPHIkSMCgDhz5oxISEgQ2dnZRR7bxYsXhUqlEp9++qn46KOPxKNHjwrd7+vb1K5dW0RERIi//vpLtGjRQjRu3Fi0bdtWnDx5Uly4cEHUqlVLfP311+rt1q5dK+zt7dWv4Y4dO4S1tbUICwsr9nXOLycnR7Ro0ULUq1dPHD58WNy8eVPs27dPHDx4UAghxOXLl0XlypXFkiVLxI0bN8SpU6dEo0aNxODBg9V9lDT2z5w5IwCII0eOiISEBPXY0NUx0PuHYYfeG8WFnT59+og6deqon78edsaMGSPat29f4FueC2tbnLp164rly5ernzs5OYkBAwaon+fm5gpbW1uxatUqIcSrb6s2NzcvMgR88sknIjg4WGPZ5s2bhb29fZE1FPYarF27VlhZWYn09HT1sgMHDohKlSqJxMTEQvvJ+4br5ORkjeVeXl6idevWGsuaNWsmpk6dKoQQ4ujRo8LCwkK8ePFCo82HH34o1qxZU+i+8v6ALV26VGN5/rDTvHlzMWrUKI02rVq1KhB2nJycNILH559/Lvr06aN+XtS3nb/u8OHDQk9PT+Mboa9evaoON0IIcfHiRQFAxMfHF9lP3rH98ccfwsfHR7Rq1arYb7R/fZt169apl23dulUAEEePHlUvmzdvnnB3d1c/d3R0VIf0PHPmzBEtW7bU6Df/65zfoUOHRKVKlcT169cLXT9w4EDx1VdfaSz7448/RKVKlURmZqYQouSx/3oIfJ2ujoHePzyNRQRACFHkKYvBgwcjJiYG7u7uGDt2LA4fPlxifxkZGZgyZQrq1q2LKlWqoHLlyvjf//6Hu3fvarRr0KCB+t8KhQJKpRJJSUkAgJiYGDRq1AjW1taF7uP8+fOYPXs2KleurH58+eWXSEhIwPPnz7U9dFy7dg0NGzaEmZmZelmrVq2Qm5uL69eva91PYccEAPb29upjOn/+PNLT02FjY6NRd3x8PG7evFlsv02bNi12/fXr1/Hxxx9rLMv/HADq1asHPT29QuvT1rVr1+Do6AhHR0f1sryf9bVr1yT1BQBffPEF0tPTcfjwYVhaWqqXf/311xqv0+tef53t7OwAQH3qLG9Z3nE9fvwY9+7dw7BhwzT6mzt3boHXvaTXOSYmBtWrV4ebm1uh68+fP4+wsDCN/fj5+SE3Nxfx8fGF1p9/7BdGl8dA7x/98i6AqCK4du2a+jqB/Bo3boz4+Hj89ttvOHLkCHr37g0fHx/85z//KbK/yZMn49ChQ/jxxx9Rq1YtmJiY4LPPPkNWVpZGOwMDA43nCoVCfWeYiYlJsTXn5uZi1qxZ6NWrV4F1xsbGxW77uuKCXnHXrBSluGPKzc2Fvb29+lqh15V0G/nrYawo+esVQkiqT1tFvWbFvZbF6dy5M8LDw3H69Gm0b99evXz27NmYNGlSodu8fhx5+8y/7PXXHQB+/vlnNG/eXKOf14MfUPLrrM24HDFiBMaOHVtgXY0aNQqtP3+9RfUL6OYY6P3DsEPvvWPHjiE2Nhbjx48vso2FhQX69OmDPn364LPPPkPHjh3x7NkzWFtbw8DAADk5ORrt//jjDwwePBg9e/YEAKSnp2t1gerrGjRogHXr1qn3k1/jxo1x/fp11KpVS+s+DQ0NC9Rat25dbNy4ERkZGeo/EqdOnUKlSpWK/N+7oaEhABToqySNGzdGYmIi9PX11Rdz64q7uzvOnDmDgQMHqpedO3dOcj+FvUb51a1bF3fv3sW9e/fUszt//fUXUlJSUKdOHcn7/Oabb+Dh4YFu3brhwIED8PLyAgDY2trq5I4zOzs7fPDBB7h16xb69+//Rn01aNAA9+/fx40bNwodH40bN8bVq1cljcv8ChtfujwGev/wNBa9V1QqFRITE/HgwQNcuHABwcHB6N69O7p27YpBgwYVus2SJUuwbds2/O9//8ONGzfw66+/QqlUqmcinJ2dcfToUSQmJiI5ORkAUKtWLezcuRMxMTG4dOkS+vXrJ3n24IsvvoBSqUSPHj1w6tQp3Lp1Czt27MCff/4JAPjXv/6FTZs2ISgoCFevXsW1a9ewfft2zJgxo8g+nZ2dcfnyZVy/fh1PnjzBy5cv0b9/fxgbGyMgIABXrlzB8ePHMWbMGAwcOFB9eiQ/JycnKBQK7N+/H48fPy7yzq38fHx80LJlS/To0QOHDh3C7du3ER0djRkzZpQqmLxuzJgxWL9+PTZu3Ii4uDjMnTsXly9fljzT4uzsjN9//x0PHjzAkydPijyOBg0aoH///rhw4QLOnDmDQYMGwcvLq9SnUMaMGYO5c+eia9euOHnyZKn6KE5QUBDmzZuHZcuW4caNG4iNjUVoaChCQkIk9ePl5YU2bdrg008/RWRkpHrWMyIiAgAwdepU/Pnnnxg1ahRiYmIQFxeHvXv3YsyYMVrvw9bWFiYmJoiIiMCjR4+QkpKi02Og9w/DDr1XIiIiYG9vD2dnZ3Ts2BHHjx/Hv//9b+zZs6fAVHieypUrY8GCBWjatCmaNWuG27dv4+DBg6hU6dWvz+LFixEZGQlHR0c0atQIwKuAZGVlBU9PT/j7+8PPzw+NGzeWVKuhoSEOHz4MW1tbdO7cGfXr18f8+fPVdfr5+WH//v2IjIxEs2bN0KJFC4SEhMDJyanIPr/88ku4u7ujadOmqFatGk6dOgVTU1McOnQIz549Q7NmzfDZZ5/B29sbK1asKLKfDz74ALNmzcK3334LOzs7jB49WqtjUigUOHjwINq0aYOhQ4fCzc0Nffv2xe3bt4sMVtrq378/pk2bhkmTJqlPPQ4ePFjSKT3g1amj27dv48MPP0S1atWKPI7du3fDysoKbdq0gY+PD2rWrInt27e/0TEEBgZi1qxZ6Ny5M6Kjo9+or/yGDx+OdevWqW+L9/LyQlhYWJGnb4uzY8cONGvWDF988QXq1q2LKVOmqGdhGjRogKioKMTFxeGTTz5Bo0aN8P3338Pe3l7r/vX19fHvf/8ba9asgYODA7p3767zY6D3i0IUdlKbiEgGfH19oVQqsXnz5vIuhYjKEa/ZISJZeP78OVavXg0/Pz/o6elh69atOHLkCCIjI8u7NCIqZ5zZISJZyMzMhL+/Py5cuACVSgV3d3fMmDGj0LvViOj9wrBDREREssYLlImIiEjWGHaIiIhI1hh2iIiISNYYdoiIiEjWGHaIiIhI1hh2iIiISNYYdoiIiEjWGHaIiIhI1hh2iIiISNb+P4apWSpgBmV7AAAAAElFTkSuQmCC",
      "text/plain": [
       "<Figure size 640x480 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "array([[  1, 470],\n",
       "       [  2, 503],\n",
       "       [  3, 375],\n",
       "       [  4, 401],\n",
       "       [  5, 262],\n",
       "       [  6, 251],\n",
       "       [  7, 235],\n",
       "       [  8, 216],\n",
       "       [  9, 192],\n",
       "       [ 10, 207],\n",
       "       [ 11, 170],\n",
       "       [ 12, 165],\n",
       "       [ 13, 165],\n",
       "       [ 14, 172],\n",
       "       [ 15, 145],\n",
       "       [ 16, 161],\n",
       "       [ 17, 148],\n",
       "       [ 18, 142],\n",
       "       [ 19, 140],\n",
       "       [ 20, 161],\n",
       "       [ 21, 152],\n",
       "       [ 22, 140],\n",
       "       [ 23, 146],\n",
       "       [ 24, 145],\n",
       "       [ 25, 146],\n",
       "       [ 26, 153]])"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def frequency_graph(kmers, model, k=53):\n",
    "    \"\"\"\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    k : int\n",
    "        an odd number that specifies that length of the kmer with the\n",
    "        phosphorylation site in the middle\n",
    "    kmer : 2D array\n",
    "        1D array with kmers\n",
    "\n",
    "    model: Title for graph\n",
    "    Returns\n",
    "    -------\n",
    "    frequencies : 2D array\n",
    "        2D array with columns distance away from center kmer PTM and \n",
    "        frequency of nearby PTMS\n",
    "\n",
    "    \"\"\"\n",
    "    around_ind = k//2\n",
    "    x=0\n",
    "    z=0\n",
    "    frequencies = {}\n",
    "    for i in range(0, len(kmers)):\n",
    "        if \"@\" in kmers[i]:\n",
    "            x+=1\n",
    "        elif \"&\" in kmers[i]:\n",
    "            z+=1\n",
    "        temp = kmers[i][around_ind+1:]\n",
    "        for j in range(len(temp)):\n",
    "            if \"@\" == temp[j]:\n",
    "                if j+1 in frequencies.keys():\n",
    "                    frequencies[j+1] += 1\n",
    "                else:\n",
    "                    frequencies[j+1] = 1\n",
    "            elif \"&\" ==temp[j]:\n",
    "                if j+1 in frequencies.keys():\n",
    "                    frequencies[j+1] += 1\n",
    "                else:\n",
    "                    frequencies[j+1] = 1        \n",
    "        \n",
    "    \n",
    "    frequencies = list(frequencies.items())\n",
    "    frequencies = np.array(frequencies)\n",
    "    print (frequencies)\n",
    "    frequencies = frequencies[frequencies[:,0].argsort()]\n",
    "\n",
    "    plt.figure(0)\n",
    "    plt.bar(frequencies[:,0], frequencies[:,1])\n",
    "    plt.xticks(frequencies[:,0])\n",
    "    plt.title(\"Frequency of nearby PTM sites\")\n",
    "    plt.ylabel(\"Frequency amount\")\n",
    "    plt.xlabel(\"Distance to the right of k-mer center\")\n",
    "    plt.savefig(model+\"_frequencies.pdf\", transparent=True)\n",
    "    plt.show()\n",
    "\n",
    "    return frequencies\n",
    "\n",
    "\n",
    "lpos_kmers, lneg_kmers=make_kmers(protein_dict, PR_site_array, PR2_site_array, verbose=\"no\")\n",
    "frequency_graph (lpos_kmers, model=f'{folder}/PreSprint')\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cdb220f7-8ce9-43be-90d9-5896b44ac4d4",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
