{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "65f9fb51-73dd-4712-a448-aed203ab87bb",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from ast import literal_eval\n",
    "from tqdm import tqdm\n",
    "import random\n",
    "import tensorflow as tf\n",
    "import re\n",
    "import matplotlib\n",
    "from matplotlib import pyplot as plt\n",
    "import sys\n",
    "import copy\n",
    "matplotlib.rcParams['pdf.fonttype'] = 42\n",
    "matplotlib.rcParams['ps.fonttype'] = 42\n",
    "\n",
    "import tensorflow as tf\n",
    "from tensorflow.keras.models import Sequential\n",
    "from tensorflow.keras.layers import Dense, Conv2D, Dropout, Flatten, MaxPooling2D\n",
    "#from tensorflow.python.ops.numpy_ops import np_config\n",
    "#np_config.enable_numpy_behavior()\n",
    "from tensorflow.keras.regularizers import l2, l1\n",
    "from tensorflow.keras.optimizers import Adam\n",
    "from tensorflow.keras.layers import Embedding, LSTM, Dense, Dropout, InputLayer, Reshape\n",
    "from sklearn.metrics import roc_curve, auc, average_precision_score, matthews_corrcoef, f1_score, precision_score"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "a86748a0-f9f7-4cc8-8279-ba9e5033cfa1",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\Clair\\AppData\\Local\\Temp\\ipykernel_12512\\1261787865.py:7: ParserWarning: Falling back to the 'python' engine because the 'c' engine does not support regex separators (separators > 1 char and different from '\\s+' are interpreted as regex); you can avoid this warning by specifying engine='python'.\n",
      "  df = pd.read_csv(file_name, sep='delimiter', header=None)\n"
     ]
    }
   ],
   "source": [
    "#Put in master excel sheet from Sugiyama et al. \n",
    "excel1 = 'Train_Data_NGlycositeAtlas.csv'\n",
    "excel2 = 'Test_Data_NGlycositeAtlas.csv'\n",
    "#file_name = 'UP000005640_9606.txt' #Reference proteome for humans\n",
    "file_name = '20240202_NGlycosite_IDs.txt' #Uniprot seqs for Nglycosite atlas\n",
    "\n",
    "df = pd.read_csv(file_name, sep='delimiter', header=None)\n",
    "\n",
    "isoform_file = '20240202_NGlycosite_IDs_with_isoforms_short.fasta' #Uniprot seqs for Nglycosite atlas"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "e6acd21d-14bf-4eea-a88b-f50b2088cebc",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ID array has been made!\n",
      "Lines have been separated!\n",
      "protein_dict has been made!\n"
     ]
    }
   ],
   "source": [
    "#Load all possible Uniprot IDs from reference proteome\n",
    "df.fillna(\"NA\", inplace=True) # just in case any \"NA\" got turned into NaNs\n",
    "ID_df = df[df[0].str.contains(\">\")][0]\n",
    "ID_list = []\n",
    "for i in range(len(ID_df)):  \n",
    "    long=ID_df.iloc[i].split(\"|\")[1]\n",
    "    ID_list.append(long.split(\" \")[0])\n",
    "ID_array = np.array(ID_list)\n",
    "#print (ID_array)\n",
    "print(\"ID array has been made!\")\n",
    "\n",
    "#Makes array of sequences\n",
    "sequences = []\n",
    "indices_of_uniprot_ids = list(df[df[0].str.contains(\">\")].index)\n",
    "for i in range(len(indices_of_uniprot_ids)-1):\n",
    "    # concatenate all into one string\n",
    "    sequences.append(df[indices_of_uniprot_ids[i]+1:indices_of_uniprot_ids[i+1]][0].sum())\n",
    "    if i == len(indices_of_uniprot_ids)-2:\n",
    "        sequences.append(df[indices_of_uniprot_ids[i+1]+1:][0].sum())\n",
    "sequences = np.array(sequences)\n",
    "print(\"Lines have been separated!\")\n",
    "\n",
    "# Creates dictionary that maps Uniprot ID to sequence\n",
    "pairs = zip(ID_array, sequences)\n",
    "protein_dict = {A:B for A,B in pairs}\n",
    "print(\"protein_dict has been made!\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "e7388750-7036-4110-bba2-a9eab70c45a4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(26743, 3)\n"
     ]
    }
   ],
   "source": [
    "#Let's start with the data from NGlycoSite Atlas\n",
    "#Load data and convert to numpy array\n",
    "sheet1 = pd.read_csv(excel1)\n",
    "Data1=sheet1[[\"PID\", \"Position\", \"label\"]]\n",
    "\n",
    "sheet2 = pd.read_csv(excel2)\n",
    "Data2=sheet2[[\"PID\", \"Position\", \"label\"]]\n",
    "\n",
    "Train_ID=Data1.iloc[:, 0] \n",
    "Train_ID=Train_ID.to_numpy(dtype=str)\n",
    "\n",
    "Train_Position=Data1.iloc[:, 1] \n",
    "Train_Position=Train_Position.to_numpy(dtype=str)\n",
    "\n",
    "Train_Label=Data1.iloc[:, 2] \n",
    "Train_Label=Train_Label.to_numpy(dtype=str)\n",
    "\n",
    "Test_ID=Data2.iloc[:, 0] \n",
    "Test_ID=Test_ID.to_numpy(dtype=str)\n",
    "\n",
    "Test_Position=Data2.iloc[:, 1] \n",
    "Test_Position=Test_Position.to_numpy(dtype=str)\n",
    "\n",
    "Test_Label=Data2.iloc[:, 2] \n",
    "Test_Label=Test_Label.to_numpy(dtype=str)\n",
    "\n",
    "dataset=np.stack((np.concatenate((Train_ID, Test_ID)), np.concatenate((Train_Position, Test_Position)), np.concatenate((Train_Label, Test_Label))), axis=1)\n",
    "print (np.shape(dataset))\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "56517392-1aa1-4fb3-9c55-bee1e3c2cd7b",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\Clair\\AppData\\Local\\Temp\\ipykernel_12512\\2477744533.py:2: ParserWarning: Falling back to the 'python' engine because the 'c' engine does not support regex separators (separators > 1 char and different from '\\s+' are interpreted as regex); you can avoid this warning by specifying engine='python'.\n",
      "  idf = pd.read_csv(isoform_file, sep='delimiter', header=None)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Isoform ID array has been made!\n",
      "122\n",
      "Lines have been separated!\n",
      "protein_dict has been made!\n"
     ]
    }
   ],
   "source": [
    "#Now we need to add isoform sequences into dictionary\n",
    "idf = pd.read_csv(isoform_file, sep='delimiter', header=None)\n",
    "idf.fillna(\"NA\", inplace=True) # just in case any \"NA\" got turned into NaNs\n",
    "\n",
    "ID_idf = idf[idf[0].str.contains(\">\")][0]\n",
    "ID_ilist = []\n",
    "for i in range(len(ID_idf)):  #This dataset uses a different Uniprot form than previous so select it differently\n",
    "    long=ID_idf.iloc[i].split(\"|\")[1]\n",
    "    if long in dataset[:, 0]:\n",
    "        ID_ilist.append(long.split(\" \")[0])\n",
    "ID_iarray = np.array(ID_ilist)\n",
    "#print (ID_array)\n",
    "print(\"Isoform ID array has been made!\")\n",
    "print (len(ID_iarray))\n",
    "\n",
    "#Makes array of sequences\n",
    "isequences = []\n",
    "indices_of_iso_ids = list(idf[idf[0].str.contains(\">\")].index)\n",
    "for i in range(len(indices_of_iso_ids)-1):\n",
    "    long=ID_idf.iloc[i].split(\"|\")[1]\n",
    "    # concatenate all into one string\n",
    "    if long in dataset[:,0]:\n",
    "        #print (long)\n",
    "        isequences.append(idf[indices_of_iso_ids[i]+1:indices_of_iso_ids[i+1]][0].sum())\n",
    "        #print (idf[indices_of_iso_ids[i]+1:indices_of_iso_ids[i+1]][0].sum())\n",
    "    if i == len(indices_of_iso_ids)-2:\n",
    "        if long in dataset[:, 0]:\n",
    "            isequences.append(idf[indices_of_uniprot_ids[i+1]+1:][0].sum())\n",
    "            #print (idf[indices_of_uniprot_ids[i+1]+1:][0].sum())\n",
    "isequences = np.array(isequences)\n",
    "print(\"Lines have been separated!\")\n",
    "\n",
    "# Creates dictionary that maps Uniprot ID to sequence\n",
    "ipairs = zip(ID_iarray, isequences)\n",
    "protein_dict.update({A:B for A,B in ipairs})\n",
    "print(\"protein_dict has been made!\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "3da0c2a1-d5d6-4ce6-986f-23c111a36af8",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Define kmer function, we'll make no labels set first\n",
    "def make_kmers (IDs, sites, y, protein_dict, k=53):\n",
    "    \"\"\"\n",
    "    Makes kmers from array of sites\n",
    "    Site is given in bio notation (not python) which starts at 1 and is a string with the identity of the amino acid \n",
    "    \"\"\"\n",
    "    kmers=[]\n",
    "    updated_labels=[]\n",
    "    updated_IDs=[]\n",
    "    updated_sites=[]\n",
    "    half=int((k-1)/2)\n",
    "    for i in range(0, len(sites)):\n",
    "        ID=IDs[i]\n",
    "        location=int(sites[i])\n",
    "        try:\n",
    "            seq=protein_dict[ID]\n",
    "            if location <= half: #To deal with sites near the n terminus\n",
    "                if location > len(seq)-half: #To deal with sequences shorter than k\n",
    "                    gap=\"-\"*(half-location+1)\n",
    "                    gap2=\"-\"*int(half-(len(seq)-location))\n",
    "                    kmer=seq[0 : int(location+half)]\n",
    "                    kmer=gap+kmer+gap2\n",
    "                else: \n",
    "                    gap=\"-\"*(half-location+1)\n",
    "                    kmer=seq[0 : int(location+half)]\n",
    "                    kmer=gap+kmer\n",
    "            elif location > len(seq)-half: #To deal with sites near the C terminus\n",
    "                gap=\"-\"*int(half-(len(seq)-location))\n",
    "                kmer=seq[int(location-half-1): len(seq)]\n",
    "                kmer=kmer+gap\n",
    "            else:\n",
    "                kmer=seq[int(location-half-1): int(location+half)]\n",
    "\n",
    "            if location > len(seq): #Will not add to list of kmers\n",
    "                print (\"Site outside of seq bounds, site: \"+str(location)+\", sequnce length: \"+str(len(seq)))\n",
    "                print (ID)\n",
    "            elif seq[int(location-1)] == \"N\" or seq[int(location-1)]==\"@\":\n",
    "                kmers.append(kmer)\n",
    "                updated_labels.append(y[i])\n",
    "                updated_IDs.append(IDs[i])\n",
    "                updated_sites.append(sites[i])\n",
    "            else:\n",
    "                #print(seq[int(location-1)])\n",
    "                print (\"Sequence key error for \"+ID+\" at \" + str(location)) #Will not add it to list of kmers because doesn't match uniprot\n",
    "                #raise Exception(\"Sequence key error for \"+ID+\" at \" + str(location))\n",
    "            if len(kmer) != k:\n",
    "                print (kmer)\n",
    "                print (len(kmer))            \n",
    "        except KeyError:\n",
    "            print(ID+\" not in protein_dict\")    \n",
    "\n",
    "    return kmers, updated_IDs, updated_sites, updated_labels"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "61eee491-ab60-4f63-84ed-3eccb6f85767",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "E7ER45 not in protein_dict\n",
      "Sequence key error for Q8TCB6 at 90\n",
      "Sequence key error for Q9UF12 at 31\n",
      "Sequence key error for P01019 at 304\n",
      "Q6IE36 not in protein_dict\n",
      "Sequence key error for Q8TD46 at 99\n",
      "Sequence key error for O15131 at 404\n",
      "Sequence key error for Q8WUH1 at 47\n",
      "Q9NY84 not in protein_dict\n",
      "Sequence key error for Q9UKU6 at 89\n",
      "Sequence key error for Q02817 at 4438\n",
      "Site outside of seq bounds, site: 3855, sequnce length: 2040\n",
      "P08519\n",
      "-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------\n",
      "1841\n",
      "Sequence key error for P15692 at 101\n",
      "Sequence key error for P01019 at 170\n",
      "Sequence key error for P01019 at 47\n",
      "Sequence key error for Q96Q06 at 409\n",
      "Sequence key error for Q02817 at 5019\n",
      "Sequence key error for Q8NFD5 at 899\n",
      "Sequence key error for Q9HCJ0 at 675\n",
      "J3QSS3 not in protein_dict\n",
      "D3DNC2 not in protein_dict\n",
      "Sequence key error for Q8NFD5 at 2017\n",
      "Sequence key error for Q9P2C4 at 185\n",
      "Sequence key error for Q8TD46 at 195\n",
      "Sequence key error for Q2KHR2 at 1179\n",
      "Sequence key error for Q9UKU6 at 800\n",
      "Sequence key error for Q9P2C4 at 299\n",
      "Sequence key error for Q8N841 at 132\n",
      "Sequence key error for H0YJS3 at 115\n",
      "Sequence key error for O43451 at 1847\n",
      "D3DP68 not in protein_dict\n",
      "Sequence key error for Q96KR4 at 615\n",
      "Sequence key error for Q02817 at 4970\n",
      "Sequence key error for Q02817 at 4752\n",
      "Sequence key error for Q5JV73 at 923\n",
      "Sequence key error for Q9HCJ0 at 770\n",
      "Sequence key error for Q9HBL0 at 132\n",
      "Sequence key error for Q8N841 at 274\n",
      "Sequence key error for Q02817 at 5069\n",
      "Sequence key error for Q9HCJ0 at 759\n",
      "Sequence key error for Q9HCJ0 at 474\n",
      "Sequence key error for Q8TD46 at 37\n",
      "Sequence key error for P18507 at 393\n",
      "Sequence key error for Q9UKU6 at 605\n",
      "Sequence key error for Q02817 at 4888\n",
      "Sequence key error for P01876 at 340\n",
      "Sequence key error for Q9UKU6 at 906\n",
      "Sequence key error for Q96JG9 at 3396\n",
      "Site outside of seq bounds, site: 197, sequnce length: 172\n",
      "Q96DA0\n",
      "Sequence key error for Q9UKU6 at 684\n",
      "Sequence key error for Q02817 at 4787\n",
      "Sequence key error for O94964 at 1102\n",
      "Sequence key error for Q9NZU1 at 277\n",
      "D3DR43 not in protein_dict\n",
      "Sequence key error for Q96T25 at 52\n",
      "Sequence key error for Q62918 at 635\n",
      "Sequence key error for Q03518 at 362\n",
      "Sequence key error for P08686 at 346\n",
      "Sequence key error for Q5FW52 at 213\n",
      "Sequence key error for Q8WUG5 at 32\n",
      "Sequence key error for Q03518 at 339\n",
      "Sequence key error for Q62918 at 53\n",
      "Sequence key error for Q4Z4D2 at 352\n",
      "Sequence key error for Q03518 at 287\n",
      "Sequence key error for Q92538 at 813\n",
      "Sequence key error for Q4Z4D2 at 184\n",
      "Sequence key error for Q92538 at 1603\n",
      "Sequence key error for Q4Z4D2 at 365\n",
      "Sequence key error for Q9HBL0 at 328\n",
      "Sequence key error for Q4Z4D2 at 23\n",
      "Sequence key error for Q4Z4D2 at 385\n",
      "Sequence key error for Q9HBL0 at 383\n",
      "Sequence key error for Q4Z4D2 at 28\n",
      "Sequence key error for Q6RKD8 at 614\n",
      "Sequence key error for Q62918 at 517\n",
      "Sequence key error for Q9HBL0 at 1063\n",
      "Sequence key error for Q62918 at 298\n",
      "Sequence key error for Q03518 at 310\n",
      "Sequence key error for Q4Z4D2 at 92\n",
      "Sequence key error for Q4Z4D2 at 305\n",
      "Sequence key error for Q13636 at 33\n",
      "Sequence key error for Q8N8F6 at 39\n",
      "Sequence key error for Q8N8F6 at 50\n",
      "Sequence key error for Q8N8F6 at 146\n",
      "Sequence key error for Q6SJ96 at 73\n",
      "Site outside of seq bounds, site: 1553, sequnce length: 1550\n",
      "Q5T890\n",
      "Sequence key error for Q2KHR2 at 1058\n",
      "Sequence key error for P59826 at 142\n",
      "Sequence key error for Q9H195 at 955\n",
      "Sequence key error for Q9HBL0 at 905\n",
      "Sequence key error for Q62918 at 293\n",
      "Split: 24176\n",
      "Test: 2476\n"
     ]
    }
   ],
   "source": [
    "test_kmers, _IDs, _sites, y_test=make_kmers (Test_ID, Test_Position, Test_Label, protein_dict)\n",
    "train_kmers, _IDs, _sites, y_train=make_kmers (Train_ID, Train_Position, Train_Label, protein_dict)\n",
    "\n",
    "#Where do the train and test sets split\n",
    "print (\"Split: \" +str(len(y_train)))\n",
    "print (\"Test: \" +str(len(y_test)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "7b995d35-8fa5-4b5b-80e2-42fae45eb677",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "int32\n"
     ]
    }
   ],
   "source": [
    "#Now let's make the labeled datasets and tensors\n",
    "def labeling (dataset, protein_dict, split, k=53):\n",
    "    \"\"\"\n",
    "    split where the train and test sets split\n",
    "    \"\"\"\n",
    "    kmers, IDs, sites, labels = make_kmers(dataset[:,0], dataset[:,1], dataset[:,2], protein_dict, k=k)\n",
    "    lprotein_dict=copy.deepcopy(protein_dict)\n",
    "    for i in range (0, len(IDs)):\n",
    "        ID=IDs[i]\n",
    "        location=int(sites[i])-1 #To turn to python index\n",
    "        seq=protein_dict[ID]\n",
    "        if location > len(seq):\n",
    "            print (len(seq))\n",
    "            print (seq)\n",
    "            print (ID)\n",
    "            print (location)\n",
    "        elif seq[location]==\"N\":\n",
    "            newseq= seq[:location] + \"@\" + seq[location + 1:]\n",
    "        elif seq[location]==\"@\": \n",
    "            print (\"Already labeled\")\n",
    "        else:\n",
    "            print (seq[location])\n",
    "            print (seq)\n",
    "            print (ID)\n",
    "            print (location)\n",
    "            raise Exception(\"Index error, no N\")\n",
    "        lprotein_dict[ID]=newseq\n",
    "    \n",
    "    kmers0, IDs0, sites0, labels0 = make_kmers(IDs, sites, labels, lprotein_dict, k=k)\n",
    "    #Now we need to revert central @ to N\n",
    "    lkmers=[]\n",
    "    half=int((k-1)/2)\n",
    "    #print (lpos_kmers0)\n",
    "    for kmer in kmers0:\n",
    "        #print (kmer)\n",
    "        if kmer[half]==\"@\":\n",
    "            rkmer=kmer[:half] + \"N\" + kmer[half+1:]\n",
    "        elif kmer[half]==\"N\": #For negatives\n",
    "            rkmer=kmer\n",
    "        else:\n",
    "            raise Exception(\"Incorrect value at kmer center, here is the kmer:\"+kmer)\n",
    "        lkmers.append(rkmer)\n",
    "    lkmers=np.array(lkmers)\n",
    "    l_dataset=np.stack((lkmers, labels), axis=1)\n",
    "    full_trainl, full_testl=np.split(l_dataset, [split])\n",
    "    #Don't need to remove central @ and & for negatives becuase they shouldn't be there!\n",
    "    return full_trainl, full_testl\n",
    "\n",
    "\n",
    "#print (dataset[:,1])\n",
    "#full_trainl, full_testl=labeling(dataset, protein_dict, 24176)\n",
    "\n",
    "\n",
    "\n",
    "alphabet_with_labels = \"ARNDCEQGHILKMFPSTWYV@-UX\" #\n",
    "alphabet_without_labels = \"ARNDCEQGHILKMFPSTWYV-UX\"\n",
    "\n",
    "def convert_to_onehot(data, alphabet):\n",
    "    #Creates a dict, that maps to every char of alphabet an unique int based on position\n",
    "    global char_to_int\n",
    "    char_to_int = dict((c,i) for i,c in enumerate(alphabet))\n",
    "    encoded_data = []\n",
    "    #Replaces every char in data with the mapped int\n",
    "    encoded_data.extend([char_to_int[char] for char in data])\n",
    "    return encoded_data\n",
    "\n",
    "def tensor_encoding(x_data, depth, type, alphabet):\n",
    "    indices = []\n",
    "    t2 = []\n",
    "    for i in tqdm(range(len(x_data))):\n",
    "        indices.append(convert_to_onehot(x_data[i], alphabet))\n",
    "        #print (convert_to_onehot(x_data[i], alphabet))\n",
    "        if len(convert_to_onehot(x_data[i], alphabet)) != 53:\n",
    "            print (x_data[i])\n",
    "            print (\"Length off\")\n",
    "    array=np.stack(indices, axis=0)\n",
    "    if type == 'emb':\n",
    "        return array\n",
    "    for i in tqdm(range(len(indices))):\n",
    "        t1 = tf.one_hot(indices[i], depth) # output: [9 x 23]\n",
    "        t2.append(t1)\n",
    "    return t2\n",
    "\n",
    "def y_to_2d (y_data):\n",
    "    b = np.zeros((y_data.size, 2))\n",
    "    b[np.arange(y_data.size), y_data] = 1\n",
    "    b=b.astype(np.int32)\n",
    "    return b\n",
    "\n",
    "a=np.array([0, 1, 0, 1])\n",
    "b=y_to_2d(a)\n",
    "print (b.dtype)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "3aee179c-92f4-4dd4-be9f-49f568fc7254",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Now let's throw it all together\n",
    "\n",
    "def make_datasets_train_models (dataset, protein_dict, split, k=53, folder=\"_model_results\", trial=\"Trial1\"):\n",
    "    \"\"\"\n",
    "    Trial1: keeps train and test sets as published\n",
    "    \"\"\"\n",
    "    kmers, IDs, sites, labels = make_kmers(dataset[:,0], dataset[:,1], dataset[:,2], protein_dict, k=k)\n",
    "    full_trainl, full_testl=labeling(dataset, protein_dict, split)\n",
    "    print (full_trainl)\n",
    "    \n",
    "    #Now let's make a full dataset and split positives and negatives\n",
    "    full_dataset=np.stack((kmers, labels), axis=1)\n",
    "    print (np.shape(full_dataset))\n",
    "    full_train, full_test = np.split(full_dataset, [split])\n",
    "    full_datasetl=np.concatenate((full_trainl, full_testl))\n",
    "    \n",
    "    if 1==1: #len(pos_kmers) != len(neg_kmers) != len(lpos_kmers) != len(lneg_kmers):\n",
    "        print (np.shape(full_dataset))\n",
    "        print (np.shape(full_datasetl))\n",
    "        print (np.shape(full_train))\n",
    "        print (np.shape(full_test))\n",
    "        print (np.shape(full_trainl))\n",
    "        print (np.shape(full_testl))\n",
    "        #raise Exception (\"Check kmer construction, not all are set equal\")\n",
    "    \n",
    "    if trial != \"Trial1\":\n",
    "        y=full_dataset[:, 1].astype(int)\n",
    "        yl=full_datasetl[:, 1].astype(int)\n",
    "        if y.all() != yl.all():\n",
    "            raise Exception (\"labeled and unlabed datasets don't match\")\n",
    "        pos_kmers=full_dataset[y == 1, 0]\n",
    "        neg_kmers=full_dataset[y == 0, 0]\n",
    "        lpos_kmers=full_datasetl[y == 1, 0]\n",
    "        lneg_kmers=full_datasetl[y == 0, 0]\n",
    "        pos_y=y[y == 1]\n",
    "        neg_y=y[y == 0]\n",
    "    \n",
    "        pos_k = np.stack((pos_kmers, pos_y), axis=1)\n",
    "        neg_k = np.stack((neg_kmers, neg_y), axis=1)\n",
    "        lpos_k = np.stack((lpos_kmers, pos_y), axis=1)\n",
    "        lneg_k = np.stack((lneg_kmers, neg_y), axis=1)\n",
    "\n",
    "        full= np.vstack([pos_k,neg_k])\n",
    "        lfull = np.vstack([lpos_k,lneg_k])\n",
    "        np.random.shuffle(full)\n",
    "        np.random.shuffle(lfull)\n",
    "\n",
    "        print (full[:, 1].dtype)\n",
    "        print (np.shape(full))\n",
    "        print (full)\n",
    "\n",
    "        x_data, y_data = np.split(full, 2, axis=1)\n",
    "        x_datal, y_datal = np.split(lfull, 2, axis=1)\n",
    "\n",
    "        #print (np.shape(x_datal))\n",
    "        y_data=y_data.astype(np.int32)\n",
    "        y_datal=y_datal.astype(np.int32)\n",
    "        #print (y_data.dtype)\n",
    "        #print (y_data)\n",
    "\n",
    "        # Split into train, validation, and test datasets\n",
    "        N = len(full)\n",
    "        x_train = x_data[0:int(0.8*N), 0]\n",
    "        #print (np.shape(x_train))\n",
    "        y_train = y_data[0:int(0.8*N)]\n",
    "        #print (y_train)\n",
    "\n",
    "        x_validation = x_data[int(0.8*N):int(0.8*N)+int(0.1*N), 0]\n",
    "        y_validation = y_data[int(0.8*N):int(0.8*N)+int(0.1*N), 0]\n",
    "\n",
    "        x_test = x_data[int(0.8*N)+int(0.1*N):, 0]\n",
    "        y_test = y_data[int(0.8*N)+int(0.1*N):, 0]\n",
    "\n",
    "        x_trainl = x_datal[0:int(0.8*N), 0]\n",
    "        y_trainl = y_datal[0:int(0.8*N), 0]\n",
    "\n",
    "        x_validationl = x_datal[int(0.8*N):int(0.8*N)+int(0.1*N), 0]\n",
    "        y_validationl = y_datal[int(0.8*N):int(0.8*N)+int(0.1*N), 0]\n",
    "\n",
    "        x_testl = x_datal[int(0.8*N)+int(0.1*N):, 0]\n",
    "        y_testl = y_datal[int(0.8*N)+int(0.1*N):, 0]\n",
    "\n",
    "    else:\n",
    "        N = len(full_train)\n",
    "        np.random.shuffle(full_train)\n",
    "        np.random.shuffle(full_test)\n",
    "        np.random.shuffle(full_trainl)\n",
    "        np.random.shuffle(full_testl)\n",
    "\n",
    "        x_train = full_train[0:int(0.89*N), 0]\n",
    "        y_train = full_train[0:int(0.89*N), 1].astype(np.int32)\n",
    "\n",
    "        x_validation = full_train[int(0.89*N):, 0]\n",
    "        y_validation = full_train[int(0.89*N):, 1].astype(np.int32)\n",
    "        \n",
    "        x_test = full_test[:, 0]\n",
    "        y_test = full_test[:, 1].astype(np.int32)\n",
    "\n",
    "        x_trainl = full_trainl[0:int(0.89*N), 0]\n",
    "        y_trainl = full_trainl[0:int(0.89*N), 1].astype(np.int32)\n",
    "\n",
    "        x_validationl = full_trainl[int(0.89*N):, 0]\n",
    "        y_validationl = full_trainl[int(0.89*N):, 1].astype(np.int32)\n",
    "\n",
    "        x_testl = full_testl[:, 0]\n",
    "        y_testl = full_testl[:, 1].astype(np.int32)\n",
    "\n",
    "    #Make tensors of x data\n",
    "    tensor1 = tensor_encoding(x_train, 23, 'emb', alphabet_without_labels)\n",
    "    tensor2 = tensor_encoding(x_validation, 23, 'emb', alphabet_without_labels)\n",
    "    tensor3 = tensor_encoding(x_test, 23, 'emb', alphabet_without_labels)\n",
    "    tensor4 = tensor_encoding(x_trainl, 23, 'emb', alphabet_with_labels)\n",
    "    tensor5 = tensor_encoding(x_validationl, 23, 'emb', alphabet_with_labels)\n",
    "    tensor6 = tensor_encoding(x_testl, 23, 'emb', alphabet_with_labels)\n",
    "\n",
    "    print (type(tensor1))\n",
    "    print (tensor1.dtype)\n",
    "    print (y_train.dtype)\n",
    "    #print (np.shape(tensor2))\n",
    "    #print (np.shape(tensor3))\n",
    "    #print (np.shape(tensor4))\n",
    "    #print (np.shape(tensor5))\n",
    "    #print (np.shape(tensor6))\n",
    "\n",
    "    directory = trial\n",
    "    parent_dir = folder+'/'\n",
    "    path = os.path.join(parent_dir, directory) \n",
    "    try:\n",
    "        os.mkdir(path) \n",
    "    except:\n",
    "        pass\n",
    "    \n",
    "    #Save datasets\n",
    "    np.save(f'{path}/{trial}_x_train_without_labels', tensor1)\n",
    "    np.save(f'{path}/{trial}_y_train_without_labels', y_train)\n",
    "    np.save(f'{path}/{trial}_x_validation_without_labels', tensor2)\n",
    "    np.save(f'{path}/{trial}_y_validation_without_labels', y_validation)\n",
    "    np.save(f'{path}/{trial}_x_test_without_labels', tensor3)\n",
    "    np.save(f'{path}/{trial}_y_test_without_labels', y_test)\n",
    "    np.save(f'{path}/{trial}_x_train_with_labels', tensor4)\n",
    "    np.save(f'{path}/{trial}_y_train_with_labels', y_trainl)\n",
    "    np.save(f'{path}/{trial}_x_validation_with_labels', tensor5)\n",
    "    np.save(f'{path}/{trial}_y_validation_with_labels', y_validationl)\n",
    "    np.save(f'{path}/{trial}_x_test_with_labels', tensor6)\n",
    "    np.save(f'{path}/{trial}_y_test_with_labels', y_testl)\n",
    "\n",
    "    #Add the model\n",
    "    callback = tf.keras.callbacks.EarlyStopping(monitor='val_loss', mode='min', verbose=1, patience=10, restore_best_weights=True)\n",
    "\n",
    "    model = Sequential()\n",
    "    model.add(Embedding(input_dim=256, output_dim=21, embeddings_initializer={'class_name': 'RandomUniform',\n",
    "         'config': {'minval': -0.05, 'maxval': 0.05, 'seed': None}}, input_length=53, name=\"emb_layer\"))\n",
    "    model.add(Reshape((53, 21, 1), input_shape=(53, 21)))\n",
    "    model.add(Conv2D(16, (3, 3), padding='valid', input_shape=(53, 21, 1), activation='relu', kernel_regularizer=l2(1e-06)))\n",
    "    model.add(MaxPooling2D(pool_size=(2,2), padding='valid'))\n",
    "    model.add(Flatten())\n",
    "    model.add(Dense(64, activation=\"relu\", kernel_regularizer=l1(1e-06)))\n",
    "    model.add(Dropout(0.1))\n",
    "    model.add(Dense(1, activation='sigmoid'))\n",
    "    model.summary()\n",
    "\n",
    "    model.compile(optimizer=Adam(1e-03), \n",
    "                  loss='BinaryCrossentropy', \n",
    "                  metrics=['accuracy'])\n",
    "    history = model.fit(x=tensor1,y=y_train,\n",
    "              batch_size=256, epochs=400,\n",
    "              validation_data=(tensor2, y_validation), \n",
    "              callbacks=[callback], verbose=1)\n",
    "    model.save(f'{path}/{directory}_emb_CNN_no_labels_{trial}.h5')\n",
    "    model.evaluate(tensor3, y_test)\n",
    "\n",
    "     # The predictions are between 0 and 1 because of the sigmoid activation\n",
    "    # function of the last layer in the NN\n",
    "    pred = model.predict(tensor3).flatten()\n",
    "\n",
    "    # We can also calculate the AUC to get an estimate how good the NN actually learned\n",
    "    print (\"Without labels\")\n",
    "    fpr, tpr, thresholds = roc_curve(y_test, pred)\n",
    "    print(\"AUC:\", round(auc(fpr, tpr), 4))\n",
    "    print(\"AUPRC:\", round(average_precision_score(y_test, pred), 4))\n",
    "    pred2 = (pred > 0.5)\n",
    "    pred2 = np.array(pred2)\n",
    "    pred2 = pred2.astype(int)\n",
    "    print(\"MCC:\", round(matthews_corrcoef(y_test, pred2), 4))\n",
    "    print(\"F1:\", round(f1_score(y_test, pred2), 4))\n",
    "    print(\"Sensitivity:\", round(np.average(tpr), 4))\n",
    "    print(\"Specificity:\", round(np.average(1-fpr), 4))\n",
    "    print(\"Precision:\", round(precision_score(y_test, pred2), 4))\n",
    "\n",
    "    #Now with labels\n",
    "    historyl = model.fit(x=tensor4,y=y_trainl,\n",
    "              batch_size=256, epochs=400,\n",
    "              validation_data=(tensor5, y_validationl), \n",
    "              callbacks=[callback], verbose=1)\n",
    "    model.save(f'{path}/{directory}_emb_CNN_with_labels_{trial}.h5')\n",
    "    model.evaluate(tensor6, y_testl)\n",
    "\n",
    "     # The predictions are between 0 and 1 because of the sigmoid activation\n",
    "    # function of the last layer in the NN\n",
    "    predl = model.predict(tensor6).flatten()\n",
    "\n",
    "    # We can also calculate the AUC to get an estimate how good the NN actually learned\n",
    "    print (\"With labels\")\n",
    "    fprl, tprl, thresholdsl = roc_curve(y_testl, predl)\n",
    "    print(\"AUC:\", round(auc(fprl, tprl), 4))\n",
    "    print(\"AUPRC:\", round(average_precision_score(y_testl, predl), 4))\n",
    "    pred2l = (predl > 0.5)\n",
    "    pred2l = np.array(pred2l)\n",
    "    pred2l = pred2l.astype(int)\n",
    "    print(\"MCC:\", round(matthews_corrcoef(y_testl, pred2l), 4))\n",
    "    print(\"F1:\", round(f1_score(y_testl, pred2l), 4))\n",
    "    print(\"Sensitivity:\", round(np.average(tprl), 4))\n",
    "    print(\"Specificity:\", round(np.average(1-fprl), 4))\n",
    "    print(\"Precision:\", round(precision_score(y_testl, pred2l), 4))\n",
    "\n",
    "#make_datasets_train_models(\"ABL1.npy\", protein_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "99389782-1752-4e68-ab20-b02dafbe9dd9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Sequence key error for Q9UF12 at 31\n",
      "Sequence key error for P01019 at 304\n",
      "Q6IE36 not in protein_dict\n",
      "Sequence key error for Q8TD46 at 99\n",
      "Sequence key error for O15131 at 404\n",
      "Sequence key error for Q8WUH1 at 47\n",
      "Q9NY84 not in protein_dict\n",
      "Sequence key error for Q9UKU6 at 89\n",
      "Sequence key error for Q02817 at 4438\n",
      "Site outside of seq bounds, site: 3855, sequnce length: 2040\n",
      "P08519\n",
      "-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------\n",
      "1841\n",
      "Sequence key error for P15692 at 101\n",
      "Sequence key error for P01019 at 170\n",
      "Sequence key error for P01019 at 47\n",
      "Sequence key error for Q96Q06 at 409\n",
      "Sequence key error for Q02817 at 5019\n",
      "Sequence key error for Q8NFD5 at 899\n",
      "Sequence key error for Q9HCJ0 at 675\n",
      "J3QSS3 not in protein_dict\n",
      "D3DNC2 not in protein_dict\n",
      "Sequence key error for Q8NFD5 at 2017\n",
      "Sequence key error for Q9P2C4 at 185\n",
      "Sequence key error for Q8TD46 at 195\n",
      "Sequence key error for Q2KHR2 at 1179\n",
      "Sequence key error for Q9UKU6 at 800\n",
      "Sequence key error for Q9P2C4 at 299\n",
      "Sequence key error for Q8N841 at 132\n",
      "Sequence key error for H0YJS3 at 115\n",
      "Sequence key error for O43451 at 1847\n",
      "D3DP68 not in protein_dict\n",
      "Sequence key error for Q96KR4 at 615\n",
      "Sequence key error for Q02817 at 4970\n",
      "Sequence key error for Q02817 at 4752\n",
      "Sequence key error for Q5JV73 at 923\n",
      "Sequence key error for Q9HCJ0 at 770\n",
      "Sequence key error for Q9HBL0 at 132\n",
      "Sequence key error for Q8N841 at 274\n",
      "Sequence key error for Q02817 at 5069\n",
      "Sequence key error for Q9HCJ0 at 759\n",
      "Sequence key error for Q9HCJ0 at 474\n",
      "Sequence key error for Q8TD46 at 37\n",
      "Sequence key error for P18507 at 393\n",
      "Sequence key error for Q9UKU6 at 605\n",
      "Sequence key error for Q02817 at 4888\n",
      "Sequence key error for P01876 at 340\n",
      "Sequence key error for Q9UKU6 at 906\n",
      "Sequence key error for Q96JG9 at 3396\n",
      "Site outside of seq bounds, site: 197, sequnce length: 172\n",
      "Q96DA0\n",
      "Sequence key error for Q9UKU6 at 684\n",
      "Sequence key error for Q02817 at 4787\n",
      "Sequence key error for O94964 at 1102\n",
      "Sequence key error for Q9NZU1 at 277\n",
      "D3DR43 not in protein_dict\n",
      "Sequence key error for Q96T25 at 52\n",
      "Sequence key error for Q62918 at 635\n",
      "Sequence key error for Q03518 at 362\n",
      "Sequence key error for P08686 at 346\n",
      "Sequence key error for Q5FW52 at 213\n",
      "Sequence key error for Q8WUG5 at 32\n",
      "Sequence key error for Q03518 at 339\n",
      "Sequence key error for Q62918 at 53\n",
      "Sequence key error for Q4Z4D2 at 352\n",
      "Sequence key error for Q03518 at 287\n",
      "Sequence key error for Q92538 at 813\n",
      "Sequence key error for Q4Z4D2 at 184\n",
      "Sequence key error for Q92538 at 1603\n",
      "Sequence key error for Q4Z4D2 at 365\n",
      "Sequence key error for Q9HBL0 at 328\n",
      "Sequence key error for Q4Z4D2 at 23\n",
      "Sequence key error for Q4Z4D2 at 385\n",
      "Sequence key error for Q9HBL0 at 383\n",
      "Sequence key error for Q4Z4D2 at 28\n",
      "Sequence key error for Q6RKD8 at 614\n",
      "Sequence key error for Q62918 at 517\n",
      "Sequence key error for Q9HBL0 at 1063\n",
      "Sequence key error for Q62918 at 298\n",
      "Sequence key error for Q03518 at 310\n",
      "Sequence key error for Q4Z4D2 at 92\n",
      "Sequence key error for Q4Z4D2 at 305\n",
      "Sequence key error for Q13636 at 33\n",
      "Sequence key error for Q8N8F6 at 39\n",
      "Sequence key error for Q8N8F6 at 50\n",
      "Sequence key error for Q8N8F6 at 146\n",
      "Sequence key error for Q6SJ96 at 73\n",
      "Site outside of seq bounds, site: 1553, sequnce length: 1550\n",
      "Q5T890\n",
      "Sequence key error for Q2KHR2 at 1058\n",
      "Sequence key error for P59826 at 142\n",
      "Sequence key error for Q9H195 at 955\n",
      "Sequence key error for Q9HBL0 at 905\n",
      "Sequence key error for Q62918 at 293\n",
      "E7ER45 not in protein_dict\n",
      "Sequence key error for Q8TCB6 at 90\n",
      "Sequence key error for Q9UF12 at 31\n",
      "Sequence key error for P01019 at 304\n",
      "Q6IE36 not in protein_dict\n",
      "Sequence key error for Q8TD46 at 99\n",
      "Sequence key error for O15131 at 404\n",
      "Sequence key error for Q8WUH1 at 47\n",
      "Q9NY84 not in protein_dict\n",
      "Sequence key error for Q9UKU6 at 89\n",
      "Sequence key error for Q02817 at 4438\n",
      "Site outside of seq bounds, site: 3855, sequnce length: 2040\n",
      "P08519\n",
      "-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------\n",
      "1841\n",
      "Sequence key error for P15692 at 101\n",
      "Sequence key error for P01019 at 170\n",
      "Sequence key error for P01019 at 47\n",
      "Sequence key error for Q96Q06 at 409\n",
      "Sequence key error for Q02817 at 5019\n",
      "Sequence key error for Q8NFD5 at 899\n",
      "Sequence key error for Q9HCJ0 at 675\n",
      "J3QSS3 not in protein_dict\n",
      "D3DNC2 not in protein_dict\n",
      "Sequence key error for Q8NFD5 at 2017\n",
      "Sequence key error for Q9P2C4 at 185\n",
      "Sequence key error for Q8TD46 at 195\n",
      "Sequence key error for Q2KHR2 at 1179\n",
      "Sequence key error for Q9UKU6 at 800\n",
      "Sequence key error for Q9P2C4 at 299\n",
      "Sequence key error for Q8N841 at 132\n",
      "Sequence key error for H0YJS3 at 115\n",
      "Sequence key error for O43451 at 1847\n",
      "D3DP68 not in protein_dict\n",
      "Sequence key error for Q96KR4 at 615\n",
      "Sequence key error for Q02817 at 4970\n",
      "Sequence key error for Q02817 at 4752\n",
      "Sequence key error for Q5JV73 at 923\n",
      "Sequence key error for Q9HCJ0 at 770\n",
      "Sequence key error for Q9HBL0 at 132\n",
      "Sequence key error for Q8N841 at 274\n",
      "Sequence key error for Q02817 at 5069\n",
      "Sequence key error for Q9HCJ0 at 759\n",
      "Sequence key error for Q9HCJ0 at 474\n",
      "Sequence key error for Q8TD46 at 37\n",
      "Sequence key error for P18507 at 393\n",
      "Sequence key error for Q9UKU6 at 605\n",
      "Sequence key error for Q02817 at 4888\n",
      "Sequence key error for P01876 at 340\n",
      "Sequence key error for Q9UKU6 at 906\n",
      "Sequence key error for Q96JG9 at 3396\n",
      "Site outside of seq bounds, site: 197, sequnce length: 172\n",
      "Q96DA0\n",
      "Sequence key error for Q9UKU6 at 684\n",
      "Sequence key error for Q02817 at 4787\n",
      "Sequence key error for O94964 at 1102\n",
      "Sequence key error for Q9NZU1 at 277\n",
      "D3DR43 not in protein_dict\n",
      "Sequence key error for Q96T25 at 52\n",
      "Sequence key error for Q62918 at 635\n",
      "Sequence key error for Q03518 at 362\n",
      "Sequence key error for P08686 at 346\n",
      "Sequence key error for Q5FW52 at 213\n",
      "Sequence key error for Q8WUG5 at 32\n",
      "Sequence key error for Q03518 at 339\n",
      "Sequence key error for Q62918 at 53\n",
      "Sequence key error for Q4Z4D2 at 352\n",
      "Sequence key error for Q03518 at 287\n",
      "Sequence key error for Q92538 at 813\n",
      "Sequence key error for Q4Z4D2 at 184\n",
      "Sequence key error for Q92538 at 1603\n",
      "Sequence key error for Q4Z4D2 at 365\n",
      "Sequence key error for Q9HBL0 at 328\n",
      "Sequence key error for Q4Z4D2 at 23\n",
      "Sequence key error for Q4Z4D2 at 385\n",
      "Sequence key error for Q9HBL0 at 383\n",
      "Sequence key error for Q4Z4D2 at 28\n",
      "Sequence key error for Q6RKD8 at 614\n",
      "Sequence key error for Q62918 at 517\n",
      "Sequence key error for Q9HBL0 at 1063\n",
      "Sequence key error for Q62918 at 298\n",
      "Sequence key error for Q03518 at 310\n",
      "Sequence key error for Q4Z4D2 at 92\n",
      "Sequence key error for Q4Z4D2 at 305\n",
      "Sequence key error for Q13636 at 33\n",
      "Sequence key error for Q8N8F6 at 39\n",
      "Sequence key error for Q8N8F6 at 50\n",
      "Sequence key error for Q8N8F6 at 146\n",
      "Sequence key error for Q6SJ96 at 73\n",
      "Site outside of seq bounds, site: 1553, sequnce length: 1550\n",
      "Q5T890\n",
      "Sequence key error for Q2KHR2 at 1058\n",
      "Sequence key error for P59826 at 142\n",
      "Sequence key error for Q9H195 at 955\n",
      "Sequence key error for Q9HBL0 at 905\n",
      "Sequence key error for Q62918 at 293\n",
      "E7ER45 not in protein_dict\n",
      "Sequence key error for Q8TCB6 at 90\n",
      "[['LISKIFPGLAADQSRALRLGDAILSVNGTDLRQATHDQAVQALKRAGKEVLLE' '1']\n",
      " ['EPSDKMDALMPISGTSFAVGIDFHAENDTIYWTDMGFNKISRAKRDQTWKEDI' '1']\n",
      " ['LYGSAPRTPSKRRGLAPNTPGKARKLNTTTMSNATANSSIRPIFGGTVYHSPV' '1']\n",
      " ...\n",
      " ['PSRVGMKVPTISVTSFHSQSSYISSDNDSELEDADIKKELQSLREKHLKEISE' '0']\n",
      " ['NTQVASGTLGRVPTITEATPIKGVAPNSTLSSRTSPVTTAGSLLEKSSITMTP' '0']\n",
      " ['TVQNCDLEQQSIVHIVQRPRRRSHETNASGGDEPQSTSEGSIWESRSLTRVDL' '0']]\n",
      "(26652, 2)\n",
      "(26652, 2)\n",
      "(26652, 2)\n",
      "(24176, 2)\n",
      "(2476, 2)\n",
      "(24176, 2)\n",
      "(2476, 2)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|█████████████████████████████████████████████████████████████████████████| 21516/21516 [00:00<00:00, 21702.54it/s]\n",
      "100%|███████████████████████████████████████████████████████████████████████████| 2660/2660 [00:00<00:00, 17668.70it/s]\n",
      "100%|███████████████████████████████████████████████████████████████████████████| 2476/2476 [00:00<00:00, 19879.70it/s]\n",
      "100%|█████████████████████████████████████████████████████████████████████████| 21516/21516 [00:01<00:00, 16115.51it/s]\n",
      "100%|███████████████████████████████████████████████████████████████████████████| 2660/2660 [00:00<00:00, 14405.49it/s]\n",
      "100%|███████████████████████████████████████████████████████████████████████████| 2476/2476 [00:00<00:00, 14889.54it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'numpy.ndarray'>\n",
      "int32\n",
      "int32\n",
      "Model: \"sequential_28\"\n",
      "_________________________________________________________________\n",
      " Layer (type)                Output Shape              Param #   \n",
      "=================================================================\n",
      " emb_layer (Embedding)       (None, 53, 21)            5376      \n",
      "                                                                 \n",
      " reshape_28 (Reshape)        (None, 53, 21, 1)         0         \n",
      "                                                                 \n",
      " conv2d_28 (Conv2D)          (None, 51, 19, 16)        160       \n",
      "                                                                 \n",
      " max_pooling2d_28 (MaxPoolin  (None, 25, 9, 16)        0         \n",
      " g2D)                                                            \n",
      "                                                                 \n",
      " flatten_28 (Flatten)        (None, 3600)              0         \n",
      "                                                                 \n",
      " dense_56 (Dense)            (None, 32)                115232    \n",
      "                                                                 \n",
      " dropout_28 (Dropout)        (None, 32)                0         \n",
      "                                                                 \n",
      " dense_57 (Dense)            (None, 1)                 33        \n",
      "                                                                 \n",
      "=================================================================\n",
      "Total params: 120,801\n",
      "Trainable params: 120,801\n",
      "Non-trainable params: 0\n",
      "_________________________________________________________________\n",
      "Epoch 1/400\n",
      "85/85 [==============================] - 6s 46ms/step - loss: 0.6574 - accuracy: 0.6555 - val_loss: 0.6495 - val_accuracy: 0.6477\n",
      "Epoch 2/400\n",
      "85/85 [==============================] - 3s 31ms/step - loss: 0.6355 - accuracy: 0.6556 - val_loss: 0.6277 - val_accuracy: 0.6470\n",
      "Epoch 3/400\n",
      "85/85 [==============================] - 3s 30ms/step - loss: 0.6244 - accuracy: 0.6578 - val_loss: 0.6174 - val_accuracy: 0.6508\n",
      "Epoch 4/400\n",
      "85/85 [==============================] - 2s 29ms/step - loss: 0.6179 - accuracy: 0.6621 - val_loss: 0.6187 - val_accuracy: 0.6515\n",
      "Epoch 5/400\n",
      "85/85 [==============================] - 3s 30ms/step - loss: 0.6134 - accuracy: 0.6615 - val_loss: 0.6098 - val_accuracy: 0.6568\n",
      "Epoch 6/400\n",
      "85/85 [==============================] - 2s 29ms/step - loss: 0.6113 - accuracy: 0.6649 - val_loss: 0.6078 - val_accuracy: 0.6639\n",
      "Epoch 7/400\n",
      "85/85 [==============================] - 2s 29ms/step - loss: 0.6064 - accuracy: 0.6687 - val_loss: 0.6101 - val_accuracy: 0.6733\n",
      "Epoch 8/400\n",
      "85/85 [==============================] - 3s 32ms/step - loss: 0.6057 - accuracy: 0.6699 - val_loss: 0.6153 - val_accuracy: 0.6782\n",
      "Epoch 9/400\n",
      "85/85 [==============================] - 2s 28ms/step - loss: 0.6081 - accuracy: 0.6719 - val_loss: 0.6113 - val_accuracy: 0.6602\n",
      "Epoch 10/400\n",
      "85/85 [==============================] - 2s 29ms/step - loss: 0.6026 - accuracy: 0.6741 - val_loss: 0.6070 - val_accuracy: 0.6771\n",
      "Epoch 11/400\n",
      "85/85 [==============================] - 2s 28ms/step - loss: 0.5999 - accuracy: 0.6786 - val_loss: 0.6112 - val_accuracy: 0.6801\n",
      "Epoch 12/400\n",
      "85/85 [==============================] - 3s 32ms/step - loss: 0.5991 - accuracy: 0.6792 - val_loss: 0.6065 - val_accuracy: 0.6737\n",
      "Epoch 13/400\n",
      "85/85 [==============================] - 2s 29ms/step - loss: 0.5928 - accuracy: 0.6830 - val_loss: 0.6076 - val_accuracy: 0.6767\n",
      "Epoch 14/400\n",
      "85/85 [==============================] - 2s 29ms/step - loss: 0.5888 - accuracy: 0.6887 - val_loss: 0.6087 - val_accuracy: 0.6722\n",
      "Epoch 15/400\n",
      "85/85 [==============================] - 3s 38ms/step - loss: 0.5861 - accuracy: 0.6896 - val_loss: 0.6135 - val_accuracy: 0.6774\n",
      "Epoch 16/400\n",
      "85/85 [==============================] - 3s 36ms/step - loss: 0.5827 - accuracy: 0.6933 - val_loss: 0.6098 - val_accuracy: 0.6771\n",
      "Epoch 17/400\n",
      "85/85 [==============================] - 3s 31ms/step - loss: 0.5763 - accuracy: 0.6968 - val_loss: 0.6134 - val_accuracy: 0.6737\n",
      "Epoch 18/400\n",
      "85/85 [==============================] - 3s 30ms/step - loss: 0.5751 - accuracy: 0.7006 - val_loss: 0.6099 - val_accuracy: 0.6782\n",
      "Epoch 19/400\n",
      "85/85 [==============================] - 3s 32ms/step - loss: 0.5713 - accuracy: 0.7055 - val_loss: 0.6181 - val_accuracy: 0.6598\n",
      "Epoch 20/400\n",
      "85/85 [==============================] - 3s 33ms/step - loss: 0.5687 - accuracy: 0.7118 - val_loss: 0.6193 - val_accuracy: 0.6481\n",
      "Epoch 21/400\n",
      "85/85 [==============================] - 3s 31ms/step - loss: 0.5695 - accuracy: 0.7078 - val_loss: 0.6109 - val_accuracy: 0.6744\n",
      "Epoch 22/400\n",
      "83/85 [============================>.] - ETA: 0s - loss: 0.5593 - accuracy: 0.7132Restoring model weights from the end of the best epoch: 12.\n",
      "85/85 [==============================] - 3s 32ms/step - loss: 0.5592 - accuracy: 0.7138 - val_loss: 0.6151 - val_accuracy: 0.6564\n",
      "Epoch 22: early stopping\n",
      "78/78 [==============================] - 0s 4ms/step - loss: 0.6125 - accuracy: 0.6729\n",
      "78/78 [==============================] - 1s 4ms/step\n",
      "Without labels\n",
      "AUC: 0.6581\n",
      "AUPRC: 0.4667\n",
      "MCC: 0.1411\n",
      "F1: 0.2387\n",
      "Sensitivity: 0.5774\n",
      "Specificity: 0.5909\n",
      "Precision: 0.5404\n",
      "Epoch 1/400\n",
      "85/85 [==============================] - 3s 32ms/step - loss: 0.5953 - accuracy: 0.6838 - val_loss: 0.5960 - val_accuracy: 0.6707\n",
      "Epoch 2/400\n",
      "85/85 [==============================] - 3s 30ms/step - loss: 0.5930 - accuracy: 0.6858 - val_loss: 0.5904 - val_accuracy: 0.6846\n",
      "Epoch 3/400\n",
      "85/85 [==============================] - 3s 32ms/step - loss: 0.5879 - accuracy: 0.6897 - val_loss: 0.5904 - val_accuracy: 0.6929\n",
      "Epoch 4/400\n",
      "85/85 [==============================] - 3s 31ms/step - loss: 0.5841 - accuracy: 0.6939 - val_loss: 0.5917 - val_accuracy: 0.6936\n",
      "Epoch 5/400\n",
      "85/85 [==============================] - 3s 32ms/step - loss: 0.5816 - accuracy: 0.6954 - val_loss: 0.5935 - val_accuracy: 0.6891\n",
      "Epoch 6/400\n",
      "85/85 [==============================] - 3s 33ms/step - loss: 0.5806 - accuracy: 0.6987 - val_loss: 0.5916 - val_accuracy: 0.6910\n",
      "Epoch 7/400\n",
      "85/85 [==============================] - 3s 33ms/step - loss: 0.5752 - accuracy: 0.7023 - val_loss: 0.5925 - val_accuracy: 0.6970\n",
      "Epoch 8/400\n",
      "85/85 [==============================] - 3s 33ms/step - loss: 0.5715 - accuracy: 0.7034 - val_loss: 0.6114 - val_accuracy: 0.6782\n",
      "Epoch 9/400\n",
      "85/85 [==============================] - 3s 34ms/step - loss: 0.5682 - accuracy: 0.7094 - val_loss: 0.5959 - val_accuracy: 0.6895\n",
      "Epoch 10/400\n",
      "85/85 [==============================] - 3s 30ms/step - loss: 0.5635 - accuracy: 0.7137 - val_loss: 0.5966 - val_accuracy: 0.6895\n",
      "Epoch 11/400\n",
      "85/85 [==============================] - 3s 32ms/step - loss: 0.5597 - accuracy: 0.7173 - val_loss: 0.6059 - val_accuracy: 0.6722\n",
      "Epoch 12/400\n",
      "85/85 [==============================] - 3s 32ms/step - loss: 0.5570 - accuracy: 0.7190 - val_loss: 0.6035 - val_accuracy: 0.6741\n",
      "Epoch 13/400\n",
      "84/85 [============================>.] - ETA: 0s - loss: 0.5543 - accuracy: 0.7217Restoring model weights from the end of the best epoch: 3.\n",
      "85/85 [==============================] - 3s 33ms/step - loss: 0.5543 - accuracy: 0.7217 - val_loss: 0.6225 - val_accuracy: 0.6774\n",
      "Epoch 13: early stopping\n",
      "78/78 [==============================] - 1s 6ms/step - loss: 0.6151 - accuracy: 0.6725\n",
      "78/78 [==============================] - 0s 5ms/step\n",
      "With labels\n",
      "AUC: 0.6575\n",
      "AUPRC: 0.4697\n",
      "MCC: 0.1675\n",
      "F1: 0.3213\n",
      "Sensitivity: 0.5788\n",
      "Specificity: 0.5906\n",
      "Precision: 0.5246\n"
     ]
    }
   ],
   "source": [
    "make_datasets_train_models (dataset, protein_dict, 24176, trial=\"Trial1\")\n",
    "#make_datasets_train_models (dataset, protein_dict, 24176, trial=\"Trial2\")\n",
    "#make_datasets_train_models (dataset, protein_dict, 24176, trial=\"Trial3\")\n",
    "#make_datasets_train_models (dataset, protein_dict, 24176, trial=\"Trial4\")\n",
    "#make_datasets_train_models (dataset, protein_dict, 24176, trial=\"Trial5\")\n",
    "#make_datasets_train_models (dataset, protein_dict, 24176, trial=\"Trial6\")\n",
    "#make_datasets_train_models (dataset, protein_dict, 24176, trial=\"Trial7\")\n",
    "#make_datasets_train_models (dataset, protein_dict, 24176, trial=\"Trial8\")\n",
    "#make_datasets_train_models (dataset, protein_dict, 24176, trial=\"Trial9\")\n",
    "#make_datasets_train_models (dataset, protein_dict, 24176, trial=\"Trial10\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "55d86bc4-e381-4cfe-bf36-e1934842cfe3",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\Clair\\AppData\\Local\\Temp\\ipykernel_12512\\30486781.py:6: ParserWarning: Falling back to the 'python' engine because the 'c' engine does not support regex separators (separators > 1 char and different from '\\s+' are interpreted as regex); you can avoid this warning by specifying engine='python'.\n",
      "  df2 = pd.read_csv(file_name2, sep='delimiter', header=None)\n"
     ]
    }
   ],
   "source": [
    "#Now let's do this for the N-GlyDE dataset (human only)\n",
    "file_name2 = 'UP000005640_9606.txt' #Reference proteome for humans\n",
    "excel3 = 'Glycobiology_NGlyDE_Training_Positive_1821_Negative_901.csv'\n",
    "excel4 = 'Glycobiology_NGlyDE_Independent_Positive_202_Negative_100.csv'\n",
    "\n",
    "df2 = pd.read_csv(file_name2, sep='delimiter', header=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "154d7e4b-c60b-41d9-8575-2b7a3c13c2d6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ID array has been made!\n",
      "Lines have been separated!\n",
      "protein_dict2 has been made!\n"
     ]
    }
   ],
   "source": [
    "#Load all possible Uniprot IDs from reference proteome\n",
    "df2.fillna(\"NA\", inplace=True) # just in case any \"NA\" got turned into NaNs\n",
    "ID_df2 = df2[df2[0].str.contains(\">\")][0]\n",
    "ID_list2 = []\n",
    "for i in range(len(ID_df2)):  #This dataset uses a different Uniprot form than previous so select it differently\n",
    "    long=ID_df2.iloc[i].split(\"|\")[1]\n",
    "    ID_list2.append(long.split(\" \")[0])\n",
    "ID_array2 = np.array(ID_list2)\n",
    "#print (ID_array)\n",
    "print(\"ID array has been made!\")\n",
    "\n",
    "#Makes array of sequences\n",
    "sequences2 = []\n",
    "indices_of_uniprot_ids2 = list(df2[df2[0].str.contains(\">\")].index)\n",
    "for i in range(len(indices_of_uniprot_ids2)-1):\n",
    "    # concatenate all into one string\n",
    "    sequences2.append(df2[indices_of_uniprot_ids2[i]+1:indices_of_uniprot_ids2[i+1]][0].sum())\n",
    "    if i == len(indices_of_uniprot_ids2)-2:\n",
    "        sequences2.append(df2[indices_of_uniprot_ids2[i+1]+1:][0].sum())\n",
    "sequences2 = np.array(sequences2)\n",
    "print(\"Lines have been separated!\")\n",
    "\n",
    "# Creates dictionary that maps Uniprot ID to sequence\n",
    "pairs2 = zip(ID_array2, sequences2)\n",
    "protein_dict2 = {A:B for A,B in pairs2}\n",
    "print(\"protein_dict2 has been made!\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "53e45d47-a86c-44e0-bd34-a51e977480b0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(3024, 3)\n"
     ]
    }
   ],
   "source": [
    "#Now let's do the data from N-GlyDE\n",
    "#Load data and convert to numpy array\n",
    "sheet3 = pd.read_csv(excel3)\n",
    "Data3=sheet3[[\"PID\", \"Position\", \"label\"]]\n",
    "\n",
    "sheet4 = pd.read_csv(excel4)\n",
    "Data4=sheet4[[\"PID\", \"Position\", \"label\"]]\n",
    "\n",
    "Train_ID2=Data3.iloc[:, 0] \n",
    "Train_ID2=Train_ID2.to_numpy(dtype=str)\n",
    "\n",
    "Train_Position2=Data3.iloc[:, 1] \n",
    "Train_Position2=Train_Position2.to_numpy(dtype=str)\n",
    "\n",
    "Train_Label2=Data3.iloc[:, 2] \n",
    "Train_Label2=Train_Label2.to_numpy(dtype=str)\n",
    "\n",
    "Test_ID2=Data4.iloc[:, 0] \n",
    "Test_ID2=Test_ID2.to_numpy(dtype=str)\n",
    "\n",
    "Test_Position2=Data4.iloc[:, 1] \n",
    "Test_Position2=Test_Position2.to_numpy(dtype=str)\n",
    "\n",
    "Test_Label2=Data4.iloc[:, 2] \n",
    "Test_Label2=Test_Label2.to_numpy(dtype=str)\n",
    "\n",
    "dataset2=np.stack((np.concatenate((Train_ID2, Test_ID2)), np.concatenate((Train_Position2, Test_Position2)), np.concatenate((Train_Label2, Test_Label2))), axis=1)\n",
    "print (np.shape(dataset2))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "962ea3f5-518b-4be4-9ecf-1c7feff78b03",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Split: 2722\n",
      "Test_size: 302\n"
     ]
    }
   ],
   "source": [
    "test_kmers2, _IDs, _sites, y_test2=make_kmers (Test_ID2, Test_Position2, Test_Label2, protein_dict2)\n",
    "train_kmers2, _IDs, _sites, y_train2=make_kmers (Train_ID2, Train_Position2, Train_Label2, protein_dict2)\n",
    "\n",
    "print (\"Split: \" +str(len(y_train2)))\n",
    "print (\"Test_size: \" +str(len(y_test2)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "ce6b87a4-d4a9-4688-b95f-96e40205de83",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[['LACPGATEAEARAALVRAAGTLEDVYNRTQSLLAERGLLQLA-----------' '1']\n",
      " ['ENLKAMATSQFHFNLRTVRDLSMFSQNMTHIIKDVIKYKEGAVDDIFVEACQG' '1']\n",
      " ['KGFLEESVSALDIKENCMRVGLVAYSNETKVINSLSMGINKSEVLQHIQNLSP' '1']\n",
      " ...\n",
      " ['PNNFRDSLEDCVTIWGPEGRWNDSPCNQSLPSICKKAGQLSQGAAEEDHGCRK' '0']\n",
      " ['RDQPGYSRGGCVALATGSAMGLWEVKNCTSFRARYICRQSLGTPVTPELPGPD' '0']\n",
      " ['QKGTDPSLSPSPAALPPAPGTELSYLNGTFRLLQKPLRWHDALLLCESRNASL' '0']]\n",
      "(3024, 2)\n",
      "(3024, 2)\n",
      "(3024, 2)\n",
      "(2722, 2)\n",
      "(302, 2)\n",
      "(2722, 2)\n",
      "(302, 2)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|███████████████████████████████████████████████████████████████████████████| 2422/2422 [00:00<00:00, 16337.36it/s]\n",
      "100%|█████████████████████████████████████████████████████████████████████████████| 300/300 [00:00<00:00, 11354.47it/s]\n",
      "100%|██████████████████████████████████████████████████████████████████████████████| 302/302 [00:00<00:00, 3048.95it/s]\n",
      "100%|███████████████████████████████████████████████████████████████████████████| 2422/2422 [00:00<00:00, 18553.99it/s]\n",
      "100%|█████████████████████████████████████████████████████████████████████████████| 300/300 [00:00<00:00, 12149.30it/s]\n",
      "100%|█████████████████████████████████████████████████████████████████████████████| 302/302 [00:00<00:00, 46514.39it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'numpy.ndarray'>\n",
      "int32\n",
      "int32\n",
      "Model: \"sequential_30\"\n",
      "_________________________________________________________________\n",
      " Layer (type)                Output Shape              Param #   \n",
      "=================================================================\n",
      " emb_layer (Embedding)       (None, 53, 21)            5376      \n",
      "                                                                 \n",
      " reshape_30 (Reshape)        (None, 53, 21, 1)         0         \n",
      "                                                                 \n",
      " conv2d_30 (Conv2D)          (None, 51, 19, 16)        160       \n",
      "                                                                 \n",
      " max_pooling2d_30 (MaxPoolin  (None, 25, 9, 16)        0         \n",
      " g2D)                                                            \n",
      "                                                                 \n",
      " flatten_30 (Flatten)        (None, 3600)              0         \n",
      "                                                                 \n",
      " dense_60 (Dense)            (None, 64)                230464    \n",
      "                                                                 \n",
      " dropout_30 (Dropout)        (None, 64)                0         \n",
      "                                                                 \n",
      " dense_61 (Dense)            (None, 1)                 65        \n",
      "                                                                 \n",
      "=================================================================\n",
      "Total params: 236,065\n",
      "Trainable params: 236,065\n",
      "Non-trainable params: 0\n",
      "_________________________________________________________________\n",
      "Epoch 1/400\n",
      "10/10 [==============================] - 3s 105ms/step - loss: 0.6519 - accuracy: 0.6685 - val_loss: 0.6375 - val_accuracy: 0.6700\n",
      "Epoch 2/400\n",
      "10/10 [==============================] - 0s 40ms/step - loss: 0.6361 - accuracy: 0.6689 - val_loss: 0.6342 - val_accuracy: 0.6700\n",
      "Epoch 3/400\n",
      "10/10 [==============================] - 0s 34ms/step - loss: 0.6319 - accuracy: 0.6689 - val_loss: 0.6327 - val_accuracy: 0.6700\n",
      "Epoch 4/400\n",
      "10/10 [==============================] - 0s 35ms/step - loss: 0.6298 - accuracy: 0.6689 - val_loss: 0.6290 - val_accuracy: 0.6700\n",
      "Epoch 5/400\n",
      "10/10 [==============================] - 0s 38ms/step - loss: 0.6234 - accuracy: 0.6689 - val_loss: 0.6262 - val_accuracy: 0.6700\n",
      "Epoch 6/400\n",
      "10/10 [==============================] - 0s 34ms/step - loss: 0.6179 - accuracy: 0.6689 - val_loss: 0.6233 - val_accuracy: 0.6700\n",
      "Epoch 7/400\n",
      "10/10 [==============================] - 0s 36ms/step - loss: 0.6099 - accuracy: 0.6689 - val_loss: 0.6206 - val_accuracy: 0.6700\n",
      "Epoch 8/400\n",
      "10/10 [==============================] - 0s 35ms/step - loss: 0.6038 - accuracy: 0.6701 - val_loss: 0.6365 - val_accuracy: 0.6700\n",
      "Epoch 9/400\n",
      "10/10 [==============================] - 0s 44ms/step - loss: 0.6046 - accuracy: 0.6813 - val_loss: 0.6155 - val_accuracy: 0.6767\n",
      "Epoch 10/400\n",
      "10/10 [==============================] - 0s 46ms/step - loss: 0.5847 - accuracy: 0.6800 - val_loss: 0.6155 - val_accuracy: 0.6800\n",
      "Epoch 11/400\n",
      "10/10 [==============================] - 0s 41ms/step - loss: 0.5756 - accuracy: 0.7027 - val_loss: 0.6110 - val_accuracy: 0.6967\n",
      "Epoch 12/400\n",
      "10/10 [==============================] - 0s 37ms/step - loss: 0.5626 - accuracy: 0.7184 - val_loss: 0.6115 - val_accuracy: 0.6900\n",
      "Epoch 13/400\n",
      "10/10 [==============================] - 0s 36ms/step - loss: 0.5500 - accuracy: 0.7238 - val_loss: 0.6099 - val_accuracy: 0.6867\n",
      "Epoch 14/400\n",
      "10/10 [==============================] - 0s 36ms/step - loss: 0.5369 - accuracy: 0.7366 - val_loss: 0.6309 - val_accuracy: 0.6867\n",
      "Epoch 15/400\n",
      "10/10 [==============================] - 0s 35ms/step - loss: 0.5230 - accuracy: 0.7469 - val_loss: 0.6131 - val_accuracy: 0.7000\n",
      "Epoch 16/400\n",
      "10/10 [==============================] - 0s 36ms/step - loss: 0.5146 - accuracy: 0.7531 - val_loss: 0.6585 - val_accuracy: 0.6933\n",
      "Epoch 17/400\n",
      "10/10 [==============================] - 0s 39ms/step - loss: 0.5116 - accuracy: 0.7581 - val_loss: 0.6233 - val_accuracy: 0.7000\n",
      "Epoch 18/400\n",
      "10/10 [==============================] - 0s 38ms/step - loss: 0.4907 - accuracy: 0.7758 - val_loss: 0.6294 - val_accuracy: 0.6567\n",
      "Epoch 19/400\n",
      "10/10 [==============================] - 0s 36ms/step - loss: 0.4806 - accuracy: 0.7742 - val_loss: 0.6352 - val_accuracy: 0.6833\n",
      "Epoch 20/400\n",
      "10/10 [==============================] - 0s 38ms/step - loss: 0.4719 - accuracy: 0.7849 - val_loss: 0.6401 - val_accuracy: 0.6600\n",
      "Epoch 21/400\n",
      "10/10 [==============================] - 0s 35ms/step - loss: 0.4536 - accuracy: 0.7903 - val_loss: 0.6473 - val_accuracy: 0.6500\n",
      "Epoch 22/400\n",
      "10/10 [==============================] - 0s 46ms/step - loss: 0.4367 - accuracy: 0.8142 - val_loss: 0.6622 - val_accuracy: 0.6667\n",
      "Epoch 23/400\n",
      " 9/10 [==========================>...] - ETA: 0s - loss: 0.4237 - accuracy: 0.8160Restoring model weights from the end of the best epoch: 13.\n",
      "10/10 [==============================] - 0s 41ms/step - loss: 0.4260 - accuracy: 0.8146 - val_loss: 0.7120 - val_accuracy: 0.6700\n",
      "Epoch 23: early stopping\n",
      "10/10 [==============================] - 0s 11ms/step - loss: 0.6220 - accuracy: 0.6689\n",
      "10/10 [==============================] - 0s 7ms/step\n",
      "Without labels\n",
      "AUC: 0.6238\n",
      "AUPRC: 0.7563\n",
      "MCC: 0.0903\n",
      "F1: 0.7942\n",
      "Sensitivity: 0.5906\n",
      "Specificity: 0.5393\n",
      "Precision: 0.6796\n",
      "Epoch 1/400\n",
      "10/10 [==============================] - 0s 45ms/step - loss: 0.5469 - accuracy: 0.7312 - val_loss: 0.5717 - val_accuracy: 0.6900\n",
      "Epoch 2/400\n",
      "10/10 [==============================] - 0s 39ms/step - loss: 0.5372 - accuracy: 0.7436 - val_loss: 0.5616 - val_accuracy: 0.7133\n",
      "Epoch 3/400\n",
      "10/10 [==============================] - 0s 34ms/step - loss: 0.5275 - accuracy: 0.7436 - val_loss: 0.5566 - val_accuracy: 0.7133\n",
      "Epoch 4/400\n",
      "10/10 [==============================] - 0s 42ms/step - loss: 0.5234 - accuracy: 0.7510 - val_loss: 0.5598 - val_accuracy: 0.7100\n",
      "Epoch 5/400\n",
      "10/10 [==============================] - 0s 32ms/step - loss: 0.5351 - accuracy: 0.7481 - val_loss: 0.5818 - val_accuracy: 0.6767\n",
      "Epoch 6/400\n",
      "10/10 [==============================] - 0s 35ms/step - loss: 0.5098 - accuracy: 0.7581 - val_loss: 0.5642 - val_accuracy: 0.7133\n",
      "Epoch 7/400\n",
      "10/10 [==============================] - 0s 35ms/step - loss: 0.4989 - accuracy: 0.7675 - val_loss: 0.5859 - val_accuracy: 0.6833\n",
      "Epoch 8/400\n",
      "10/10 [==============================] - 0s 33ms/step - loss: 0.4894 - accuracy: 0.7733 - val_loss: 0.5614 - val_accuracy: 0.7067\n",
      "Epoch 9/400\n",
      "10/10 [==============================] - 0s 33ms/step - loss: 0.4798 - accuracy: 0.7845 - val_loss: 0.5573 - val_accuracy: 0.7233\n",
      "Epoch 10/400\n",
      "10/10 [==============================] - 0s 34ms/step - loss: 0.4774 - accuracy: 0.7985 - val_loss: 0.5641 - val_accuracy: 0.7100\n",
      "Epoch 11/400\n",
      "10/10 [==============================] - 0s 34ms/step - loss: 0.4600 - accuracy: 0.7944 - val_loss: 0.6015 - val_accuracy: 0.6967\n",
      "Epoch 12/400\n",
      "10/10 [==============================] - 0s 38ms/step - loss: 0.4735 - accuracy: 0.7762 - val_loss: 0.5640 - val_accuracy: 0.7000\n",
      "Epoch 13/400\n",
      " 9/10 [==========================>...] - ETA: 0s - loss: 0.4421 - accuracy: 0.8095Restoring model weights from the end of the best epoch: 3.\n",
      "10/10 [==============================] - 0s 40ms/step - loss: 0.4455 - accuracy: 0.8068 - val_loss: 0.5696 - val_accuracy: 0.7100\n",
      "Epoch 13: early stopping\n",
      "10/10 [==============================] - 0s 5ms/step - loss: 0.6285 - accuracy: 0.6821\n",
      "10/10 [==============================] - 0s 5ms/step\n",
      "With labels\n",
      "AUC: 0.614\n",
      "AUPRC: 0.7462\n",
      "MCC: 0.1693\n",
      "F1: 0.7966\n",
      "Sensitivity: 0.5558\n",
      "Specificity: 0.5625\n",
      "Precision: 0.6963\n"
     ]
    }
   ],
   "source": [
    "make_datasets_train_models (dataset2, protein_dict2, 2722, folder=\"_model_results_NGlyDE\", trial=\"Trial1\")\n",
    "#make_datasets_train_models (dataset2, protein_dict2, 2722, folder=\"_model_results_NGlyDE\", trial=\"Trial4\")\n",
    "#make_datasets_train_models (dataset2, protein_dict2, 2722, folder=\"_model_results_NGlyDE\", trial=\"Trial5\")\n",
    "#make_datasets_train_models (dataset2, protein_dict2, 2722, folder=\"_model_results_NGlyDE\", trial=\"Trial6\")\n",
    "#make_datasets_train_models (dataset2, protein_dict2, 2722, folder=\"_model_results_NGlyDE\", trial=\"Trial7\")\n",
    "#make_datasets_train_models (dataset2, protein_dict2, 2722, folder=\"_model_results_NGlyDE\", trial=\"Trial8\")\n",
    "#make_datasets_train_models (dataset2, protein_dict2, 2722, folder=\"_model_results_NGlyDE\", trial=\"Trial9\")\n",
    "#make_datasets_train_models (dataset2, protein_dict2, 2722, folder=\"_model_results_NGlyDE\", trial=\"Trial10\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "id": "7bbf12bc-3d24-4a53-b6a6-93f78e217795",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Sequence key error for Q9UF12 at 31\n",
      "Sequence key error for P01019 at 304\n",
      "Q6IE36 not in protein_dict\n",
      "Sequence key error for Q8TD46 at 99\n",
      "Sequence key error for O15131 at 404\n",
      "Sequence key error for Q8WUH1 at 47\n",
      "Q9NY84 not in protein_dict\n",
      "Sequence key error for Q9UKU6 at 89\n",
      "Sequence key error for Q02817 at 4438\n",
      "Site outside of seq bounds, site: 3855, sequnce length: 2040\n",
      "P08519\n",
      "-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------\n",
      "1841\n",
      "Sequence key error for P15692 at 101\n",
      "Sequence key error for P01019 at 170\n",
      "Sequence key error for P01019 at 47\n",
      "Sequence key error for Q96Q06 at 409\n",
      "Sequence key error for Q02817 at 5019\n",
      "Sequence key error for Q8NFD5 at 899\n",
      "Sequence key error for Q9HCJ0 at 675\n",
      "J3QSS3 not in protein_dict\n",
      "D3DNC2 not in protein_dict\n",
      "Sequence key error for Q8NFD5 at 2017\n",
      "Sequence key error for Q9P2C4 at 185\n",
      "Sequence key error for Q8TD46 at 195\n",
      "Sequence key error for Q2KHR2 at 1179\n",
      "Sequence key error for Q9UKU6 at 800\n",
      "Sequence key error for Q9P2C4 at 299\n",
      "Sequence key error for Q8N841 at 132\n",
      "Sequence key error for H0YJS3 at 115\n",
      "Sequence key error for O43451 at 1847\n",
      "D3DP68 not in protein_dict\n",
      "Sequence key error for Q96KR4 at 615\n",
      "Sequence key error for Q02817 at 4970\n",
      "Sequence key error for Q02817 at 4752\n",
      "Sequence key error for Q5JV73 at 923\n",
      "Sequence key error for Q9HCJ0 at 770\n",
      "Sequence key error for Q9HBL0 at 132\n",
      "Sequence key error for Q8N841 at 274\n",
      "Sequence key error for Q02817 at 5069\n",
      "Sequence key error for Q9HCJ0 at 759\n",
      "Sequence key error for Q9HCJ0 at 474\n",
      "Sequence key error for Q8TD46 at 37\n",
      "Sequence key error for P18507 at 393\n",
      "Sequence key error for Q9UKU6 at 605\n",
      "Sequence key error for Q02817 at 4888\n",
      "Sequence key error for P01876 at 340\n",
      "Sequence key error for Q9UKU6 at 906\n",
      "Sequence key error for Q96JG9 at 3396\n",
      "Site outside of seq bounds, site: 197, sequnce length: 172\n",
      "Q96DA0\n",
      "Sequence key error for Q9UKU6 at 684\n",
      "Sequence key error for Q02817 at 4787\n",
      "Sequence key error for O94964 at 1102\n",
      "Sequence key error for Q9NZU1 at 277\n",
      "D3DR43 not in protein_dict\n",
      "Sequence key error for Q96T25 at 52\n",
      "Sequence key error for Q62918 at 635\n",
      "Sequence key error for Q03518 at 362\n",
      "Sequence key error for P08686 at 346\n",
      "Sequence key error for Q5FW52 at 213\n",
      "Sequence key error for Q8WUG5 at 32\n",
      "Sequence key error for Q03518 at 339\n",
      "Sequence key error for Q62918 at 53\n",
      "Sequence key error for Q4Z4D2 at 352\n",
      "Sequence key error for Q03518 at 287\n",
      "Sequence key error for Q92538 at 813\n",
      "Sequence key error for Q4Z4D2 at 184\n",
      "Sequence key error for Q92538 at 1603\n",
      "Sequence key error for Q4Z4D2 at 365\n",
      "Sequence key error for Q9HBL0 at 328\n",
      "Sequence key error for Q4Z4D2 at 23\n",
      "Sequence key error for Q4Z4D2 at 385\n",
      "Sequence key error for Q9HBL0 at 383\n",
      "Sequence key error for Q4Z4D2 at 28\n",
      "Sequence key error for Q6RKD8 at 614\n",
      "Sequence key error for Q62918 at 517\n",
      "Sequence key error for Q9HBL0 at 1063\n",
      "Sequence key error for Q62918 at 298\n",
      "Sequence key error for Q03518 at 310\n",
      "Sequence key error for Q4Z4D2 at 92\n",
      "Sequence key error for Q4Z4D2 at 305\n",
      "Sequence key error for Q13636 at 33\n",
      "Sequence key error for Q8N8F6 at 39\n",
      "Sequence key error for Q8N8F6 at 50\n",
      "Sequence key error for Q8N8F6 at 146\n",
      "Sequence key error for Q6SJ96 at 73\n",
      "Site outside of seq bounds, site: 1553, sequnce length: 1550\n",
      "Q5T890\n",
      "Sequence key error for Q2KHR2 at 1058\n",
      "Sequence key error for P59826 at 142\n",
      "Sequence key error for Q9H195 at 955\n",
      "Sequence key error for Q9HBL0 at 905\n",
      "Sequence key error for Q62918 at 293\n",
      "E7ER45 not in protein_dict\n",
      "Sequence key error for Q8TCB6 at 90\n",
      "[[14 29]\n",
      " [11 34]\n",
      " [25 30]\n",
      " [ 5 33]\n",
      " [22 15]\n",
      " [20 27]\n",
      " [ 7 28]\n",
      " [12 20]\n",
      " [ 4 28]\n",
      " [ 9 31]\n",
      " [10 29]\n",
      " [ 8 29]\n",
      " [23 20]\n",
      " [24 27]\n",
      " [21 27]\n",
      " [26 25]\n",
      " [18 29]\n",
      " [ 1 39]\n",
      " [ 3 25]\n",
      " [17 19]\n",
      " [ 6 31]\n",
      " [19 15]\n",
      " [15 25]\n",
      " [16 17]\n",
      " [13 28]]\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAjMAAAHFCAYAAAAHcXhbAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjguMCwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy81sbWrAAAACXBIWXMAAA9hAAAPYQGoP6dpAABSeElEQVR4nO3dd1QUZ/s+8GukLEVAUWkBASNgN7bYBQvWoGKJxoYtibG3aCx5xQaWSPSNiSUqtp8licYudkliiVhQNL6KCogKYgtNXQSe3x8e9stK20FwGbw+5+w5mWdn7rlneQiXM7O7khBCgIiIiEihyui7ASIiIqK3wTBDREREisYwQ0RERIrGMENERESKxjBDREREisYwQ0RERIrGMENERESKxjBDREREisYwQ0RERIrGMEOlxvr16yFJUq6PyZMn67u999axY8fQsGFDmJubQ5Ik7Nq1S98tyZI1r86fP1+s+/H399eas8bGxnB1dcW4cePw77//AkCe8/vNx8mTJxEdHa1Z9vf3z3WfQ4cO1axTFLy8vODl5aVZfv78Ofz9/XHy5MkiqU+UF0N9N0BU1IKDg1GtWjWtMQcHBz11834TQuDTTz+Fu7s79uzZA3Nzc3h4eOi7rRItJCQEVlZWSE5OxoEDB7Bs2TKcO3cOp0+fxpkzZ7TWnTt3Lk6cOIHjx49rjdeoUQNPnz4FAFhYWGD9+vX4z3/+gzJl/u/frykpKfj1119haWmJpKSkIun9p59+0lp+/vw5Zs+eDQBaIYeoqDHMUKlTq1YtNGzYUKd1X716BUmSYGjIX4Xi8ODBAzx9+hS+vr5o27atvtuRJWtuvGsNGjRAxYoVAQDe3t548uQJNm3ahNOnT6N58+Za61aqVAllypRBkyZNctTJCjN9+vTBmjVrcOzYMXh7e2ue3759OzIyMtC9e3ds3ry5SHqvUaNGkdQhkouXmei9cfLkSUiShE2bNmHSpEn44IMPoFKpcOvWLQDA0aNH0bZtW1haWsLMzAzNmzfHsWPHctTZv38/PvroI6hUKri6uuK7777TXCLIknWKf/369Tm2z+20f2RkJPr16wcbGxuoVCpUr14dP/74Y679b926FTNmzICDgwMsLS3Rrl073LhxI8d+QkJC0LZtW1hZWcHMzAzVq1dHYGAgAGDTpk2QJCnHv/QBYM6cOTAyMsKDBw/yfT3/+usvtG3bFhYWFjAzM0OzZs2wf/9+zfP+/v5wdHQEAEydOhWSJMHFxSXPenKPT5ef161btzBkyBC4ubnBzMwMH3zwAXx8fBAREZHrvvOaGwDw7NkzDBkyBNbW1jA3N4ePjw/u3LmjeX7u3LkwNDREbGxsjl6HDh2KChUq4OXLl/m+prnJCioxMTGytwUADw8PNGvWDOvWrdMaX7duHXr06AErKyud6ty5cwd9+/aFg4MDVCoVbG1t0bZtW4SHh2vWyX6ZKTo6GpUqVQIAzJ49W3M5a/DgwZr1dZn3mZmZmDdvHjw8PGBqaopy5cqhTp06WLZsmfwXg0othhkqdTIyMpCenq71yG7atGm4e/cuVq5cib1798LGxgabN29G+/btYWlpiQ0bNuCXX36BtbU1OnTooPUH8tixY+jWrRssLCywbds2LF68GL/88guCg4ML3e8///yDRo0a4erVq1iyZAn27duHLl26YOzYsZpT9NlNnz4dMTExWLNmDVavXo3IyEj4+PggIyNDs87atWvRuXNnZGZmao5z7NixuHfvHoDX/1q3s7PL8YcjPT0dq1atgq+vb76X5kJDQ9GmTRskJiZi7dq12Lp1KywsLODj44Pt27cDAIYPH46dO3cCAMaMGYMzZ87g999/L/D10OX4dP15PXjwABUqVMCCBQsQEhKCH3/8EYaGhmjcuHGuASm3uZFl2LBhKFOmDLZs2YKlS5fi3Llz8PLy0tzP8uWXX8LQ0BCrVq3Sqvn06VNs27YNw4YNg4mJSYHH/6asQJUVDApj2LBh2LVrF549ewYAuHHjBk6fPo1hw4bpXKNz5864cOECFi1ahCNHjmDFihWoV6+e5vjfZG9vj5CQEM3+z5w5gzNnzuDbb78FoPu8X7RoEfz9/fHZZ59h//792L59O4YNG5bnfuk9JYhKieDgYAEg18erV6/EiRMnBADRqlUrre1SU1OFtbW18PHx0RrPyMgQdevWFR9//LFmrHHjxsLBwUG8ePFCM5aUlCSsra1F9l+nqKgoAUAEBwfn6BOAmDVrlma5Q4cOwtHRUSQmJmqtN3r0aGFiYiKePn0qhBCa/jt37qy13i+//CIAiDNnzgghhEhOThaWlpaiRYsWIjMzM8/Xa9asWcLY2Fg8fPhQM7Z9+3YBQISGhua5nRBCNGnSRNjY2Ijk5GTNWHp6uqhVq5ZwdHTU7DfrdVi8eHG+9eQcn5yf15vS09NFWlqacHNzExMmTMix7zfnhhD/N698fX21xk+dOiUAiHnz5mnG/Pz8hI2NjVCr1ZqxhQsXijJlyoioqKh8j3/WrFkCgIiPjxevXr0Sz549E5s3bxampqbCyclJa85l35+5uXmu9bK/9snJyaJs2bJi+fLlQgghvv76a+Hq6ioyMzPFqFGjREF/Ch4/fiwAiKVLl+a7nqenp/D09NQsP3r0KMd8z6LrvP/kk0/ERx99lO9+iXhmhkqdjRs3IiwsTOuR/Z6Ynj17aq1/+vRpPH36FH5+flpnczIzM9GxY0eEhYUhNTUVqampCAsLQ48ePbT+hZ11RqIwXr58iWPHjsHX1xdmZmZa++/cuTNevnyJs2fPam3TtWtXreU6deoA+L/LEKdPn0ZSUhJGjhyZ7z0fX331FQDg559/1owtX74ctWvXRqtWrfLcLjU1FX///Td69eqFsmXLasYNDAwwcOBA3Lt3L9ezHrrS5fh0+XkBr880BQQEoEaNGjA2NoahoSGMjY0RGRmJ69ev59j3m3Mju/79+2stN2vWDM7Ozjhx4oRmbNy4cUhISMCvv/4K4PUlkhUrVqBLly75XmLLzs7ODkZGRihfvjwGDBiA+vXrIyQkpFBndbKULVsWvXv3xrp165Ceno6NGzdiyJAhOt8TZG1tjQ8//BCLFy9GUFAQLl26hMzMzEL3I2fef/zxx7h8+TJGjhyJQ4cOFdnNylS68K5HKnWqV6+e7w3A9vb2WssPHz4EAPTq1SvPbZ4+fQpJkpCZmQk7O7scz+c2posnT54gPT0dP/zwA3744Ydc13n8+LHWcoUKFbSWVSoVAODFixcAgEePHgGA5n6VvNja2qJPnz5YtWoVvvnmG1y7dg1//vlnjsskb3r27BmEEDleR+D/3jX25MmTfGvkp6Dj0/XnZW5ujokTJ+LHH3/E1KlT4enpifLly6NMmTIYPny4pl52uR1Tlrx+7tmPtV69emjZsiV+/PFH9O/fH/v27UN0dHSBr2l2R48ehZWVFYyMjODo6Jjj9SisYcOGoUWLFpg/fz4ePXqkde9KQSRJwrFjxzBnzhwsWrQIkyZNgrW1Nfr374/58+fDwsJCVi9y5v20adNgbm6OzZs3Y+XKlTAwMECrVq2wcOFCnW/0p9KPYYbeO2/+azTrnSM//PBDru8KAV7/4c96d0t8fHyO598cy/pXtFqt1hp/8498+fLlNWc0Ro0aleu+XV1d8zmanLLurci6PyY/48aNw6ZNm7B7926EhISgXLlyOc5AvCkrEMTFxeV4Luum4azXtDjo+vMCXt9bM2jQIAQEBGg9//jxY5QrVy7Hdvmdqcjr5161alWtsbFjx6J37964ePEili9fDnd3d613ERWkbt26xfL6NW/eHB4eHpgzZw68vb3h5OQka3tnZ2esXbsWAHDz5k388ssv8Pf3R1paGlauXCmrlpx5b2hoiIkTJ2LixIn4999/cfToUUyfPh0dOnRAbGwszMzMZO2bSieGGXrvNW/eHOXKlcM///yD0aNH57mesbExPv74Y+zcuROLFy/WBJbk5GTs3btXa11bW1uYmJjgypUrWuO7d+/WWjYzM0Pr1q1x6dIl1KlTB8bGxm99PM2aNYOVlRVWrlyJvn375vsHukGDBmjWrBkWLlyIq1ev4osvvoC5uXm+9c3NzdG4cWPs3LkT3333HUxNTQG8vqSyefNmODo6wt3d/a2PIy+6/ryA1+Ek68xOlv379+P+/fs5QkhB/t//+39al6FOnz6NmJgYDB8+XGs9X19fVK5cGZMmTUJoaCi+//57vbzFOzczZ87Eb7/9lmeA0JW7uztmzpyJHTt24OLFi3mu9+ZZtSyFnfflypVDr169cP/+fYwfPx7R0dF8OzgBYJghQtmyZfHDDz/Az88PT58+Ra9evWBjY4NHjx7h8uXLePToEVasWAHg9dtvO3bsCG9vb0yaNAkZGRlYuHAhzM3NNZ/rAbz+IzpgwACsW7cOH374IerWrYtz585hy5YtOfa/bNkytGjRAi1btsRXX30FFxcXJCcn49atW9i7d2+OD0TT5XiWLFmC4cOHo127dvj8889ha2uLW7du4fLly1i+fLnW+uPGjUOfPn0gSRJGjhyp0z4CAwPh7e2N1q1bY/LkyTA2NsZPP/2Eq1evYuvWrcX6x1vOz+uTTz7B+vXrUa1aNdSpUwcXLlzA4sWLC7wEl5vz589j+PDh6N27N2JjYzFjxgx88MEHOV4zAwMDjBo1ClOnToW5ubmsyznFbcCAARgwYIDs7a5cuYLRo0ejd+/ecHNzg7GxMY4fP44rV67gm2++yXM7CwsLODs7Y/fu3Wjbti2sra1RsWJFuLi46DzvfXx8NJ8dValSJcTExGDp0qVwdnaGm5tboV8LKl0YZojw+n/ylStXxqJFi/Dll18iOTkZNjY2+Oijj7T+GHl7e2PXrl2YOXOm5u3NI0eOxIsXL3K8jXrJkiUAXr+1NCUlBW3atMG+ffty3Ahao0YNXLx4EXPnzsXMmTORkJCAcuXKwc3NDZ07dy7U8QwbNgwODg5YuHAhhg8fDiEEXFxc4Ofnl2Pd7t27Q6VSoXXr1jr/cfD09MTx48cxa9YsDB48GJmZmahbty727NmDTz75pFA9y6Hrz2vZsmUwMjJCYGAgUlJSUL9+fezcuRMzZ86Uvc+1a9di06ZN6Nu3L9RqNVq3bo1ly5bB2to6x7p9+vTB1KlTMXDgQJ0/x6Uks7Ozw4cffoiffvoJsbGxkCQJVapUwZIlSzBmzJh8t127di2+/vprdO3aFWq1Gn5+fli/fr3O875169bYsWMH1qxZg6SkJNjZ2cHb2xvffvstjIyMivvQSSEkIYTQdxNESufv74/Zs2dDib9Oe/fuRdeuXbF///5ChyfS9sMPP2Ds2LG4evUqatasqe92iEo9npkhek/9888/iImJwaRJk/DRRx+hU6dO+m5J8S5duoSoqCjMmTMH3bp1Y5AhekcYZojeUyNHjsSpU6dQv359bNiwocTcpKpkvr6+iI+PR8uWLWW/w4eICo+XmYiIiEjR+AnAREREpGgMM0RERKRoDDNERESkaKX+BuDMzEw8ePAAFhYWvMGRiIhIIYQQSE5OhoODA8qUyf/cS6kPMw8ePJD9HSRERERUMsTGxhb4qd2lPsxkfZtrbGwsLC0t9dwNERER6SIpKQlOTk46fSt7qQ8zWZeWLC0tGWaIiIgURpdbRHgDMBERESlaiQkzgYGBkCQJ48eP14wJIeDv7w8HBweYmprCy8sL165d01+TREREVOKUiDATFhaG1atXo06dOlrjixYtQlBQEJYvX46wsDDNt6UmJyfrqVMiIiIqafQeZlJSUtC/f3/8/PPPKF++vGZcCIGlS5dixowZ6NGjB2rVqoUNGzbg+fPn2LJlix47JiIiopJE72Fm1KhR6NKlC9q1a6c1HhUVhfj4eLRv314zplKp4OnpidOnT7/rNomIiKiE0uu7mbZt24aLFy8iLCwsx3Px8fEAAFtbW61xW1tbxMTE5FlTrVZDrVZrlpOSkoqoWyIiIiqJ9HZmJjY2FuPGjcPmzZthYmKS53pvviVLCJHv27QCAwNhZWWlefAD84iIiEo3vYWZCxcuICEhAQ0aNIChoSEMDQ0RGhqK//73vzA0NNSckck6Q5MlISEhx9ma7KZNm4bExETNIzY2tliPg4iIiPRLb5eZ2rZti4iICK2xIUOGoFq1apg6dSqqVKkCOzs7HDlyBPXq1QMApKWlITQ0FAsXLsyzrkqlgkqlKtbeiYiIqOTQW5ixsLBArVq1tMbMzc1RoUIFzfj48eMREBAANzc3uLm5ISAgAGZmZujXr58+WiYiIqISqER/ncGUKVPw4sULjBw5Es+ePUPjxo1x+PBhnb6ngYiIiN4PkhBC6LuJ4pSUlAQrKyskJibyu5mIiIgUQs7fb71/zgwRERHR22CYISIiIkVjmCEiIiJFK9E3ACuByzf7C7Vd9IIuRdwJERHR+4lnZoiIiEjRGGaIiIhI0RhmiIiISNEYZoiIiEjRGGaIiIhI0RhmiIiISNEYZoiIiEjRGGaIiIhI0RhmiIiISNEYZoiIiEjRGGaIiIhI0RhmiIiISNEYZoiIiEjRGGaIiIhI0RhmiIiISNEYZoiIiEjRGGaIiIhI0RhmiIiISNEYZoiIiEjRGGaIiIhI0RhmiIiISNEYZoiIiEjRGGaIiIhI0RhmiIiISNEYZoiIiEjRGGaIiIhI0RhmiIiISNEYZoiIiEjRGGaIiIhI0RhmiIiISNEYZoiIiEjR9BpmVqxYgTp16sDS0hKWlpZo2rQpDh48qHl+8ODBkCRJ69GkSRM9dkxEREQljaE+d+7o6IgFCxagatWqAIANGzagW7duuHTpEmrWrAkA6NixI4KDgzXbGBsb66VXIiIiKpn0GmZ8fHy0lufPn48VK1bg7NmzmjCjUqlgZ2enj/aIiIhIAUrMPTMZGRnYtm0bUlNT0bRpU834yZMnYWNjA3d3d3z++edISEjIt45arUZSUpLWg4iIiEovvYeZiIgIlC1bFiqVCiNGjMDvv/+OGjVqAAA6deqE//f//h+OHz+OJUuWICwsDG3atIFarc6zXmBgIKysrDQPJyend3UoREREpAeSEELos4G0tDTcvXsX//77L3bs2IE1a9YgNDRUE2iyi4uLg7OzM7Zt24YePXrkWk+tVmuFnaSkJDg5OSExMRGWlpZF3r/LN/sLtV30gi5F3AkREVHpkZSUBCsrK53+fuv1nhng9Q29WTcAN2zYEGFhYVi2bBlWrVqVY117e3s4OzsjMjIyz3oqlQoqlarY+iUiIqKSRe+Xmd4khMjzMtKTJ08QGxsLe3v7d9wVERERlVR6PTMzffp0dOrUCU5OTkhOTsa2bdtw8uRJhISEICUlBf7+/ujZsyfs7e0RHR2N6dOno2LFivD19dVn20RERFSC6DXMPHz4EAMHDkRcXBysrKxQp04dhISEwNvbGy9evEBERAQ2btyIf//9F/b29mjdujW2b98OCwsLfbZNREREJYhew8zatWvzfM7U1BSHDh16h90QERGREpW4e2aIiIiI5GCYISIiIkXT+1uziejd4+cjEVFpwjMzREREpGgMM0RERKRoDDNERESkaAwzREREpGgMM0RERKRoDDNERESkaAwzREREpGgMM0RERKRo/NA8KnL8QDYiInqXeGaGiIiIFI1hhoiIiBSNYYaIiIgUjWGGiIiIFI1hhoiIiBSNYYaIiIgUjWGGiIiIFI1hhoiIiBSNYYaIiIgUjWGGiIiIFI1hhoiIiBSNYYaIiIgUjWGGiIiIFI1hhoiIiBSNYYaIiIgUjWGGiIiIFI1hhoiIiBTNUN8NEOXF5Zv9hdouekGXIu6kaJS24yEiKil4ZoaIiIgUjWGGiIiIFI1hhoiIiBSNYYaIiIgUTa9hZsWKFahTpw4sLS1haWmJpk2b4uDBg5rnhRDw9/eHg4MDTE1N4eXlhWvXrumxYyIiIipp9BpmHB0dsWDBApw/fx7nz59HmzZt0K1bN01gWbRoEYKCgrB8+XKEhYXBzs4O3t7eSE5O1mfbREREVILoNcz4+Pigc+fOcHd3h7u7O+bPn4+yZcvi7NmzEEJg6dKlmDFjBnr06IFatWphw4YNeP78ObZs2aLPtomIiKgEKTH3zGRkZGDbtm1ITU1F06ZNERUVhfj4eLRv316zjkqlgqenJ06fPp1nHbVajaSkJK0HERERlV56/9C8iIgING3aFC9fvkTZsmXx+++/o0aNGprAYmtrq7W+ra0tYmJi8qwXGBiI2bNnF2vPREREJcn7/qGcej8z4+HhgfDwcJw9exZfffUV/Pz88M8//2ielyRJa30hRI6x7KZNm4bExETNIzY2tth6JyIiIv3T+5kZY2NjVK1aFQDQsGFDhIWFYdmyZZg6dSoAID4+Hvb29pr1ExIScpytyU6lUkGlUhVv00RERFRi6P3MzJuEEFCr1XB1dYWdnR2OHDmieS4tLQ2hoaFo1qyZHjskIiKikkSvZ2amT5+OTp06wcnJCcnJydi2bRtOnjyJkJAQSJKE8ePHIyAgAG5ubnBzc0NAQADMzMzQr18/fbZNREREJYhew8zDhw8xcOBAxMXFwcrKCnXq1EFISAi8vb0BAFOmTMGLFy8wcuRIPHv2DI0bN8bhw4dhYWGhz7aJiIioBNFrmFm7dm2+z0uSBH9/f/j7+7+bhoiIiEhxStw9M0RERERyMMwQERGRoun9rdlESlCSPpCqJPVSkvB1yR1fF3of8MwMERERKRrDDBERESkawwwREREpGsMMERERKRrDDBERESkawwwREREpGsMMERERKRrDDBERESkaPzSPtJS2D9gqbcdTGvFnRERvi2dmiIiISNEYZoiIiEjRGGaIiIhI0WSHmY0bN0KtVucYT0tLw8aNG4ukKSIiIiJdyQ4zQ4YMQWJiYo7x5ORkDBkypEiaIiIiItKV7DAjhIAkSTnG7927BysrqyJpioiIiEhXOr81u169epAkCZIkoW3btjA0/L9NMzIyEBUVhY4dOxZLk0RERER50TnMdO/eHQAQHh6ODh06oGzZsprnjI2N4eLigp49exZ5g0RERET50TnMzJo1CwDg4uKCPn36wMTEpNiaIiKi0qcoPiCxNH7IYmk8pndN9icA+/n5AXj97qWEhARkZmZqPV+5cuWi6YyIiIhIB7LDTGRkJIYOHYrTp09rjWfdGJyRkVFkzREREREVRHaYGTx4MAwNDbFv3z7Y29vn+s4mIiIiondFdpgJDw/HhQsXUK1ateLoh4iIiEgW2Z8zU6NGDTx+/Lg4eiEiIiKSTXaYWbhwIaZMmYKTJ0/iyZMnSEpK0noQERERvUuyLzO1a9cOANC2bVutcd4ATERERPogO8ycOHGiOPogIiIiKhTZYcbT07M4+iAiIiIqFNlh5o8//sj3+VatWhW6GSIiIiK5ZIcZLy+vHGPZP2uG98wQERHRuyT73UzPnj3TeiQkJCAkJASNGjXC4cOHi6NHIiIiojzJPjNjZWWVY8zb2xsqlQoTJkzAhQsXiqQxIiIiIl3IPjOTl0qVKuHGjRuytgkMDESjRo1gYWEBGxsbdO/ePUeNwYMHQ5IkrUeTJk2Kqm0iIiJSONlnZq5cuaK1LIRAXFwcFixYgLp168qqFRoailGjRqFRo0ZIT0/HjBkz0L59e/zzzz8wNzfXrNexY0cEBwdrlo2NjeW2TURERKWU7DDz0UcfQZIkCCG0xps0aYJ169bJqhUSEqK1HBwcDBsbG1y4cEHrXVEqlQp2dnZyWyUiIqL3gOwwExUVpbVcpkwZVKpUCSYmJm/dTGJiIgDA2tpaa/zkyZOwsbFBuXLl4Onpifnz58PGxibXGmq1Gmq1WrPMr1ggIiIq3WSHGWdn5+LoA0IITJw4ES1atECtWrU04506dULv3r3h7OyMqKgofPvtt2jTpg0uXLgAlUqVo05gYCBmz55dLD2WdC7f7C/UdtELuhRxJ0TKxN8het8V5negJMz/Qt0AHBoaCh8fH1StWhVubm7o2rUr/vzzz7dqZPTo0bhy5Qq2bt2qNd6nTx906dIFtWrVgo+PDw4ePIibN29i//7cX/Bp06YhMTFR84iNjX2rvoiIiKhkkx1mNm/ejHbt2sHMzAxjx47F6NGjYWpqirZt22LLli2FamLMmDHYs2cPTpw4AUdHx3zXtbe3h7OzMyIjI3N9XqVSwdLSUutBREREpZfsy0zz58/HokWLMGHCBM3YuHHjEBQUhLlz56Jfv3461xJCYMyYMfj9999x8uRJuLq6FrjNkydPEBsbC3t7e7mtExERUSkk+8zMnTt34OPjk2O8a9euOW4OLsioUaOwefNmbNmyBRYWFoiPj0d8fDxevHgBAEhJScHkyZNx5swZREdH4+TJk/Dx8UHFihXh6+srt3UiIiIqhWSHGScnJxw7dizH+LFjx+Dk5CSr1ooVK5CYmAgvLy/Y29trHtu3bwcAGBgYICIiAt26dYO7uzv8/Pzg7u6OM2fOwMLCQm7rREREVArJvsw0adIkjB07FuHh4WjWrBkkScJff/2F9evXY9myZbJqvflZNW8yNTXFoUOH5LZIRERE7xHZYearr76CnZ0dlixZgl9++QUAUL16dWzfvh3dunUr8gaJiIiI8iM7zACAr68v71khIiKiEqFQYSZLSkoKMjMztcb4VmgiIiJ6l2TfABwVFYUuXbrA3NwcVlZWKF++PMqXL49y5cqhfPnyxdEjERERUZ5kn5np378/AGDdunWwtbWFJElF3hQRERGRrmSHmStXruDChQvw8PAojn6IiIiIZJF9malRo0b8viMiIiIqMWSfmVmzZg1GjBiB+/fvo1atWjAyMtJ6vk6dOkXWHBEREVFBZIeZR48e4fbt2xgyZIhmTJIkCCEgSRIyMjKKtEEiIiKi/MgOM0OHDkW9evWwdetW3gBMREREeic7zMTExGDPnj2oWrVqcfRDREREJIvsG4DbtGmDy5cvF0cvRERERLLJPjPj4+ODCRMmICIiArVr185xA3DXrl2LrDkiIiKigsgOMyNGjAAAzJkzJ8dzvAGYiIiI3jXZYebN72IiIiIi0ifZ98wQERERlSSF+tbs1NRUhIaG4u7du0hLS9N6buzYsUXSGBEREZEuZIeZS5cuoXPnznj+/DlSU1NhbW2Nx48fw8zMDDY2NgwzRERE9E7Jvsw0YcIE+Pj44OnTpzA1NcXZs2cRExODBg0a4LvvviuOHomIiIjyJDvMhIeHY9KkSTAwMICBgQHUajWcnJywaNEiTJ8+vTh6JCIiIsqT7DBjZGSk+QoDW1tb3L17FwBgZWWl+W8iIiKid0X2PTP16tXD+fPn4e7ujtatW+M///kPHj9+jE2bNqF27drF0SMRERFRnmSfmQkICIC9vT0AYO7cuahQoQK++uorJCQkYPXq1UXeIBEREVF+ZJ+Zadiwoea/K1WqhAMHDhRpQ0RERERy8EPziIiISNEYZoiIiEjRGGaIiIhI0RhmiIiISNFkh5moqKji6IOIiIioUGSHmapVq6J169bYvHkzXr58WRw9EREREelMdpi5fPky6tWrh0mTJsHOzg5ffvklzp07Vxy9ERERERVIdpipVasWgoKCcP/+fQQHByM+Ph4tWrRAzZo1ERQUhEePHhVHn0RERES5KvQNwIaGhvD19cUvv/yChQsX4vbt25g8eTIcHR0xaNAgxMXFFWWfRERERLkqdJg5f/48Ro4cCXt7ewQFBWHy5Mm4ffs2jh8/jvv376Nbt25F2ScRERFRrmSHmaCgINSuXRvNmjXDgwcPsHHjRsTExGDevHlwdXVF8+bNsWrVKly8eLHAWoGBgWjUqBEsLCxgY2OD7t2748aNG1rrCCHg7+8PBwcHmJqawsvLC9euXZPbNhEREZVSssPMihUr0K9fP9y9exe7du3CJ598gjJltMtUrlwZa9euLbBWaGgoRo0ahbNnz+LIkSNIT09H+/btkZqaqlln0aJFCAoKwvLlyxEWFgY7Ozt4e3sjOTlZbutERERUCsn+osnIyMgC1zE2Noafn1+B64WEhGgtBwcHw8bGBhcuXECrVq0ghMDSpUsxY8YM9OjRAwCwYcMG2NraYsuWLfjyyy/ltk9ERESljOwzM8HBwfj1119zjP/666/YsGHDWzWTmJgIALC2tgbw+gP64uPj0b59e806KpUKnp6eOH36dK411Go1kpKStB5ERERUesk+M7NgwQKsXLkyx7iNjQ2++OILnc7I5EYIgYkTJ6JFixaoVasWACA+Ph4AYGtrq7Wura0tYmJicq0TGBiI2bNnF6oHIpLH5Zv9hdouekGXIu6E6N3j/C85ZJ+ZiYmJgaura45xZ2dn3L17t9CNjB49GleuXMHWrVtzPCdJktayECLHWJZp06YhMTFR84iNjS10T0RERFTyyQ4zNjY2uHLlSo7xy5cvo0KFCoVqYsyYMdizZw9OnDgBR0dHzbidnR2A/ztDkyUhISHH2ZosKpUKlpaWWg8iIiIqvWSHmb59+2Ls2LE4ceIEMjIykJGRgePHj2PcuHHo27evrFpCCIwePRo7d+7E8ePHc5zxcXV1hZ2dHY4cOaIZS0tLQ2hoKJo1aya3dSIiIiqFZN8zM2/ePMTExKBt27YwNHy9eWZmJgYNGoSAgABZtUaNGoUtW7Zg9+7dsLCw0JyBsbKygqmpKSRJwvjx4xEQEAA3Nze4ubkhICAAZmZm6Nevn9zWiYiIqBSSHWaMjY2xfft2zJ07F5cvX4apqSlq164NZ2dn2TtfsWIFAMDLy0trPDg4GIMHDwYATJkyBS9evMDIkSPx7NkzNG7cGIcPH4aFhYXs/REREVHpIzvMZHF3d4e7u/tb7VwIUeA6kiTB398f/v7+b7UvIiIiKp1kh5mMjAysX78ex44dQ0JCAjIzM7WeP378eJE1R0RERFQQ2WFm3LhxWL9+Pbp06YJatWrl+RZpIiIiondBdpjZtm0bfvnlF3Tu3Lk4+iEiIvAD2YjkkP3WbGNjY1StWrU4eiEiIiKSTXaYmTRpEpYtW6bTzbtERERExU32Zaa//voLJ06cwMGDB1GzZk0YGRlpPb9z584ia46IiIioILLDTLly5eDr61scvRARERHJJjvMBAcHF0cfRERERIUi+54ZAEhPT8fRo0exatUqJCcnAwAePHiAlJSUIm2OiIiIqCCyz8zExMSgY8eOuHv3LtRqNby9vWFhYYFFixbh5cuXWLlyZXH0SURERJQr2Wdmxo0bh4YNG+LZs2cwNTXVjPv6+uLYsWNF2hwRERFRQQr1bqZTp07B2NhYa9zZ2Rn3798vssaIiIiIdCH7zExmZiYyMjJyjN+7d4/fZE1ERETvnOww4+3tjaVLl2qWJUlCSkoKZs2axa84ICIiondO9mWm77//Hq1bt0aNGjXw8uVL9OvXD5GRkahYsSK2bt1aHD0SERER5Ul2mHFwcEB4eDi2bt2KixcvIjMzE8OGDUP//v21bggmIiIiehdkhxkAMDU1xdChQzF06NCi7oeIiIhIFtlhZuPGjfk+P2jQoEI3Q0RERCSX7DAzbtw4reVXr17h+fPnMDY2hpmZGcMMERERvVOy38307NkzrUdKSgpu3LiBFi1a8AZgIiIieucKdc/Mm9zc3LBgwQIMGDAA//vf/4qiJBGRYrl8s79Q20Uv6FLEnRC9Hwr1RZO5MTAwwIMHD4qqHBEREZFOZJ+Z2bNnj9ayEAJxcXFYvnw5mjdvXmSNEREREelCdpjp3r271rIkSahUqRLatGmDJUuWFFVfRERERDqRHWYyMzOLow8iIiKiQimye2aIiIiI9EH2mZmJEyfqvG5QUJDc8kRERESyyA4zly5dwsWLF5Geng4PDw8AwM2bN2FgYID69etr1pMkqei6JCIiIsqD7DDj4+MDCwsLbNiwAeXLlwfw+oP0hgwZgpYtW2LSpElF3iQRERFRXmTfM7NkyRIEBgZqggwAlC9fHvPmzeO7mYiIiOidkx1mkpKS8PDhwxzjCQkJSE5OLpKmiIiIiHQlO8z4+vpiyJAh+O2333Dv3j3cu3cPv/32G4YNG4YePXoUR49EREREeZJ9z8zKlSsxefJkDBgwAK9evXpdxNAQw4YNw+LFi4u8QSIiIqL8yA4zZmZm+Omnn7B48WLcvn0bQghUrVoV5ubmxdEfERERUb4K/aF5cXFxiIuLg7u7O8zNzSGEkF3jjz/+gI+PDxwcHCBJEnbt2qX1/ODBgyFJktajSZMmhW2ZiIiISiHZYebJkydo27Yt3N3d0blzZ8TFxQEAhg8fLvtt2ampqahbty6WL1+e5zodO3bUBKe4uDgcOHBAbstERERUism+zDRhwgQYGRnh7t27qF69uma8T58+mDBhgqy3Z3fq1AmdOnXKdx2VSgU7Ozu5bRIREdF7QnaYOXz4MA4dOgRHR0etcTc3N8TExBRZY1lOnjwJGxsblCtXDp6enpg/fz5sbGzyXF+tVkOtVmuWk5KSirwnIiIiKjlkh5nU1FSYmZnlGH/8+DFUKlWRNJWlU6dO6N27N5ydnREVFYVvv/0Wbdq0wYULF/LcV2BgIGbPnl2kfRARvc9cvtlfqO2iF3Qp4k6KTmk8pveZ7HtmWrVqhY0bN2qWJUlCZmYmFi9ejNatWxdpc3369EGXLl1Qq1Yt+Pj44ODBg7h58yb27897Ek6bNg2JiYmaR2xsbJH2RERERCWL7DMzixcvhpeXF86fP4+0tDRMmTIF165dw9OnT3Hq1Kni6FHD3t4ezs7OiIyMzHMdlUpV5GeIiIiIqOSSfWamRo0auHLlCj7++GN4e3sjNTUVPXr0wKVLl/Dhhx8WR48aT548QWxsLOzt7Yt1P0RERKQcss7MvHr1Cu3bt8eqVauK5L6UlJQU3Lp1S7McFRWF8PBwWFtbw9raGv7+/ujZsyfs7e0RHR2N6dOno2LFivD19X3rfRMREVHpICvMGBkZ4erVq5AkqUh2fv78ea37bCZOnAgA8PPzw4oVKxAREYGNGzfi33//hb29PVq3bo3t27fDwsKiSPZPREREyif7nplBgwZh7dq1WLBgwVvv3MvLK99PDj506NBb74OIiIhKN9lhJi0tDWvWrMGRI0fQsGHDHN/JFBQUVGTNERERERVEdpi5evUq6tevDwC4efOm1nNFdfmJiIiISFc6h5k7d+7A1dUVJ06cKM5+iIiIiGTR+a3Zbm5uePTokWa5T58+ePjwYbE0RURERKQrncPMmzfqHjhwAKmpqUXeEBEREZEcsj80j4iIiKgk0TnMSJKU4wZf3vBLRERE+qbzDcBCCAwePFjzvUcvX77EiBEjcrw1e+fOnUXbIREREVE+dA4zfn5+WssDBgwo8maIiIiI5NI5zAQHBxdnH0RERESFwhuAiYiISNEYZoiIiEjRGGaIiIhI0RhmiIiISNEYZoiIiEjRGGaIiIhI0RhmiIiISNEYZoiIiEjRGGaIiIhI0RhmiIiISNEYZoiIiEjRGGaIiIhI0RhmiIiISNEYZoiIiEjRGGaIiIhI0RhmiIiISNEYZoiIiEjRGGaIiIhI0RhmiIiISNEYZoiIiEjRGGaIiIhI0RhmiIiISNEYZoiIiEjRGGaIiIhI0fQaZv744w/4+PjAwcEBkiRh165dWs8LIeDv7w8HBweYmprCy8sL165d00+zREREVCLpNcykpqaibt26WL58ea7PL1q0CEFBQVi+fDnCwsJgZ2cHb29vJCcnv+NOiYiIqKQy1OfOO3XqhE6dOuX6nBACS5cuxYwZM9CjRw8AwIYNG2Bra4stW7bgyy+/fJetEhERUQlVYu+ZiYqKQnx8PNq3b68ZU6lU8PT0xOnTp/PcTq1WIykpSetBREREpVeJDTPx8fEAAFtbW61xW1tbzXO5CQwMhJWVlebh5ORUrH0SERGRfpXYMJNFkiStZSFEjrHspk2bhsTERM0jNja2uFskIiIiPdLrPTP5sbOzA/D6DI29vb1mPCEhIcfZmuxUKhVUKlWx90dEREQlQ4k9M+Pq6go7OzscOXJEM5aWlobQ0FA0a9ZMj50RERFRSaLXMzMpKSm4deuWZjkqKgrh4eGwtrZG5cqVMX78eAQEBMDNzQ1ubm4ICAiAmZkZ+vXrp8euiYiIqCTRa5g5f/48WrdurVmeOHEiAMDPzw/r16/HlClT8OLFC4wcORLPnj1D48aNcfjwYVhYWOirZSIiIiph9BpmvLy8IITI83lJkuDv7w9/f/931xQREREpSom9Z4aIiIhIFwwzREREpGgMM0RERKRoDDNERESkaAwzREREpGgMM0RERKRoDDNERESkaAwzREREpGgMM0RERKRoDDNERESkaAwzREREpGgMM0RERKRoDDNERESkaAwzREREpGgMM0RERKRoDDNERESkaAwzREREpGgMM0RERKRoDDNERESkaAwzREREpGgMM0RERKRoDDNERESkaAwzREREpGgMM0RERKRoDDNERESkaAwzREREpGgMM0RERKRoDDNERESkaAwzREREpGgMM0RERKRoDDNERESkaAwzREREpGgMM0RERKRoDDNERESkaCU6zPj7+0OSJK2HnZ2dvtsiIiKiEsRQ3w0UpGbNmjh69Khm2cDAQI/dEBERUUlT4sOMoaEhz8YQERFRnkr0ZSYAiIyMhIODA1xdXdG3b1/cuXNH3y0RERFRCVKiz8w0btwYGzduhLu7Ox4+fIh58+ahWbNmuHbtGipUqJDrNmq1Gmq1WrOclJT0rtolIiIiPSjRZ2Y6deqEnj17onbt2mjXrh32798PANiwYUOe2wQGBsLKykrzcHJyelftEhERkR6U6DDzJnNzc9SuXRuRkZF5rjNt2jQkJiZqHrGxse+wQyIiInrXSvRlpjep1Wpcv34dLVu2zHMdlUoFlUr1DrsiIiIifSrRZ2YmT56M0NBQREVF4e+//0avXr2QlJQEPz8/fbdGREREJUSJPjNz7949fPbZZ3j8+DEqVaqEJk2a4OzZs3B2dtZ3a0RERFRClOgws23bNn23QERERCVcib7MRERERFQQhhkiIiJSNIYZIiIiUjSGGSIiIlI0hhkiIiJSNIYZIiIiUjSGGSIiIlI0hhkiIiJSNIYZIiIiUjSGGSIiIlI0hhkiIiJSNIYZIiIiUjSGGSIiIlI0hhkiIiJSNIYZIiIiUjSGGSIiIlI0hhkiIiJSNIYZIiIiUjSGGSIiIlI0hhkiIiJSNIYZIiIiUjSGGSIiIlI0hhkiIiJSNIYZIiIiUjSGGSIiIlI0hhkiIiJSNIYZIiIiUjSGGSIiIlI0hhkiIiJSNIYZIiIiUjSGGSIiIlI0hhkiIiJSNIYZIiIiUjSGGSIiIlI0RYSZn376Ca6urjAxMUGDBg3w559/6rslIiIiKiFKfJjZvn07xo8fjxkzZuDSpUto2bIlOnXqhLt37+q7NSIiIioBSnyYCQoKwrBhwzB8+HBUr14dS5cuhZOTE1asWKHv1oiIiKgEKNFhJi0tDRcuXED79u21xtu3b4/Tp0/rqSsiIiIqSQz13UB+Hj9+jIyMDNja2mqN29raIj4+Ptdt1Go11Gq1ZjkxMREAkJSUVCw9ZqqfF2q74uinKHopKTXYC3thL+yFvSijl+L6+5pVVwhR8MqiBLt//74AIE6fPq01Pm/ePOHh4ZHrNrNmzRIA+OCDDz744IOPUvCIjY0tMC+U6DMzFStWhIGBQY6zMAkJCTnO1mSZNm0aJk6cqFnOzMzE06dPUaFCBUiSVKz9ZpeUlAQnJyfExsbC0tLyne23uHopKTXYC3thL+yFvSivRmEIIZCcnAwHB4cC1y3RYcbY2BgNGjTAkSNH4Ovrqxk/cuQIunXrlus2KpUKKpVKa6xcuXLF2Wa+LC0t9R5mshRFLyWlBnthL+yFvbAX5dWQy8rKSqf1SnSYAYCJEydi4MCBaNiwIZo2bYrVq1fj7t27GDFihL5bIyIiohKgxIeZPn364MmTJ5gzZw7i4uJQq1YtHDhwAM7OzvpujYiIiEqAEh9mAGDkyJEYOXKkvtuQRaVSYdasWTkueSm1l5JSg72wF/bCXtiL8moUN0kIXd7zRERERFQylegPzSMiIiIqCMMMERERKRrDDBERESkawwwREREpGsNMEfvjjz/g4+MDBwcHSJKEXbt26aWPFStWoE6dOpoPOWratCkOHjz4VjUDAwMhSRLGjx8vazt/f39IkqT1sLOzK1QP9+/fx4ABA1ChQgWYmZnho48+woULF3Te3sXFJUcvkiRh1KhROtdIT0/HzJkz4erqClNTU1SpUgVz5sxBZmam7ONJTk7G+PHj4ezsDFNTUzRr1gxhYWH5blPQHBNCwN/fHw4ODjA1NYWXlxeuXbsmq8bOnTvRoUMHVKxYEZIkITw8XFYfr169wtSpU1G7dm2Ym5vDwcEBgwYNwoMHD2Qfj7+/P6pVqwZzc3OUL18e7dq1w99//y2rRnZffvklJEnC0qVLZfcyePDgHHOnSZMmsnu5fv06unbtCisrK1hYWKBJkya4e/eurDq5zWNJkrB48WKda6SkpGD06NFwdHSEqakpqlevjhUrVsh+XR4+fIjBgwfDwcEBZmZm6NixIyIjI7XWCQwMRKNGjWBhYQEbGxt0794dN27c0FqnoLmrSw1d5m5BdXSZv7r0osvc1aVOdrnNX11q6DJ3de1Fl/n7rjHMFLHU1FTUrVsXy5cv12sfjo6OWLBgAc6fP4/z58+jTZs26NatW44/aroKCwvD6tWrUadOnUJtX7NmTcTFxWkeERERsms8e/YMzZs3h5GREQ4ePIh//vkHS5YskfUJz2FhYVp9HDlyBADQu3dvnWssXLgQK1euxPLly3H9+nUsWrQIixcvxg8//CD3kDB8+HAcOXIEmzZtQkREBNq3b4927drh/v37eW5T0BxbtGgRgoKCsHz5coSFhcHOzg7e3t5ITk7WuUZqaiqaN2+OBQsWFKqP58+f4+LFi/j2229x8eJF7Ny5Ezdv3kTXrl1lH4+7uzuWL1+OiIgI/PXXX3BxcUH79u3x6NEjnWtk2bVrF/7+++88Px5dlzodO3bUmkMHDhyQVeP27dto0aIFqlWrhpMnT+Ly5cv49ttvYWJiIqtO9h7i4uKwbt06SJKEnj176lxjwoQJCAkJwebNm3H9+nVMmDABY8aMwe7du3XuRQiB7t27486dO9i9ezcuXboEZ2dntGvXDqmpqZr1QkNDMWrUKJw9exZHjhxBeno62rdvr7VOQXNXlxq6zN2C6ugyf3XpRZe5q0udLHnNX11rFDR3damj6/x9597yuyApHwDE77//ru82NMqXLy/WrFkje7vk5GTh5uYmjhw5Ijw9PcW4ceNkbT9r1ixRt25d2ft909SpU0WLFi3euk5248aNEx9++KHIzMzUeZsuXbqIoUOHao316NFDDBgwQNa+nz9/LgwMDMS+ffu0xuvWrStmzJihU40351hmZqaws7MTCxYs0Iy9fPlSWFlZiZUrV+pUI7uoqCgBQFy6dElWH7k5d+6cACBiYmLeqk5iYqIAII4ePSqrxr1798QHH3wgrl69KpydncX333+f735yq+Pn5ye6deuW73YF1ejTp4/suaLL69KtWzfRpk0bWTVq1qwp5syZozVWv359MXPmTJ3r3LhxQwAQV69e1Yylp6cLa2tr8fPPP+dZJyEhQQAQoaGhQojCzd03a2Sn69wtqE6WguavLjUKmrv51ZEzf3OrIXfu5lWnMPP3XeCZmfdARkYGtm3bhtTUVDRt2lT29qNGjUKXLl3Qrl27QvcQGRkJBwcHuLq6om/fvrhz547sGnv27EHDhg3Ru3dv2NjYoF69evj5558L3VNaWho2b96MoUOHyvoS0hYtWuDYsWO4efMmAODy5cv466+/0LlzZ1n7T09PR0ZGRo5/0ZiamuKvv/6SVStLVFQU4uPj0b59e82YSqWCp6cnTp8+XaiaRSUxMRGSJL3Vd6WlpaVh9erVsLKyQt26dXXeLjMzEwMHDsTXX3+NmjVrFnr/AHDy5EnY2NjA3d0dn3/+ORISEmT1sX//fri7u6NDhw6wsbFB48aN3/py9MOHD7F//34MGzZM1nYtWrTAnj17cP/+fQghcOLECdy8eRMdOnTQuYZarQYArXlsYGAAY2PjfOdxYmIiAMDa2hpA4ebumzUKS5c6Bc3fgmroOndzqyN3/ubVi9y5+2ad4pq/RULfaao0g57PzFy5ckWYm5sLAwMDYWVlJfbv3y+7xtatW0WtWrXEixcvhBCiUGdmDhw4IH777Tdx5coVzdkdW1tb8fjxY1l1VCqVUKlUYtq0aeLixYti5cqVwsTERGzYsEFWnSzbt28XBgYG4v79+7K2y8zMFN98842QJEkYGhoKSZJEQEBAoXpo2rSp8PT0FPfv3xfp6eli06ZNQpIk4e7urtP2b86xU6dOCQA5junzzz8X7du316lGdkV1ZubFixeiQYMGon///oWqs3fvXmFubi4kSRIODg7i3LlzsmoEBAQIb29vzRm4wp6Z2bZtm9i3b5+IiIgQe/bsEXXr1hU1a9YUL1++1KlGXFycACDMzMxEUFCQuHTpkggMDBSSJImTJ0/K6iW7hQsXivLly2t+T3WtoVarxaBBgwQAYWhoKIyNjcXGjRvzrJFbnbS0NOHs7Cx69+4tnj59KtRqtQgMDBQA8pxzmZmZwsfHR+tMq9y5m1uN7HSduwXVEaLg+ZtfDTlzN686cuZvXjXkzt3c6hR2/r4LDDPFSN9hRq1Wi8jISBEWFia++eYbUbFiRXHt2jWdt797966wsbER4eHhmrHChJk3paSkCFtbW7FkyRJZ2xkZGYmmTZtqjY0ZM0Y0adKkUH20b99efPLJJ7K327p1q3B0dBRbt24VV65cERs3bhTW1tZi/fr1smvdunVLtGrVSgAQBgYGolGjRqJ///6ievXqOm2fV5h58OCB1nrDhw8XHTp00KlGdkURZtLS0kS3bt1EvXr1RGJiYqHqpKSkiMjISHHmzBkxdOhQ4eLiIh4+fKhTjfPnzwtbW1utP5KFDTNvevDggTAyMhI7duzQqcb9+/cFAPHZZ59prefj4yP69u1b6F48PDzE6NGj8+01txqLFy8W7u7uYs+ePeLy5cvihx9+EGXLlhVHjhyRVef8+fOibt26mnncoUMH0alTJ9GpU6dca4wcOVI4OzuL2NhYzZjcuZtbjex0nbsF1dFl/uZXQ87cza2O3Plb0PFkKWju5lansPP3XWCYKUb6DjNvatu2rfjiiy90Xv/333/X/M8p6wFASJIkDAwMRHp6eqF7adeunRgxYoSsbSpXriyGDRumNfbTTz8JBwcH2fuPjo4WZcqUEbt27ZK9raOjo1i+fLnW2Ny5c4WHh4fsWllSUlI0/xP/9NNPRefOnXXa7s05dvv2bQFAXLx4UWu9rl27ikGDBulUI7u3DTNpaWmie/fuok6dOjqdidP1d6Zq1ap5ng17s8b333+vmbPZ53GZMmWEs7NzkfSS/T6P/Gqo1WphaGgo5s6dq7XelClTRLNmzQrVyx9//CEAaP2jQ5caz58/F0ZGRjnu2Ro2bFiewbegXv7991+RkJAghBDi448/FiNHjsyxzujRo4Wjo6O4c+eO1ricuZtXjex0mbsF1dFl/urSS3Z5zd286siZv4XpJbe5m1edws7fd4H3zLxHhBCa69u6aNu2LSIiIhAeHq55NGzYEP3790d4eDgMDAwK1Ydarcb169dhb28va7vmzZvneJvgzZs3C/UN6sHBwbCxsUGXLl1kb/v8+XOUKaP9q2NgYFCot2ZnMTc3h729PZ49e4ZDhw6hW7duharj6uoKOzs7zbu0gNfX6kNDQ9GsWbNC91cYr169wqefforIyEgcPXoUFSpUKLLacubywIEDceXKFa157ODggK+//hqHDh16qz6ePHmC2NhYneeysbExGjVqVGTzGADWrl2LBg0ayLqHCHj983n16lWRzmUrKytUqlQJkZGROH/+vNY8FkJg9OjR2LlzJ44fPw5XV1etbXWZuwXV0JUudQqav4Xt5c25W1AdXeZvYXrJbe4WVKc45m9RUcS3ZitJSkoKbt26pVmOiopCeHg4rK2tUbly5XfWx/Tp09GpUyc4OTkhOTkZ27Ztw8mTJxESEqJzDQsLC9SqVUtrzNzcHBUqVMgxnp/JkyfDx8cHlStXRkJCAubNm4ekpCT4+fnpXAN4/TbSZs2aISAgAJ9++inOnTuH1atXY/Xq1bLqZGZmIjg4GH5+fjA0lP8r4OPjg/nz56Ny5cqoWbMmLl26hKCgIAwdOlR2rUOHDkEIAQ8PD9y6dQtff/01PDw8MGTIkDy3KWiOjR8/HgEBAXBzc4ObmxsCAgJgZmaGfv366Vzj6dOnuHv3ruZzNbL+52VnZ6f5jKD8ajg4OKBXr164ePEi9u3bh4yMDMTHxwN4fTOhsbGxTr1UqFAB8+fPR9euXWFvb48nT57gp59+wr1797TeTl/Q8bz5h8jIyAh2dnbw8PDQ+bW1traGv78/evbsCXt7e0RHR2P69OmoWLEifH19de7l66+/Rp8+fdCqVSu0bt0aISEh2Lt3L06ePCnr5wwASUlJ+PXXX7FkyRLkpqAanp6e+Prrr2FqagpnZ2eEhoZi48aNCAoKklXn119/RaVKlVC5cmVERERg3Lhx6N69u9bNvKNGjcKWLVuwe/duWFhYaOaDlZUVTE1NNZ9hld/cLagGAJ3mbkF10tPTC5y/BdVITU3Vae4WVKdChQoFzt+CaqSkpOg0d3V5fXWdv++cns4IlVonTpwQAHI8/Pz83mkfQ4cOFc7OzsLY2FhUqlRJtG3bVhw+fPit6xbmnpk+ffoIe3t7YWRkJBwcHESPHj1k3buT3d69e0WtWrWESqUS1apVE6tXr5Zd49ChQwKAuHHjRqF6SEpKEuPGjROVK1cWJiYmokqVKmLGjBlCrVbLrrV9+3ZRpUoVYWxsLOzs7MSoUaPEv//+m+82Bc2xzMxMMWvWLGFnZydUKpVo1aqViIiIkFUjODg41+dnzZqlU42sU/y5PU6cOKFzLy9evBC+vr7CwcFBGBsbC3t7e9G1a9ccN1HK/b3L656D/Oo8f/5ctG/fXlSqVEkYGRmJypUrCz8/P3H37l3Zvaxdu1ZUrVpVmJiYiLp16+Z6uVOXOqtWrRKmpqZ5zpmCasTFxYnBgwcLBwcHYWJiIjw8PMSSJUtyfFRBQXWWLVsmHB0dNa/LzJkzc/w+5DUfgoODNesUNHd1qaHL3C2oji7zt6Aaus5dXY7pTW/O34Jq6Dp3de1Fl/n7rklCCAEiIiIiheI9M0RERKRoDDNERESkaAwzREREpGgMM0RERKRoDDNERESkaAwzREREpGgMM0RERKRoDDP03pEkqWR8Zb3CeHl5Yfz48e9kX+vXr0e5cuVkbTN48GB07969WPrJzerVq+Hk5IQyZcpg6dKlOZ6Pjo6GJEkIDw9/Zz0Rva8YZqhUGDx4MCRJgiRJMDIygq2tLby9vbFu3boc3zMTFxeHTp066VRXScGnqP6Ynzx5EpIk4d9//33rWoXVp08f3Lx5s8jruri45Bo85EpKSsLo0aMxdepU3L9/H1988cXbN/eeYMij4sAwQ6VGx44dERcXh+joaBw8eBCtW7fGuHHj8MknnyA9PV2znp2dHVQqlR47pfy8evUKpqamsLGx0Xcrebp79y5evXqFLl26wN7eHmZmZvpuSbZXr17pu4W3VhqOgYoGwwyVGiqVCnZ2dvjggw9Qv359TJ8+Hbt378bBgwexfv16zXrZz7akpaVh9OjRsLe3h4mJCVxcXBAYGAjg9b/iAcDX1xeSJGmWb9++jW7dusHW1hZly5ZFo0aNcPToUa1eXFxcEBAQgKFDh8LCwgKVK1fO8YWY9+7dQ9++fWFtbQ1zc3M0bNgQf//9t+b5vXv3okGDBjAxMUGVKlUwe/ZsrVCWnb+/PzZs2IDdu3drzlBlffFbREQE2rRpo/nSui+++AIpKSm51omOjkbr1q0BAOXLl4ckSRg8eLDm+czMTEyZMgXW1taws7ODv7+/1vaJiYn44osvYGNjA0tLS7Rp0waXL1/OdV9Z+5MkCb/88gu8vLxgYmKCzZs353qZad68ebCxsYGFhQWGDx+Ob775Bh999FGOmt999x3s7e1RoUIFjBo1SvMHz8vLCzExMZgwYYLmNcrL3bt30a1bN5QtWxaWlpb49NNP8fDhQwCvL4HVrl0bAFClShVIkoTo6Og8a2V/7T7//HO4u7sjJiamwNejZcuWMDU1RaNGjXDz5k2EhYWhYcOGKFu2LDp27IhHjx5pbRscHIzq1avDxMQE1apVw08//ZRr3eyvc27+/fdffPHFF7C1tYWJiQlq1aqFffv2aZ4/ffo0WrVqBVNTUzg5OWHs2LFITU3VPF/Q3M/6JuZ69epBkiR4eXkV+THQe0jfXw5FVBT8/PxEt27dcn2ubt26olOnTpplAOL3338XQgixePFi4eTkJP744w8RHR0t/vzzT7FlyxYhhBAJCQmaL1mLi4sTCQkJQgghwsPDxcqVK8WVK1fEzZs3xYwZM4SJiYmIiYnR7MPZ2VlYW1uLH3/8UURGRorAwEBRpkwZcf36dSGEEMnJyaJKlSqiZcuW4s8//xSRkZFi+/bt4vTp00IIIUJCQoSlpaVYv369uH37tjh8+LBwcXER/v7+uR5jcnKy+PTTT0XHjh1FXFyciIuLE2q1WqSmpmq+3DMiIkIcO3ZMuLq65vkFjOnp6WLHjh2aL+KMi4vTfImhp6ensLS0FP7+/uLmzZtiw4YNQpIkzReYZmZmiubNmwsfHx8RFhYmbt68KSZNmiQqVKggnjx5kuv+sr7Qz8XFRezYsUPcuXNH3L9/XwQHBwsrKyvNeps3bxYmJiZi3bp14saNG2L27NnC0tJS1K1bV7OOn5+fsLS0FCNGjBDXr18Xe/fuFWZmZpovI33y5IlwdHQUc+bM0bxGucnMzBT16tUTLVq0EOfPnxdnz54V9evXF56enkKI11/ad/ToUQFAnDt3TsTFxYn09PQ8j+3SpUtCrVaLnj17io8++kg8fPgw1/1m36ZatWoiJCRE/PPPP6JJkyaifv36wsvLS/z111/i4sWLomrVqmLEiBGa7VavXi3s7e01r+GOHTuEtbW1WL9+fb6v85syMjJEkyZNRM2aNcXhw4fF7du3xd69e8WBAweEEEJcuXJFlC1bVnz//ffi5s2b4tSpU6JevXpi8ODBmhoFzf1z584JAOLo0aMiLi5OMzeK6hjo/cQwQ6VCfmGmT58+onr16prl7GFmzJgxok2bNjm+JTi3dfNTo0YN8cMPP2iWnZ2dxYABAzTLmZmZwsbGRqxYsUII8frbji0sLPL8I9+yZUsREBCgNbZp0yZhb2+fZw+5vQarV68W5cuXFykpKZqx/fv3izJlyoj4+Phc62R9Q/KzZ8+0xj09PUWLFi20xho1aiSmTp0qhBDi2LFjwtLSUrx8+VJrnQ8//FCsWrUq131l/YFaunSp1vibYaZx48Zi1KhRWus0b948R5hxdnbWCha9e/cWffr00Szn9W3Z2R0+fFgYGBhofaPwtWvXNOFFCCEuXbokAIioqKg862Qd259//inatWsnmjdvXuA3omdts2bNGs3Y1q1bBQBx7NgxzVhgYKDw8PDQLDs5OWlCeJa5c+eKpk2batV983V+06FDh0SZMmXy/Eb5gQMHii+++EJr7M8//xRlypQRL168EEIUPPezh7zsiuoY6P3Ey0xU6gkh8rykMHjwYISHh8PDwwNjx47F4cOHC6yXmpqKKVOmoEaNGihXrhzKli2L//3vf7h7967WenXq1NH8tyRJsLOzQ0JCAgAgPDwc9erVg7W1da77uHDhAubMmYOyZctqHp9//jni4uLw/PlzXQ8d169fR926dWFubq4Za968OTIzM3Hjxg2d6+R2TABgb2+vOaYLFy4gJSUFFSpU0Oo7KioKt2/fzrduw4YN833+xo0b+Pjjj7XG3lwGgJo1a8LAwCDX/nR1/fp1ODk5wcnJSTOW9bO+fv26rFoA8NlnnyElJQWHDx+GlZWVZnzEiBFar1N22V9nW1tbANBc2soayzquR48eITY2FsOGDdOqN2/evByve0Gvc3h4OBwdHeHu7p7r8xcuXMD69eu19tOhQwdkZmYiKioq1/7fnPu5KcpjoPeTob4bICpu169f11ynf1P9+vURFRWFgwcP4ujRo/j000/Rrl07/Pbbb3nW+/rrr3Ho0CF89913qFq1KkxNTdGrVy+kpaVprWdkZKS1LEmS5p1Vpqam+facmZmJ2bNno0ePHjmeMzExyXfb7PILcvndM5KX/I4pMzMT9vb2mnt1sivobdbZw1Ze3uxXCCGrP13l9Zrl91rmp3Pnzti8eTPOnj2LNm3aaMbnzJmDyZMn57pN9uPI2uebY9lfdwD4+eef0bhxY6062YMdUPDrrMu8/PLLLzF27Ngcz1WuXDnX/t/sN6+6QNEcA72fGGaoVDt+/DgiIiIwYcKEPNextLREnz590KdPH/Tq1QsdO3bE06dPYW1tDSMjI2RkZGit/+eff2Lw4MHw9fUFAKSkpOh0A2h2derUwZo1azT7eVP9+vVx48YNVK1aVeeaxsbGOXqtUaMGNmzYgNTUVM0fgVOnTqFMmTJ5/uvb2NgYAHLUKkj9+vURHx8PQ0NDzc3SRcXDwwPnzp3DwIEDNWPnz5+XXSe31+hNNWrUwN27dxEbG6s5O/PPP/8gMTER1atXl73Pr776CrVq1ULXrl2xf/9+eHp6AgBsbGyK5B1btra2+OCDD3Dnzh3079//rWrVqVMH9+7dw82bN3OdH/Xr18e1a9dkzcs35Ta/ivIY6P3Ey0xUaqjVasTHx+P+/fu4ePEiAgIC0K1bN3zyyScYNGhQrtt8//332LZtG/73v//h5s2b+PXXX2FnZ6c5k+Di4oJjx44hPj4ez549AwBUrVoVO3fuRHh4OC5fvox+/frJ/tf/Z599Bjs7O3Tv3h2nTp3CnTt3sGPHDpw5cwYA8J///AcbN26Ev78/rl27huvXr2P79u2YOXNmnjVdXFxw5coV3LhxA48fP8arV6/Qv39/mJiYwM/PD1evXsWJEycwZswYDBw4UHP54k3Ozs6QJAn79u3Do0eP8nzn05vatWuHpk2bonv37jh06BCio6Nx+vRpzJw5s1DBI7sxY8Zg7dq12LBhAyIjIzFv3jxcuXJF9pkSFxcX/PHHH7h//z4eP36c53HUqVMH/fv3x8WLF3Hu3DkMGjQInp6ehb7EMWbMGMybNw+ffPIJ/vrrr0LVyI+/vz8CAwOxbNky3Lx5ExEREQgODkZQUJCsOp6enmjVqhV69uyJI0eOaM5ahoSEAACmTp2KM2fOYNSoUQgPD0dkZCT27NmDMWPG6LwPGxsbmJqaIiQkBA8fPkRiYmKRHgO9nxhmqNQICQmBvb09XFxc0LFjR5w4cQL//e9/sXv37hynqrOULVsWCxcuRMOGDdGoUSNER0fjwIEDKFPm9a/GkiVLcOTIETg5OaFevXoAXgeg8uXLo1mzZvDx8UGHDh1Qv359Wb0aGxvj8OHDsLGxQefOnVG7dm0sWLBA02eHDh2wb98+HDlyBI0aNUKTJk0QFBQEZ2fnPGt+/vnn8PDwQMOGDVGpUiWcOnUKZmZmOHToEJ4+fYpGjRqhV69eaNu2LZYvX55nnQ8++ACzZ8/GN998A1tbW4wePVqnY5IkCQcOHECrVq0wdOhQuLu7o2/fvoiOjs4zOOmqf//+mDZtGiZPnqy5NDh48GBZl9yA15d2oqOj8eGHH6JSpUp5HseuXbtQvnx5tGrVCu3atUOVKlWwffv2tzqG8ePHY/bs2ejcuTNOnz79VrXeNHz4cKxZs0bztnFPT0+sX78+z8ur+dmxYwcaNWqEzz77DDVq1MCUKVM0Z1Hq1KmD0NBQREZGomXLlqhXrx6+/fZb2Nvb61zf0NAQ//3vf7Fq1So4ODigW7duRX4M9P6RRG4XnomISjhvb2/Y2dlh06ZN+m6FiPSM98wQUYn3/PlzrFy5Eh06dICBgQG2bt2Ko0eP4siRI/pujYhKAJ6ZIaIS78WLF/Dx8cHFixehVqvh4eGBmTNn5vpuLyJ6/zDMEBERkaLxBmAiIiJSNIYZIiIiUjSGGSIiIlI0hhkiIiJSNIYZIiIiUjSGGSIiIlI0hhkiIiJSNIYZIiIiUjSGGSIiIlK0/w8SrvaS8CgGXwAAAABJRU5ErkJggg==",
      "text/plain": [
       "<Figure size 640x480 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "array([[ 1, 39],\n",
       "       [ 3, 25],\n",
       "       [ 4, 28],\n",
       "       [ 5, 33],\n",
       "       [ 6, 31],\n",
       "       [ 7, 28],\n",
       "       [ 8, 29],\n",
       "       [ 9, 31],\n",
       "       [10, 29],\n",
       "       [11, 34],\n",
       "       [12, 20],\n",
       "       [13, 28],\n",
       "       [14, 29],\n",
       "       [15, 25],\n",
       "       [16, 17],\n",
       "       [17, 19],\n",
       "       [18, 29],\n",
       "       [19, 15],\n",
       "       [20, 27],\n",
       "       [21, 27],\n",
       "       [22, 15],\n",
       "       [23, 20],\n",
       "       [24, 27],\n",
       "       [25, 30],\n",
       "       [26, 25]])"
      ]
     },
     "execution_count": 75,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def frequency(k, kmers, model):\n",
    "    \"\"\"\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    k : int\n",
    "        an odd number that specifies that length of the kmer with the\n",
    "        phosphorylation site in the middle\n",
    "    kmer : 2D array\n",
    "        1D array with kmers\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "    frequencies : 2D array\n",
    "        2D array with columns distance away from center kmer PTM and \n",
    "        frequency of nearby PTMS\n",
    "\n",
    "    \"\"\"\n",
    "    around_ind = k//2\n",
    "    x=0\n",
    "    frequencies = {}\n",
    "    for i in range(kmers.shape[0]):\n",
    "        if \"@\" in kmers[i]:\n",
    "            x+=1\n",
    "        temp = kmers[i][around_ind+1:]\n",
    "        for j in range(len(temp)):\n",
    "            if \"@\" == temp[j]:\n",
    "                if j+1 in frequencies.keys():\n",
    "                    frequencies[j+1] += 1\n",
    "                else:\n",
    "                    frequencies[j+1] = 1\n",
    "    \n",
    "    frequencies = list(frequencies.items())\n",
    "    frequencies = np.array(frequencies)\n",
    "    print (frequencies)\n",
    "    frequencies = frequencies[frequencies[:,0].argsort()]\n",
    "\n",
    "    plt.figure(0)\n",
    "    plt.bar(frequencies[:,0], frequencies[:,1])\n",
    "    plt.xticks(frequencies[:,0])\n",
    "    plt.title(\"Frequency of nearby PTM sites\")\n",
    "    plt.ylabel(\"Frequency amount\")\n",
    "    plt.xlabel(\"Distance to the right of k-mer center\")\n",
    "    plt.savefig(model+\"_frequencies.pdf\", transparent=True)\n",
    "    plt.show()\n",
    "\n",
    "    return frequencies\n",
    "\n",
    "\n",
    "full_trainl, full_testl=labeling(dataset, protein_dict, 24176)\n",
    "full_datasetl=np.concatenate((full_trainl, full_testl))\n",
    "frequency (53, full_datasetl[:,0], model=\"NGlycoSite\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "16746bcf-db8f-45ea-bb40-63f1e8767089",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "95a4d3af-1455-4421-bf2f-44d5ebc002d8",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
